{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "#make files if don't already exist\n",
    "os.makedirs(\"profile_o\", exist_ok=True)\n",
    "os.makedirs(\"sorted_reads\", exist_ok=True)\n",
    "os.makedirs(\"seq_reads\", exist_ok=True)\n",
    "os.makedirs(\"summary_o\", exist_ok=True)\n",
    "os.makedirs(\"classifier\", exist_ok=True)\n",
    "os.makedirs(\"classifier_training_data\", exist_ok=True)\n",
    "\n",
    "#downstream analysis: https://www.youtube.com/watch?v=9AmM-BHYxJo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/miloscola/augene/training_data/processed/ERR1351807_1.fastq.gz\n",
      "PHRED 33 encoding detected.\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$fastq_file\"\n",
    "\n",
    "#check if files are PHRED33\n",
    "check_fastq_encoding() {\n",
    "    # Check if a file is provided as an argument\n",
    "    if [ $# -ne 1 ]; then\n",
    "        echo \"Usage: $0 <fastq_file>\"\n",
    "        exit 1\n",
    "    fi\n",
    "    \n",
    "    fastq_file=\"$1\"\n",
    "\n",
    "    # Read the first quality line from the FASTQ file\n",
    "    quality_line=$(head -n 40 \"$fastq_file\" | grep -E '^\\+[[:space:]]*$' -A 1 | tail -n 1)\n",
    "\n",
    "    # Loop through each character in the quality line and check its ASCII value\n",
    "    phred33=1\n",
    "    phred64=1\n",
    "    for (( i=0; i<${#quality_line}; i++ )); do\n",
    "        char=$(printf \"%d\" \"'${quality_line:$i:1}\")\n",
    "        if [ $char -lt 33 ] || [ $char -gt 73 ]; then\n",
    "            phred33=0\n",
    "        fi\n",
    "        if [ $char -lt 64 ] || [ $char -gt 104 ]; then\n",
    "            phred64=0\n",
    "        fi\n",
    "    done\n",
    "\n",
    "    # Output the result\n",
    "    if [ $phred33 -eq 1 ]; then\n",
    "        echo \"PHRED 33 encoding detected.\"\n",
    "    elif [ $phred64 -eq 1 ]; then\n",
    "        echo \"PHRED 64 encoding detected.\"\n",
    "    else\n",
    "        echo \"Unknown encoding format.\"\n",
    "    fi\n",
    "}\n",
    "\n",
    "processed=\"$(pwd)/processed\"\n",
    "\n",
    "echo \"$processed/ERR1351807_1.fastq.gz\"\n",
    "\n",
    "check_fastq_encoding \"$processed/ERR1351807_1.fastq.gz\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/miloscola/augene/training_data/seq_reads/*.fastq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "gzip: /home/miloscola/augene/training_data/seq_reads/*.fastq: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "current_dir=$(pwd)\n",
    "\n",
    "#gzip any randomly unzipped files\n",
    "for file in $current_dir/seq_reads/*.fastq; do\n",
    "    # Re-zip each file\n",
    "    gzip -f \"$file\"\n",
    "    #remove unzipped file\n",
    "    #rm \"$file\"\n",
    "done\n",
    "\n",
    "for file in $current_dir/seq_reads/*_R1_001.fastq.gz; do\n",
    "    # Extract the prefix (e.g., ERR1351807) from the filename\n",
    "    prefix=\"${file%_R1_001.fastq.gz}\"\n",
    "    \n",
    "    # Rename the forward read file\n",
    "    mv \"$file\" \"${prefix}_1.fastq.gz\"\n",
    "    \n",
    "done\n",
    "\n",
    "for file in $current_dir/seq_reads/*_R2_001.fastq.gz; do\n",
    "    # Extract the prefix (e.g., ERR1351807) from the filename\n",
    "    prefix=\"${file%_R2_001.fastq.gz}\"\n",
    "    \n",
    "    # Rename the forward read file\n",
    "    mv \"$file\" \"${prefix}_2.fastq.gz\"\n",
    "    \n",
    "done\n",
    "                                                                                                                                    \n",
    "for file in $current_dir/seq_reads/*.fastq; do\n",
    "    if [[ $file == *.fastq ]]; then\n",
    "        echo \"$file\"\n",
    "    fi\n",
    "done "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "#LIST FILES\n",
    "current_dir=$(pwd)\n",
    "\n",
    "for file in $current_dir/seq_reads/*; do\n",
    "    # Extract the prefix (e.g., ERR1351807) from the filename\n",
    "    echo $file\n",
    "    \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## SORT RAW SKINOME READS ##\n",
    "\n",
    "import os\n",
    "import shutil\n",
    "import pandas as pd\n",
    "import glob\n",
    "\n",
    "# Define paths\n",
    "current_dir = os.getcwd()\n",
    "sorted_reads = os.path.join(current_dir, 'sorted_reads')\n",
    "seq_reads = os.path.join(current_dir, 'seq_reads')\n",
    "master_manifest = os.path.join(current_dir, 'DataFrame_2_Pruned.tsv')\n",
    "\n",
    "# load in df\n",
    "df = pd.read_csv(master_manifest, sep='\\t', encoding='latin1')\n",
    "df = df.dropna(subset=['Submission'])\n",
    "submissions = df['Submission'].unique()\n",
    "submissions = submissions.tolist()\n",
    "\n",
    "#create seprate file for each submission\n",
    "for sub in submissions: \n",
    "    new_direc = os.path.join(sorted_reads, sub)\n",
    "    os.makedirs(new_direc, exist_ok=True)\n",
    "    pruned_df = df[df['Submission'] == sub]\n",
    "    p_s = pruned_df['Library_Layout'].unique().tolist()[0]\n",
    "    if p_s == 'PAIRED':\n",
    "        for run in pruned_df['Run'].tolist():\n",
    "            sample_direc = os.path.join(new_direc, run)\n",
    "            sample_1_path = os.path.join(seq_reads, run + '_1.fastq.gz')\n",
    "            sample_2_path = os.path.join(seq_reads, run + '_2.fastq.gz')\n",
    "            \n",
    "            if (not os.path.exists(sample_1_path)) or  (not os.path.exists(sample_2_path)):\n",
    "                \n",
    "                if not os.path.exists(sample_1_path):\n",
    "                    print('file path ' + sample_1_path + ' was not found for paired sample ' + run)\n",
    "                    \n",
    "                if not os.path.exists(sample_1_path):\n",
    "                    print('file path ' + sample_1_path + ' was not found for paired sample ' + run)\n",
    "                \n",
    "                continue\n",
    "                \n",
    "            os.makedirs(sample_direc, exist_ok=True)\n",
    "            shutil.move(sample_1_path, sample_direc)\n",
    "            shutil.move(sample_2_path, sample_direc)\n",
    "    elif p_s == 'SINGLE':\n",
    "        for run in pruned_df['Run'].tolist():\n",
    "            sample_direc = os.path.join(new_direc, run)\n",
    "            sample_1_path = os.path.join(seq_reads, run + '_1.fastq.gz')\n",
    "            \n",
    "            if not os.path.exists(sample_1_path):\n",
    "                print('file path ' + sample_1_path + ' was not found for single sample ' + run)\n",
    "                continue\n",
    "            \n",
    "            os.makedirs(sample_direc, exist_ok=True)\n",
    "            shutil.move(sample_1_path, sample_direc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (1.79)\r\n",
      "Requirement already satisfied: numpy in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from biopython) (1.19.1)\r\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Qiime2 ##\n",
    "\n",
    "import os\n",
    "import shutil\n",
    "import pandas as pd\n",
    "import glob\n",
    "import subprocess\n",
    "import re\n",
    "import qiime2\n",
    "from qiime2.plugins.dada2.methods import denoise_single, denoise_paired\n",
    "from qiime2.plugins.feature_table.visualizers import summarize\n",
    "from qiime2.plugins.feature_classifier.methods import classify_sklearn\n",
    "from qiime2.plugins.taxa.visualizers import barplot\n",
    "from Bio import SeqIO\n",
    "from qiime2.plugins.feature_table.methods import relative_frequency\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from qiime2 import Metadata\n",
    "\n",
    "\n",
    "def qiime2_profile(sub_num, sub_path, master_manifest_path, profile_o_path, summary_o_folder, classifier_map, trim_left_f=13, trim_left_r=13, trunc_len_f=150, trunc_len_r=150, n_threads=8):\n",
    "    \"\"\"\n",
    "    calculates microbiome profiles from a submission of reads and stores the results in a .qza file in the specified folder\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    sub_num : str\n",
    "        ID of the submission containing the reads.\n",
    "    sub_path : str\n",
    "        path to the submission.\n",
    "    master_manifest_path : str\n",
    "        path to the master meta data file (should .tsv file) Manifest must \n",
    "        have the following columns: Run or id, Library_Layout, Region_16S, \n",
    "        Submission\n",
    "    profile_o_path : str\n",
    "        path to the output folder for microbiome profile\n",
    "    summary_o_path : str:\n",
    "        path to the output folder for read summary table\n",
    "    classifier_map : dict\n",
    "        map of 16s rRNA region to classifier built to classify that region\n",
    "    trim_left_f : int\n",
    "        number of nucleotides to trim from the 5' end of the forward read before denoising\n",
    "    trim_left_r : int\n",
    "        number of nucleotides to trim from the 5' end of the reverse read before denoising\n",
    "    trim_left_f : int\n",
    "        specifies the length at which the forward read will be truncated\n",
    "    trim_left_r : int\n",
    "        specifies the length at which the reverse read will be truncated\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    none\n",
    "    \"\"\"\n",
    "    \n",
    "    #define variables\n",
    "    current_dir = os.getcwd()\n",
    "    summary_o_path = os.path.join(summary_o_folder, sub_num)\n",
    "    manifest_path = os.path.join(summary_o_path, f\"{sub_num}-manifest.tsv\")\n",
    "    \n",
    "    #string representing weather reads are paired or single\n",
    "    p_s = None           \n",
    "    #string representing what 16s regions were used\n",
    "    region_16s = None           \n",
    "\n",
    "    #validate inputs\n",
    "    if not type(sub_num) == str:\n",
    "        raise TypeEror(\"sub_num must be a string\")\n",
    "    if not os.path.exists(sub_path):\n",
    "        raise FileNotFoundError(f\"{sub_num} submission path is invalid\")\n",
    "    if not os.path.exists(master_manifest_path):\n",
    "        raise FileNotFoundError(\"master manifest path is invalid\")\n",
    "    if not type(classifier_map) == dict:\n",
    "        raise TypeError(\"classifier_map must be a dict\")\n",
    "    for classifier in classifier_map.values():\n",
    "        if not os.path.exists(classifier):\n",
    "            raise FileNotFoundError(f\"classifier map contains invalid paths. {classifier} is not a valid path\")\n",
    "    \n",
    "    #if output folders do not exist, create them\n",
    "    os.makedirs(profile_o_path, exist_ok=True)\n",
    "    os.makedirs(summary_o_folder, exist_ok=True)\n",
    "    os.makedirs(summary_o_path, exist_ok=True)\n",
    "    \n",
    "    #generate manifest\n",
    "    manifest = pd.read_csv(master_manifest_path, sep='\\t', encoding='latin1')\n",
    "    \n",
    "    #Rename \"Run\" to 'id' to support compatability with qiime2\n",
    "    if 'Run' in manifest.columns:\n",
    "        manifest.rename(columns={'Run': 'id'}, inplace=True)\n",
    "    \n",
    "    #prune manifest\n",
    "    manifest = manifest[manifest['Submission'] == sub_num]\n",
    "    reads = os.listdir(sub_path)\n",
    "    len_manifest = len(manifest)\n",
    "    len_reads = len(reads)\n",
    "    manifest = manifest[manifest['id'].isin(reads)]\n",
    "    len_filtered_manifest = len(manifest)\n",
    "    \n",
    "    #check that names and number of samples match on manifest and in folder\n",
    "    if manifest.empty:\n",
    "        raise ValueError(\"Pruned manifest is empty. Submission number may not be in manifest or reads may be improperly named.\")\n",
    "    if len_manifest > len_reads:\n",
    "        print(\"Warning: more entries in manifest than in folder\")\n",
    "    if len_manifest < len_reads:\n",
    "        print(\"Warning: more entries in folder than in manifest\")\n",
    "    if len_manifest == len_reads and len_filtered_manifest < len_reads:\n",
    "        print(\"Warning: mismatch between some names on manifest and in folder\")\n",
    "        \n",
    "    #check that the submission does not have mixed single and paired reads\n",
    "    unique_layouts = manifest['Library_Layout'].unique().tolist()\n",
    "    \n",
    "    if len(unique_layouts) > 1:\n",
    "        raise ValueError(\"Multiple layouts detected. Submission may contain single and paired reads.\")\n",
    "    p_s = unique_layouts[0]\n",
    "    \n",
    "    #check that the submission does not have samples with different 16s regions\n",
    "    unique_16s = manifest['Region_16S'].unique().tolist()\n",
    "    if len(unique_layouts) > 1:\n",
    "        raise ValueError(\"Multiple regions detected. Submission may contain reads from different 16s regions.\")\n",
    "    region_16s = unique_16s[0]\n",
    "    \n",
    "    #add file paths to manifest\n",
    "    if p_s == 'PAIRED':\n",
    "        manifest['forward-absolute-filepath'] = 0\n",
    "        manifest['reverse-absolute-filepath'] = 0\n",
    "        for sample in manifest['id'].tolist():\n",
    "            sample_path = os.path.join(sub_path, sample)\n",
    "            reads = os.listdir(sample_path)\n",
    "            #check for correct number of reads\n",
    "            if len(reads) != 2:\n",
    "                raise ValueError(f\"Incorrect number of files for {sample}. Should be 2. Got {len(reads)}.\")\n",
    "            #add read paths to manifest\n",
    "            manifest.loc[manifest['id'] == sample, 'forward-absolute-filepath'] = os.path.join(sample_path, reads[0])\n",
    "            manifest.loc[manifest['id'] == sample, 'reverse-absolute-filepath'] = os.path.join(sample_path, reads[1])\n",
    "    elif p_s == 'SINGLE':\n",
    "        manifest['absolute-filepath'] = 0\n",
    "        for sample in manifest['id'].tolist():\n",
    "            sample_path = os.path.join(sub_path, sample)\n",
    "            reads = os.listdir(sample_path)\n",
    "            #check for correct number of reads\n",
    "            if len(reads) != 1:\n",
    "                raise ValueError(f\"incorrect number of files for {sample}. Should be 1. Got {len(reads)}.\")\n",
    "            #add read path to manifest\n",
    "            manifest.loc[manifest['id'] == sample, 'absolute-filepath'] = os.path.join(sample_path, reads[0])\n",
    "    else:\n",
    "        raise ValueError(\"invalid value for library layout\")\n",
    "        \n",
    "    #save the manifest\n",
    "    manifest.to_csv(manifest_path, sep='\\t', index=False)\n",
    "\n",
    "    ## execute qiime2 pipeline ##\n",
    "    \n",
    "    print(f\"Processing {sub_num}\")\n",
    "       \n",
    "    # Execute import command\n",
    "    if p_s == 'PAIRED':\n",
    "        demux = qiime2.Artifact.import_data(\n",
    "            'SampleData[PairedEndSequencesWithQuality]', \n",
    "            manifest_path, \n",
    "            view_type='PairedEndFastqManifestPhred33V2'\n",
    "        )\n",
    "    elif p_s == 'SINGLE':\n",
    "        demux = qiime2.Artifact.import_data(\n",
    "            'SampleData[SequencesWithQuality]', \n",
    "            manifest_path, \n",
    "            view_type='SingleEndFastqManifestPhred33V2'\n",
    "        )\n",
    "\n",
    "    # Execute denoise command\n",
    "    denoising_stats_output = os.path.join(summary_o_path, f\"{sub_num}-denoise-stats.qza\") \n",
    "\n",
    "    if p_s == 'PAIRED':\n",
    "        denoise_results = denoise_paired(\n",
    "            demultiplexed_seqs=demux,\n",
    "            trim_left_f=trim_left_f,\n",
    "            trim_left_r=trim_left_r,\n",
    "            trunc_len_f=trunc_len_f,\n",
    "            trunc_len_r=trunc_len_r, \n",
    "            n_threads=n_threads\n",
    "        )\n",
    "    elif p_s == 'SINGLE':\n",
    "        denoise_results = denoise_single(\n",
    "            demultiplexed_seqs=demux,\n",
    "            trim_left=trim_left_f,\n",
    "            trunc_len=trunc_len_f, \n",
    "            n_threads=n_threads\n",
    "        )\n",
    "\n",
    "    denoise_results.denoising_stats.save(denoising_stats_output)\n",
    "    \n",
    "    #temporay? Keep for longer?\n",
    "    denoise_results.representative_sequences.save(os.path.join(summary_o_path, f\"{sub_num}-rep_seqs\"))\n",
    "    denoise_results.table.save(os.path.join(summary_o_path, f\"{sub_num}-freq_table\"))\n",
    "    denoise_results.denoising_stats.save(denoising_stats_output)\n",
    "    print(f\"{sub_num} denoising complete\")\n",
    "\n",
    "    # Execute summary command\n",
    "    visualization_output = os.path.join(summary_o_path, f\"{sub_num}-summary\")\n",
    "    metadata = qiime2.Metadata.load(manifest_path)\n",
    "\n",
    "    summary_result = summarize(\n",
    "        table=denoise_results.table,\n",
    "        sample_metadata=metadata\n",
    "    )\n",
    "    summary_result.visualization.save(visualization_output)\n",
    "    print(f\"{sub_num} summary table complete\")\n",
    "\n",
    "    # Execute classify command\n",
    "    classifier = qiime2.Artifact.load(classifier_map[region_16s])\n",
    "    taxonomy_output = os.path.join(summary_o_path, f\"{sub_num}-taxonomy\")\n",
    "\n",
    "    taxonomy = classify_sklearn(\n",
    "        reads=denoise_results.representative_sequences,\n",
    "        classifier=classifier, \n",
    "        n_jobs=-1,\n",
    "    )\n",
    "    taxonomy.classification.save(taxonomy_output)\n",
    "    \n",
    "    print(f\"{sub_num} taxonomy classification complete\")\n",
    "    \n",
    "    \"\"\"from qiime2 import Artifact\n",
    "    current_dir = os.getcwd()\n",
    "    profile_o_path = os.path.join(current_dir, \"profile_o\")\n",
    "    taxonomy_path = os.path.join(profile_o_path, f\"{sub_num}-taxonomy.qza\")\n",
    "    temp_path = os.path.join(current_dir, \"temp\")\n",
    "    denoise_path = os.path.join(temp_path, \"denoise.qza\")\n",
    "    manifest=os.path.join(temp_path, \"manifest.tsv\")\n",
    "    taxonomy = Artifact.load(taxonomy_path)\n",
    "    denoise_results = Artifact.load(denoise_path)\"\"\"\n",
    "    \n",
    "    #get assignments from BLAST\n",
    "    \n",
    "    \n",
    "    #Generate Relitive Frequencies\n",
    "    rel_freq = relative_frequency(denoise_results.table)[0] #add .table for denoise results\n",
    "    print(f\"{sub_num} relitive frequency table complete\")\n",
    "    \n",
    "    #Generate Taxonomy Bar Graph\n",
    "    barplot_path = os.path.join(summary_o_path, f\"{sub_num}-barplot\") \n",
    "    metadata = Metadata.load(manifest_path)\n",
    "    \n",
    "    barplot_vis = barplot(\n",
    "        table=denoise_results.table, \n",
    "        taxonomy=taxonomy.classification, \n",
    "        metadata=metadata\n",
    "    )\n",
    "    barplot_vis.visualization.save(barplot_path)\n",
    "    print(f\"{sub_num} barplot complete\")\n",
    "    \n",
    "    #generate final output data\n",
    "    rel_freq_df = rel_freq.view(pd.DataFrame)\n",
    "    rel_freq_df = rel_freq_df.transpose()\n",
    "    taxonomy_df = taxonomy.classification.view(pd.DataFrame)\n",
    "    \n",
    "    #display_table(rel_freq_df)\n",
    "    #display_table(taxonomy_df)\n",
    "    \n",
    "    merged_df = pd.merge(rel_freq_df, taxonomy_df, left_index=True, right_index=True, how='inner')\n",
    "    merged_df = merged_df.drop(columns=['Confidence']) #used to also drop feature ID, will see if it works\n",
    "    #sum accross multiple OTUs that are the same species\n",
    "    merged_df = merged_df.groupby('Taxon').sum()\n",
    "    \n",
    "    #DEBUG: \n",
    "    #display_table(merged_df)\n",
    "    \n",
    "    \n",
    "    #print(\"\\n========== fel_freq_df ==========\\n\")\n",
    "    #print(set(rel_freq_df.index))\n",
    "    #print(\"\\n========== taxon_df ==========\\n\")\n",
    "    #print(set(taxonomy_df['Feature ID']))\n",
    "    #print(\"\\n========== join ==========\\n\")\n",
    "    #print(set(rel_freq_df.index) & set(taxonomy_df['Feature ID']))\n",
    "    \n",
    "    #save final output data\n",
    "    profile_output = os.path.join(profile_o_path, f\"{sub_num}-profile.csv\")\n",
    "    merged_df.to_csv(profile_output, index=True)\n",
    "    print(f\"==================== {sub_num} processing complete ====================\")\n",
    "\n",
    "\n",
    "#TODO: double check sequence assignments with blast\n",
    "def display_table(table):\n",
    "    display(table.head())\n",
    "    num_rows, num_cols = table.shape\n",
    "    print(\"Number of rows:\", num_rows)\n",
    "    print(\"Number of columns:\", num_cols)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing submissions {'ERA1111396', 'ERA292152', 'SRA578693', 'SRA1038019', 'SRA298410', 'SRA860003', 'SRA575681', 'SRA996410'}\n",
      "\n",
      "Warning: more entries in manifest than in folder\n",
      "Processing ERA1111396\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-fz0pn7ls/2d82f39b-0744-4caa-8965-e42a868b27f3/data /tmp/tmp6_w64ewo/output.tsv.biom /tmp/tmp6_w64ewo/track.tsv /tmp/tmp6_w64ewo 0 0 2.0 2 Inf independent consensus 1.0 8 1000000 NULL 16\n",
      "\n",
      "ERA1111396 denoising complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:122: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'sample-frequency-detail.csv'))\n",
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:124: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'feature-frequency-detail.csv'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERA1111396 summary table complete\n",
      "ERA1111396 taxonomy classification complete\n",
      "ERA1111396 relitive frequency table complete\n",
      "ERA1111396 barplot complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ERR2162452</th>\n",
       "      <th>ERR2162453</th>\n",
       "      <th>ERR2162454</th>\n",
       "      <th>ERR2162455</th>\n",
       "      <th>ERR2162456</th>\n",
       "      <th>ERR2162457</th>\n",
       "      <th>ERR2162458</th>\n",
       "      <th>ERR2162459</th>\n",
       "      <th>ERR2162460</th>\n",
       "      <th>ERR2162461</th>\n",
       "      <th>...</th>\n",
       "      <th>ERR2163209</th>\n",
       "      <th>ERR2163210</th>\n",
       "      <th>ERR2163211</th>\n",
       "      <th>ERR2163212</th>\n",
       "      <th>ERR2163213</th>\n",
       "      <th>ERR2163214</th>\n",
       "      <th>ERR2163215</th>\n",
       "      <th>ERR2163216</th>\n",
       "      <th>ERR2163217</th>\n",
       "      <th>ERR2163218</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Unassigned</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.052964</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.029153</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Archaea</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Archaea; p__Crenarchaeota; c__Thaumarchaeota; o__Cenarchaeales; f__Cenarchaeaceae; g__Nitrosopumilus; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Archaea; p__Crenarchaeota; c__Thaumarchaeota; o__Nitrososphaerales; f__Nitrososphaeraceae</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Archaea; p__Crenarchaeota; c__Thaumarchaeota; o__Nitrososphaerales; f__Nitrososphaeraceae; g__Candidatus Nitrososphaera; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 761 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    ERR2162452  ERR2162453  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                                 0.0    0.000261   \n",
       "k__Archaea                                                 0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "\n",
       "                                                    ERR2162454  ERR2162455  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                            0.052964         0.0   \n",
       "k__Archaea                                            0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "\n",
       "                                                    ERR2162456  ERR2162457  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                                 0.0         0.0   \n",
       "k__Archaea                                                 0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "\n",
       "                                                    ERR2162458  ERR2162459  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                                 0.0    0.001969   \n",
       "k__Archaea                                                 0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "\n",
       "                                                    ERR2162460  ERR2162461  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                            0.000192         0.0   \n",
       "k__Archaea                                            0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "\n",
       "                                                    ...  ERR2163209  \\\n",
       "Taxon                                               ...               \n",
       "Unassigned                                          ...         0.0   \n",
       "k__Archaea                                          ...         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...  ...         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...  ...         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...  ...         0.0   \n",
       "\n",
       "                                                    ERR2163210  ERR2163211  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                                 0.0    0.029153   \n",
       "k__Archaea                                                 0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "\n",
       "                                                    ERR2163212  ERR2163213  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                            0.000133    0.000081   \n",
       "k__Archaea                                            0.000000    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000    0.000000   \n",
       "\n",
       "                                                    ERR2163214  ERR2163215  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                            0.000034         0.0   \n",
       "k__Archaea                                            0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "\n",
       "                                                    ERR2163216  ERR2163217  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                                 0.0         0.0   \n",
       "k__Archaea                                                 0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "\n",
       "                                                    ERR2163218  \n",
       "Taxon                                                           \n",
       "Unassigned                                                 0.0  \n",
       "k__Archaea                                                 0.0  \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0  \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0  \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0  \n",
       "\n",
       "[5 rows x 761 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 2181\n",
      "Number of columns: 761\n",
      "==================== ERA1111396 processing complete ====================\n",
      "Warning: more entries in manifest than in folder\n",
      "Processing ERA292152\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-3ffhmlxu/b1dea680-b371-4f96-9a85-3c2f07afbf8f/data /tmp/tmpcgfdg1sd/output.tsv.biom /tmp/tmpcgfdg1sd/track.tsv /tmp/tmpcgfdg1sd 0 0 2.0 2 Inf independent consensus 1.0 8 1000000 NULL 16\n",
      "\n",
      "Error processing ERA292152: An error was encountered while running DADA2 in R (return code 1), please inspect stdout and stderr to learn more.\n",
      "==================== ERA292152 processing failed ====================\n",
      "Error processing SRA578693: Pruned manifest is empty. Submission number may not be in manifest or reads may be improperly named.\n",
      "==================== SRA578693 processing failed ====================\n",
      "Processing SRA1038019\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_paired.R /tmp/tmpjbvblg_6/forward /tmp/tmpjbvblg_6/reverse /tmp/tmpjbvblg_6/output.tsv.biom /tmp/tmpjbvblg_6/track.tsv /tmp/tmpjbvblg_6/filt_f /tmp/tmpjbvblg_6/filt_r 0 0 0 0 2.0 2.0 2 independent consensus 1.0 8 1000000\n",
      "\n",
      "SRA1038019 denoising complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:122: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'sample-frequency-detail.csv'))\n",
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:124: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'feature-frequency-detail.csv'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRA1038019 summary table complete\n",
      "SRA1038019 taxonomy classification complete\n",
      "SRA1038019 relitive frequency table complete\n",
      "SRA1038019 barplot complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRR11027332</th>\n",
       "      <th>SRR11027333</th>\n",
       "      <th>SRR11027334</th>\n",
       "      <th>SRR11027335</th>\n",
       "      <th>SRR11027336</th>\n",
       "      <th>SRR11027337</th>\n",
       "      <th>SRR11027338</th>\n",
       "      <th>SRR11027339</th>\n",
       "      <th>SRR11027340</th>\n",
       "      <th>SRR11027341</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR11027344</th>\n",
       "      <th>SRR11027345</th>\n",
       "      <th>SRR11027346</th>\n",
       "      <th>SRR11027347</th>\n",
       "      <th>SRR11027348</th>\n",
       "      <th>SRR11027349</th>\n",
       "      <th>SRR11027350</th>\n",
       "      <th>SRR11027351</th>\n",
       "      <th>SRR11027352</th>\n",
       "      <th>SRR11027353</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>k__Bacteria</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.901408</td>\n",
       "      <td>0.34127</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.111842</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales; f__Porphyromonadaceae; g__Porphyromonas; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.412214</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Firmicutes; c__Bacilli; o__Bacillales; f__Staphylococcaceae; g__Staphylococcus</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Clostridiaceae; g__Clostridium; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156489</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.425</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__[Tissierellaceae]; g__Anaerococcus; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.526316</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.148855</td>\n",
       "      <td>...</td>\n",
       "      <td>0.888158</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    SRR11027332  SRR11027333  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "\n",
       "                                                    SRR11027334  SRR11027335  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                            0.000000          0.0   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...     0.333333          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...     0.000000          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...     0.000000          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...     0.500000          0.0   \n",
       "\n",
       "                                                    SRR11027336  SRR11027337  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                            0.901408      0.34127   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...     0.000000      0.00000   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...     0.000000      0.00000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...     0.000000      0.00000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...     0.000000      0.00000   \n",
       "\n",
       "                                                    SRR11027338  SRR11027339  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0     0.000000   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.526316   \n",
       "\n",
       "                                                    SRR11027340  SRR11027341  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0     0.000000   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0     0.412214   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.156489   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.148855   \n",
       "\n",
       "                                                    ...  SRR11027344  \\\n",
       "Taxon                                               ...                \n",
       "k__Bacteria                                         ...     0.111842   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...  ...     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...  ...     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...  ...     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...  ...     0.888158   \n",
       "\n",
       "                                                    SRR11027345  SRR11027346  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0        0.000   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0        0.000   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0        0.000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0        0.425   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0        0.000   \n",
       "\n",
       "                                                    SRR11027347  SRR11027348  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0     0.083333   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0     0.916667   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.000000   \n",
       "\n",
       "                                                    SRR11027349  SRR11027350  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          1.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "\n",
       "                                                    SRR11027351  SRR11027352  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "\n",
       "                                                    SRR11027353  \n",
       "Taxon                                                            \n",
       "k__Bacteria                                                 0.0  \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0  \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0  \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0  \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 15\n",
      "Number of columns: 22\n",
      "==================== SRA1038019 processing complete ====================\n",
      "Warning: more entries in manifest than in folder\n",
      "Processing SRA298410\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_paired.R /tmp/tmpscm0387u/forward /tmp/tmpscm0387u/reverse /tmp/tmpscm0387u/output.tsv.biom /tmp/tmpscm0387u/track.tsv /tmp/tmpscm0387u/filt_f /tmp/tmpscm0387u/filt_r 0 0 0 0 2.0 2.0 2 independent consensus 1.0 8 1000000\n",
      "\n",
      "Error processing SRA298410: An error was encountered while running DADA2 in R (return code -11), please inspect stdout and stderr to learn more.\n",
      "==================== SRA298410 processing failed ====================\n",
      "Processing SRA860003\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_paired.R /tmp/tmprx5v08qh/forward /tmp/tmprx5v08qh/reverse /tmp/tmprx5v08qh/output.tsv.biom /tmp/tmprx5v08qh/track.tsv /tmp/tmprx5v08qh/filt_f /tmp/tmprx5v08qh/filt_r 0 0 0 0 2.0 2.0 2 independent consensus 1.0 8 1000000\n",
      "\n",
      "SRA860003 denoising complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:122: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'sample-frequency-detail.csv'))\n",
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:124: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'feature-frequency-detail.csv'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRA860003 summary table complete\n",
      "SRA860003 taxonomy classification complete\n",
      "SRA860003 relitive frequency table complete\n",
      "SRA860003 barplot complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRR8728265</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Corynebacteriaceae; g__Corynebacterium; s__</th>\n",
       "      <td>0.082248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Dermabacteraceae; g__Brachybacterium</th>\n",
       "      <td>0.020562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Dietziaceae; g__Dietzia; s__</th>\n",
       "      <td>0.002443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Micrococcaceae</th>\n",
       "      <td>0.002850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Micrococcaceae; g__Kocuria; s__palustris</th>\n",
       "      <td>0.036849</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    SRR8728265\n",
       "Taxon                                                         \n",
       "k__Bacteria; p__Actinobacteria; c__Actinobacter...    0.082248\n",
       "k__Bacteria; p__Actinobacteria; c__Actinobacter...    0.020562\n",
       "k__Bacteria; p__Actinobacteria; c__Actinobacter...    0.002443\n",
       "k__Bacteria; p__Actinobacteria; c__Actinobacter...    0.002850\n",
       "k__Bacteria; p__Actinobacteria; c__Actinobacter...    0.036849"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 49\n",
      "Number of columns: 1\n",
      "==================== SRA860003 processing complete ====================\n",
      "Error processing SRA575681: Pruned manifest is empty. Submission number may not be in manifest or reads may be improperly named.\n",
      "==================== SRA575681 processing failed ====================\n",
      "Processing SRA996410\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-uoy8c3cn/d6fc8df8-2285-46bb-a7e6-d4210da98c38/data /tmp/tmp9eye8jzr/output.tsv.biom /tmp/tmp9eye8jzr/track.tsv /tmp/tmp9eye8jzr 0 0 2.0 2 Inf independent consensus 1.0 8 1000000 NULL 16\n",
      "\n",
      "SRA996410 denoising complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:122: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'sample-frequency-detail.csv'))\n",
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:124: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'feature-frequency-detail.csv'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRA996410 summary table complete\n",
      "SRA996410 taxonomy classification complete\n",
      "SRA996410 relitive frequency table complete\n",
      "SRA996410 barplot complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRR10447626</th>\n",
       "      <th>SRR10447627</th>\n",
       "      <th>SRR10447628</th>\n",
       "      <th>SRR10447629</th>\n",
       "      <th>SRR10447630</th>\n",
       "      <th>SRR10447631</th>\n",
       "      <th>SRR10447632</th>\n",
       "      <th>SRR10447633</th>\n",
       "      <th>SRR10447634</th>\n",
       "      <th>SRR10447635</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR10447734</th>\n",
       "      <th>SRR10447735</th>\n",
       "      <th>SRR10447736</th>\n",
       "      <th>SRR10447737</th>\n",
       "      <th>SRR10447738</th>\n",
       "      <th>SRR10447739</th>\n",
       "      <th>SRR10447740</th>\n",
       "      <th>SRR10447741</th>\n",
       "      <th>SRR10447742</th>\n",
       "      <th>SRR10447743</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>k__Archaea; p__Crenarchaeota; c__Thaumarchaeota; o__Nitrososphaerales; f__Nitrososphaeraceae; g__Candidatus Nitrososphaera; s__SCA1170</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006458</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003238</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__; c__; o__; f__; g__; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Acidobacteria</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Acidobacteria; c__Acidobacteria-5; o__; f__; g__; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 118 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    SRR10447626  SRR10447627  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447628  SRR10447629  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447630  SRR10447631  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447632  SRR10447633  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447634  SRR10447635  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    ...  SRR10447734  \\\n",
       "Taxon                                               ...                \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...  ...          0.0   \n",
       "k__Bacteria                                         ...          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__           ...          0.0   \n",
       "k__Bacteria; p__Acidobacteria                       ...          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...  ...          0.0   \n",
       "\n",
       "                                                    SRR10447735  SRR10447736  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0     0.000000   \n",
       "k__Bacteria                                                 0.0     0.006458   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0     0.000000   \n",
       "k__Bacteria; p__Acidobacteria                               0.0     0.000000   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0     0.000000   \n",
       "\n",
       "                                                    SRR10447737  SRR10447738  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0     0.000000   \n",
       "k__Bacteria                                                 0.0     0.003238   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0     0.000000   \n",
       "k__Bacteria; p__Acidobacteria                               0.0     0.000000   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0     0.000000   \n",
       "\n",
       "                                                    SRR10447739  SRR10447740  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447741  SRR10447742  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447743  \n",
       "Taxon                                                            \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0  \n",
       "k__Bacteria                                                 0.0  \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0  \n",
       "k__Bacteria; p__Acidobacteria                               0.0  \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0  \n",
       "\n",
       "[5 rows x 118 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 994\n",
      "Number of columns: 118\n",
      "==================== SRA996410 processing complete ====================\n"
     ]
    }
   ],
   "source": [
    "#Run Pipeline\n",
    "#generate file paths\n",
    "current_dir = os.getcwd()\n",
    "classifier_folder = os.path.join(current_dir, \"classifier\")\n",
    "master_manifest_path = os.path.join(current_dir, \"DataFrame_2_Pruned.tsv\")\n",
    "profile_o_path = os.path.join(current_dir, \"profile_o\")\n",
    "summary_o_path = os.path.join(current_dir, \"summary_o\")\n",
    "gg_nb = os.path.join(classifier_folder, \"gg-13-8-99-nb-classifier.qza\") \n",
    "\n",
    "#only run the pipeline on unprocessed reads\n",
    "sub_nums = set(os.listdir(os.path.join(current_dir, \"sorted_reads\")))\n",
    "completed = set(os.listdir(profile_o_path))\n",
    "completed = [x[:-len(\"-profile.csv\")] for x in completed]\n",
    "sub_nums.difference_update(completed)\n",
    "\n",
    "\n",
    "classifier_map = {\n",
    "    \"V4\":gg_nb, \n",
    "    \"V1-V3\":gg_nb, \n",
    "    \"V1-V2\":gg_nb, \n",
    "    \"V3-V4\":gg_nb, \n",
    "    \"V1-V2 and v4\":gg_nb, \n",
    "    \"V4-V5\":gg_nb, \n",
    "    \"V1-V4\":gg_nb\n",
    "}\n",
    "\n",
    "#BUG: wrong set being used to join taxon df and freq df\n",
    "\n",
    "print(f\"Processing submissions {sub_nums}\\n\")\n",
    "for sub_num in sub_nums:\n",
    "    sub_path = os.path.join(current_dir, \"sorted_reads\", sub_num)\n",
    "    try:\n",
    "        qiime2_profile(sub_num, sub_path, master_manifest_path, profile_o_path, summary_o_path, classifier_map, trim_left_f=0, trim_left_r=0, trunc_len_f=0, trunc_len_r=0, )\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing {sub_num}: {e}\")\n",
    "        print(f\"==================== {sub_num} processing failed ====================\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test Pipeline\n",
    "\n",
    "#generate file paths\n",
    "current_dir = os.getcwd()\n",
    "sub_num = \"ERA1522711\"\n",
    "classifier_folder = os.path.join(current_dir, \"classifier\")\n",
    "sub_path = os.path.join(current_dir, f\"sorted_reads/{sub_num}\")\n",
    "master_manifest_path = os.path.join(current_dir, \"DataFrame_2_Pruned.tsv\")\n",
    "profile_o_path = os.path.join(current_dir, \"profile_o\")\n",
    "summary_o_path = os.path.join(current_dir, \"summary_o\")\n",
    "\n",
    "#classifiers\n",
    "silva_nb = os.path.join(classifier_folder, \"silva-138-99-nb-classifier.qza\") #warning: species are unreliable\n",
    "gg_nb = os.path.join(classifier_folder, \"gg-13-8-99-nb-classifier.qza\") \n",
    "\n",
    "classifier_map = {\n",
    "    \"V4\":gg_nb, \n",
    "    \"V1-V3\":gg_nb, \n",
    "    \"V1-V2\":gg_nb, \n",
    "    \"V3-V4\":gg_nb, \n",
    "    \"V1-V2 and v4\":gg_nb, \n",
    "    \"V4-V5\":gg_nb, \n",
    "    \"V1-V4\":gg_nb\n",
    "}\n",
    "\n",
    "qiime2_profile(sub_num, sub_path, master_manifest_path, profile_o_path, summary_o_path, classifier_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#no trim test\n",
    "#generate bar plot graph\n",
    "from qiime2 import Visualization\n",
    "from qiime2 import Artifact\n",
    "from qiime2.plugins import demux, dada2, feature_classifier, feature_table\n",
    "from qiime2 import Metadata\n",
    "import os\n",
    "from qiime2.plugins.taxa.visualizers import barplot\n",
    "\n",
    "current_dir = os.getcwd()\n",
    "sub_num = \"ERA1777984\"\n",
    "classifier_folder = os.path.join(current_dir, \"classifier\")\n",
    "sub_path = os.path.join(current_dir, f\"sorted_reads/{sub_num}\")\n",
    "master_manifest_path = os.path.join(current_dir, \"DataFrame_2_Pruned.tsv\")\n",
    "profile_o_path = os.path.join(current_dir, \"profile_o\")\n",
    "summary_o_path = os.path.join(current_dir, \"summary_o\")\n",
    "\n",
    "#classifiers\n",
    "silva_nb = os.path.join(classifier_folder, \"silva-138-99-nb-classifier.qza\") #warning: species are unreliable\n",
    "gg_nb = os.path.join(classifier_folder, \"gg-13-8-99-nb-classifier.qza\") \n",
    "\n",
    "classifier_map = {\n",
    "    \"V4\":gg_nb, \n",
    "    \"V1-V3\":gg_nb, \n",
    "    \"V1-V2\":gg_nb, \n",
    "    \"V3-V4\":gg_nb, \n",
    "    \"V1-V2 and v4\":gg_nb, \n",
    "    \"V4-V5\":gg_nb, \n",
    "    \"V1-V4\":gg_nb\n",
    "}\n",
    "\n",
    "qiime2_profile(sub_num, sub_path, master_manifest_path, profile_o_path, summary_o_path, classifier_map, trim_left_f=0, trim_left_r=0, trunc_len_f=0, trunc_len_r=0)\n",
    "\n",
    "\n",
    "barplot_path = os.path.join(summary_o_path, sub_num, f\"{sub_num}-barplot.qza\")\n",
    "barplot_vis = qiime2.Visualization.load('path_to_barplot.qzv')\n",
    "barplot_vis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-rahtj2le')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 8a0c45c1-9865-4725-8d55-ada57cdaed15>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#generate bar plot graph\n",
    "from qiime2 import Visualization\n",
    "from qiime2 import Artifact\n",
    "from qiime2.plugins import demux, dada2, feature_classifier, feature_table\n",
    "from qiime2 import Metadata\n",
    "import os\n",
    "from qiime2.plugins.taxa.visualizers import barplot\n",
    "\n",
    "\n",
    "sub_num = \"ERA1522711\"\n",
    "current_dir = os.getcwd()\n",
    "profile_o_path = os.path.join(current_dir, \"profile_o\")\n",
    "taxonomy_path = os.path.join(profile_o_path, f\"{sub_num}-taxonomy.qza\")\n",
    "temp_path = os.path.join(current_dir, \"temp\")\n",
    "denoise_path = os.path.join(temp_path, \"denoise.qza\")\n",
    "manifest=os.path.join(temp_path, \"manifest.tsv\")\n",
    "\n",
    "\n",
    "metadata = Metadata.load(manifest)\n",
    "\n",
    "taxonomy_qza = Artifact.load(taxonomy_path)\n",
    "\n",
    "denoise_qza = Artifact.load(denoise_path)\n",
    "\n",
    "# Calculate the relative frequency of features\n",
    "relative_frequency_result = feature_table.methods.relative_frequency(denoise_qza)\n",
    "\n",
    "\n",
    "# Generate the barplot visualization\n",
    "barplot_viz = barplot(table=denoise_qza, taxonomy=taxonomy_qza, metadata=metadata)\n",
    "\n",
    "# Visualize the barplot\n",
    "barplot_viz.visualization\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "/home/miloscola/augene/aggit2/SKINOME_training_data/profile_o/ERA1522711-taxonomy.qza does not exist.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-394ec0119db2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[0mdenoise_qza\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mArtifact\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdenoise_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m \u001b[0mseq_qza\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mArtifact\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 32\u001b[0;31m \u001b[0mtaxonomy_qza\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mArtifact\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtaxonomy_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     33\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[0;31m#get realitive frequencies\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/qiime2/lib/python3.6/site-packages/qiime2/sdk/result.py\u001b[0m in \u001b[0;36mload\u001b[0;34m(cls, filepath)\u001b[0m\n\u001b[1;32m     64\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfilepath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     65\u001b[0m         \u001b[0;34m\"\"\"Factory for loading Artifacts and Visualizations.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 66\u001b[0;31m         \u001b[0marchiver\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0marchive\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mArchiver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     67\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     68\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mArtifact\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_is_valid_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marchiver\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/qiime2/lib/python3.6/site-packages/qiime2/core/archive/archiver.py\u001b[0m in \u001b[0;36mload\u001b[0;34m(cls, filepath)\u001b[0m\n\u001b[1;32m    297\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mclassmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    298\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfilepath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 299\u001b[0;31m         \u001b[0marchive\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_archive\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    300\u001b[0m         \u001b[0mFormat\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_format_class\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marchive\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mversion\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    301\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mFormat\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/qiime2/lib/python3.6/site-packages/qiime2/core/archive/archiver.py\u001b[0m in \u001b[0;36mget_archive\u001b[0;34m(cls, filepath)\u001b[0m\n\u001b[1;32m    257\u001b[0m         \u001b[0mfilepath\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpathlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPath\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    258\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mfilepath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 259\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"%s does not exist.\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mfilepath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    260\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    261\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0m_ZipArchive\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_archive_type\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: /home/miloscola/augene/aggit2/SKINOME_training_data/profile_o/ERA1522711-taxonomy.qza does not exist."
     ]
    }
   ],
   "source": [
    "#test conversion to machine learning data\n",
    "from qiime2 import Visualization\n",
    "from qiime2 import Artifact\n",
    "from qiime2.plugins import demux, dada2, feature_classifier, feature_table\n",
    "from qiime2 import Metadata\n",
    "import os\n",
    "import pandas as pd\n",
    "from qiime2.plugins.taxa.visualizers import barplot\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio import SeqIO\n",
    "\n",
    "sub_num = \"ERA1522711\"\n",
    "current_dir = os.getcwd()\n",
    "profile_o_path = os.path.join(current_dir, \"profile_o\")\n",
    "summary_o_path = os.path.join(current_dir, \"summary_o\", sub_num)\n",
    "taxonomy_path = os.path.join(profile_o_path, f\"{sub_num}-taxonomy.qza\")\n",
    "temp_path = os.path.join(current_dir, \"temp\")\n",
    "#denoise_path = os.path.join(temp_path, \"denoise.qza\")\n",
    "manifest=os.path.join(temp_path, \"manifest.tsv\")\n",
    "seq_path = os.path.join(summary_o_path, f\"{sub_num}-rep_seqs.qza\")\n",
    "denoise_path = os.path.join(summary_o_path, f\"{sub_num}-freq_table.qza\")\n",
    "profile_o_path = os.path.join(current_dir, \"profile_o\")\n",
    "    \n",
    "\n",
    "#assign new file paths\n",
    "#denoise_biom_path = os.path.join(temp_path, f\"{sub_num}-denoise.biom\")\n",
    "#taxonomy_tsv_path = os.path.join(temp_path, f\"{sub_num}-taxonomy.tsv\")\n",
    "\n",
    "#re-format old files\n",
    "denoise_qza = Artifact.load(denoise_path)\n",
    "seq_qza = Artifact.load(seq_path)\n",
    "taxonomy_qza = Artifact.load(taxonomy_path)\n",
    "\n",
    "#get realitive frequencies\n",
    "rel_freq_qza = feature_table.methods.relative_frequency(denoise_qza)[0]\n",
    "#print(rel_freq_qza[0])\n",
    "#rel_freq_qza.save(os.path.join(temp_path, \"relative_frequency.qza\"))\n",
    "#denoise_qza.save(denoise_biom_path)\n",
    "#taxonomy_qza.save(taxonomy_tsv_path)\n",
    "\n",
    "rel_freq_df = rel_freq_qza.view(pd.DataFrame)\n",
    "rel_freq_df = rel_freq_df.transpose()\n",
    "taxonomy_df = taxonomy_qza.view(pd.DataFrame)\n",
    "taxonomy_df.index=taxonomy_df['Feature ID']\n",
    "\n",
    "#denoise_df = pd.read_csv(denoise_biom_path, sep='\\t', index_col=0, skiprows=[1])\n",
    "#taxonomy_df = pd.read_csv(taxonomy_tsv_path, sep='\\t', index_col=0)\n",
    "\n",
    "\n",
    "# Subset the DataFrames to only include common indices\n",
    "#denoise_df = denoise_df.loc[common_indices]\n",
    "#taxonomy_df = taxonomy_df.loc[common_indices]\n",
    "\n",
    "merged_df = rel_freq_df.merge(taxonomy_df['Taxon'], left_index=True, right_index=True)\n",
    "merged_df.set_index('Taxon', inplace=True)\n",
    "merged_df = merged_df.groupby('Taxon').sum()\n",
    "\n",
    "display(rel_freq_df.head())\n",
    "\n",
    "num_rows, num_cols = rel_freq_df.shape\n",
    "\n",
    "print(\"Number of rows:\", num_rows)\n",
    "print(\"Number of columns:\", num_cols)\n",
    "\n",
    "display(taxonomy_df.head())\n",
    "\n",
    "num_rows, num_cols = taxonomy_df.shape\n",
    "\n",
    "print(\"Number of rows:\", num_rows)\n",
    "print(\"Number of columns:\", num_cols)\n",
    "\n",
    "display(merged_df.head())\n",
    "\n",
    "num_rows, num_cols = merged_df.shape\n",
    "\n",
    "print(\"Number of rows:\", num_rows)\n",
    "print(\"Number of columns:\", num_cols)\n",
    "\n",
    "merged_df.to_csv(os.path.join(profile_o_path, f\"{sub_num}-profile.csv\"), index=False)\n",
    "\n",
    "\"\"\"# Iterate over samples and write to CSV\n",
    "for sample in feature_table_df.columns:\n",
    "    sample_df = merged_df[[sample, 'Taxon']]\n",
    "    sample_df.columns = ['Relative Frequency', 'Bacteria']\n",
    "    sample_df = sample_df[['Bacteria', 'Relative Frequency']]\n",
    "    sample_df.to_csv(os.path.join(output_dir, f'{sample}.csv'), index=False)\"\"\"\n",
    "\n",
    "def displaylay_file(denoise, temp_path, sub_num):\n",
    "    # Convert the denoised sequences to a pandas DataFrame\n",
    "    sequences_df = denoise.view(pd.DataFrame)\n",
    "\n",
    "    display(sequences_df.head())\n",
    "\n",
    "    num_rows, num_cols = sequences_df.shape\n",
    "\n",
    "    print(\"Number of rows:\", num_rows)\n",
    "    print(\"Number of columns:\", num_cols)\n",
    "    \n",
    "def blast(denoise, temp_path, sub_num, blast_program=\"blastp\", database=\"nr\", num_alignments=10, evalue=0.01):\n",
    "    sequences_df = denoise.view(pd.DataFrame)\n",
    "    \n",
    "    fasta_path = os.path.joinoin(temp_path, f'{sub_num}_output_sequences.fasta')\n",
    "    \n",
    "    with open(fasta_path, 'w') as fasta_file:\n",
    "        for index, row in sequences_df.iterrows():\n",
    "            fasta_file.write(f'>{index}\\n{row[0]}\\n')\n",
    "    \n",
    "    # Read the FASTA file\n",
    "    query_sequence = SeqIO.read(fasta_path, \"fasta\")\n",
    "\n",
    "    # Perform BLAST search\n",
    "    result_handle = NCBIWWW.qblast(blast_program, database, query_sequence.seq, expect=evalue, hitlist_size=num_alignments)\n",
    "    \n",
    "    #convert blast results into a pandas dataframe\n",
    "    rows = []\n",
    "    for record in blast_records:\n",
    "        for alignment in record.alignments:\n",
    "            for hsp in alignment.hsps:\n",
    "                rows.append({\n",
    "                    \"Query ID\": record.query,\n",
    "                    \"Hit ID\": alignment.title,\n",
    "                    \"Length\": alignment.length,\n",
    "                    \"E-value\": hsp.expect,\n",
    "                    \"Score\": hsp.score,\n",
    "                    \"Query Seq\": hsp.query,\n",
    "                    \"Hit Seq\": hsp.sbjct\n",
    "                })\n",
    "    return pd.DataFrame(rows)\n",
    "    \n",
    "displaylay_file(seq_qza)\n",
    "blast_resultslts = blast(seq_qza, temp_path, sub_num)\n",
    "displaylay_file(blast_results)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Visuialize data with pandas (quick and dirty)\n",
    "\n",
    "import qiime2\n",
    "import pandas as pd\n",
    "import os\n",
    "from qiime2.plugins import demux, dada2, feature_classifier, feature_table\n",
    "\n",
    "current_dir = os.getcwd()\n",
    "\n",
    "# Define path to taxonomy.qza file\n",
    "taxonomy_qza_path = os.path.join(current_dir, \"profile_o/ERA1522711-taxonomy.qza\")\n",
    "\n",
    "# Load the taxonomy.qza file\n",
    "taxonomy_artifact = qiime2.Artifact.load(taxonomy_qza_path)\n",
    "\n",
    "# Extract the data from the artifact\n",
    "taxonomy_df = taxonomy_artifact.view(pd.DataFrame)\n",
    "\n",
    "# Display the taxonomy dataframe\n",
    "print(taxonomy_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    #OLD CODE PROB WONT NEED\n",
    "  \n",
    "\n",
    "    #execute import command\n",
    "    if p_s == 'PAIRED':\n",
    "        demux_output = os.path.join(temp_path, \"paired-end-demux.qza\")\n",
    "        import_command = f\"\"\"qiime tools import --type SampleData[PairedEndSequencesWithQuality] \\\n",
    "            --input-path {manifest_path} \\\n",
    "            --output-path {demux_output} \\\n",
    "            --input-format PairedEndFastqManifestPhred33V2\"\"\" \n",
    "    elif p_s == 'SINGLE':\n",
    "        demux_output = os.path.join(temp_path, \"single-end-demux.qza\")\n",
    "        import_command = f\"\"\"qiime tools import --type SampleData[SequencesWithQuality] \\\n",
    "            --input-path {manifest_path} \\\n",
    "            --output-path {demux_output} \\\n",
    "            --input-format SingleEndFastqManifestPhred33V2\"\"\"\n",
    "        \n",
    "    #run_command(import_command)\n",
    "    \n",
    "    #execute denoise command\n",
    "    rep_seqs_dada2_output = os.path.join(temp_path, \"rep-seqs-dada2.qza\")\n",
    "    table_dada2_output = os.path.join(temp_path, \"table-dada2.qza\")\n",
    "    denoising_stats_output = os.path.join(summary_o_path, f\"{sub_num}-stats-dada2.qza\") \n",
    "    \n",
    "    if p_s == 'PAIRED':\n",
    "        denoise_command = f\"\"\"qiime dada2 denoise-paired \\\n",
    "        --i-demultiplexed-seqs {demux_output} \\\n",
    "        --p-trim-left-f {trim_left_f} \\\n",
    "        --p-trim-left-r {trim_left_r} \\\n",
    "        --p-trunc-len-f {trunc_len_f} \\\n",
    "        --p-trunc-len-r {trunc_len_r} \\\n",
    "        --o-representative-sequences {rep_seqs_dada2_output} \\\n",
    "        --o-table {table_dada2_output} \\\n",
    "        --o-denoising-stats {denoising_stats_output}\"\"\"\n",
    "    elif p_s == 'SINGLE':\n",
    "        denoise_command = f\"\"\"qiime dada2 denoise-paired \\\n",
    "        --i-demultiplexed-seqs {demux_output} \\\n",
    "        --p-trim-left {trim_left_f} \\\n",
    "        --p-trunc-len {trunc_len_f} \\\n",
    "        --o-representative-sequences {rep_seqs_dada2_output} \\\n",
    "        --o-table {table_dada2_output} \\\n",
    "        --o-denoising-stats {denoising_stats_output}\"\"\"\n",
    "    \n",
    "    #run_command(denoise_command)\n",
    "    \n",
    "    #execute summary command\n",
    "    visuialization_output = os.path.join(summary_o_path, f\"{sub_num}-table-dada2.qzv\")\n",
    "    \n",
    "    summerize_command = f\"\"\"qiime feature-table summarize \\\n",
    "    --i-table {table_dada2_output} \\\n",
    "    --o-visualization {visuialization_output} \\\n",
    "    --m-sample-metadata-file {manifest_path}\"\"\"\n",
    "    \n",
    "    #run_command(summerize_command)\n",
    "    \n",
    "    #execute classify command\n",
    "    classifier = classifier_map[region_16s]\n",
    "    taxonomy_output = os.path.join(profile_o_path, f\"{sub_num}-taxonomy.qza\")\n",
    "    \n",
    "    classify_command = f\"\"\"qiime feature-classifier classify-sklearn \\\n",
    "        --i-classifier {classifier} \\\n",
    "        --i-reads {rep_seqs_dada2_output} \\\n",
    "        --o-classification {taxonomy_output}\"\"\"\n",
    "    \n",
    "    run_command(classify_command) \n",
    "    \n",
    "    \n",
    "def run_command(command):\n",
    "    \"\"\"\n",
    "    runs a bash command\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    command : str\n",
    "        string of the command to execute\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    none\n",
    "    \"\"\"\n",
    "    print(\"__________________________________________________________\")\n",
    "    pretty_command = re.sub('--', '\\n--', command)\n",
    "    print(f\"Running: {pretty_command}\")\n",
    "    try:\n",
    "        result = subprocess.run(command, shell=True, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, universal_newlines=True)\n",
    "        print(f\"\\n{result.stdout}\")\n",
    "        if result.stderr:\n",
    "            print(f\"\\n{result.stderr}\")\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(f\"\\nCommand failed with exit code {e.returncode}\")\n",
    "        print(f\"\\nstdout: {e.stdout}\")\n",
    "        print(f\"\\nstderr: {e.stderr}\")\n",
    "        raise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#PYTHON TRANSLATION OF THE ABOVE CODE:\n",
    "# Execute import command\n",
    "if p_s == 'PAIRED':\n",
    "    demux_output = os.path.join(temp_path, \"paired-end-demux.qza\")\n",
    "    demux = qiime2.Artifact.import_data('SampleData[PairedEndSequencesWithQuality]', manifest_path, view_type='PairedEndFastqManifestPhred33V2')\n",
    "elif p_s == 'SINGLE':\n",
    "    demux_output = os.path.join(temp_path, \"single-end-demux.qza\")\n",
    "    demux = qiime2.Artifact.import_data('SampleData[SequencesWithQuality]', manifest_path, view_type='SingleEndFastqManifestPhred33V2')\n",
    "    \n",
    "demux.save(demux_output)\n",
    "\n",
    "# Execute denoise command\n",
    "rep_seqs_dada2_output = os.path.join(temp_path, \"rep-seqs-dada2.qza\")\n",
    "table_dada2_output = os.path.join(temp_path, \"table-dada2.qza\")\n",
    "denoising_stats_output = os.path.join(summary_o_path, f\"{sub_num}-stats-dada2.qza\") \n",
    "\n",
    "if p_s == 'PAIRED':\n",
    "    denoise_results = denoise_paired(\n",
    "        demultiplexed_seqs=demux,\n",
    "        trim_left_f=trim_left_f,\n",
    "        trim_left_r=trim_left_r,\n",
    "        trunc_len_f=trunc_len_f,\n",
    "        trunc_len_r=trunc_len_r\n",
    "    )\n",
    "elif p_s == 'SINGLE':\n",
    "    denoise_results = denoise_single(\n",
    "        demultiplexed_seqs=demux,\n",
    "        trim_left=trim_left_f,\n",
    "        trunc_len=trunc_len_f\n",
    "    )\n",
    "\n",
    "denoise_results.representative_sequences.save(rep_seqs_dada2_output)\n",
    "denoise_results.table.save(table_dada2_output)\n",
    "denoise_results.denoising_stats.save(denoising_stats_output)\n",
    "\n",
    "# Execute summary command\n",
    "visualization_output = os.path.join(summary_o_path, f\"{sub_num}-table-dada2.qzv\")\n",
    "metadata = qiime2.Metadata.load(manifest_path)\n",
    "\n",
    "summary_result = summarize(\n",
    "    table=denoise_results.table,\n",
    "    sample_metadata=metadata\n",
    ")\n",
    "summary_result.visualization.save(visualization_output)\n",
    "\n",
    "# Execute classify command\n",
    "classifier = qiime2.Artifact.load(classifier_map[region_16s])\n",
    "taxonomy_output = os.path.join(profile_o_path, f\"{sub_num}-taxonomy.qza\")\n",
    "\n",
    "classification_result = classify_sklearn(\n",
    "    reads=denoise_results.representative_sequences,\n",
    "    classifier=classifier\n",
    ")\n",
    "classification_result.classification.save(taxonomy_output)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Licenzing and Citations:\n",
    "QIIME2 overall:\n",
    "\n",
    "\n",
    "QIIME2 pluging:\n",
    "\n",
    "\n",
    "classifiers:\n",
    "\n",
    "Silva:\n",
    "Michael S Robeson II, Devon R O’Rourke, Benjamin D Kaehler, Michal Ziemski, Matthew R Dillon, Jeffrey T Foster, Nicholas A Bokulich. RESCRIPt: Reproducible sequence taxonomy reference database management for the masses. bioRxiv 2020.10.05.326504; doi: https://doi.org/10.1101/2020.10.05.326504\n",
    "\n",
    "See the SILVA website for the latest citation information for SILVA. https://www.arb-silva.de/\n",
    "\n",
    "GTDB:\n",
    "UUID: 5d5461cc-6a51-434b-90ab-040f388e4221\n",
    "SHA256: 07aadcf7472d9cc6f853f6b4615348619f1a3eceb56c1fb1b6d8dbb20554765f\n",
    "Sklearn Version: 1.4.2\n",
    "Date Trained: 2024-05-30\n",
    "Citations: Parks et al. (2021), Parks et al. (2020), Parks et al. (2018), Rinke et al. (2021)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERR1351807\n",
      "<bound method NDFrame.head of               id  Year_of_release     Release_Date       Load_Date Submission  \\\n",
      "0     ERR1351807             2017  8/21/2017 23:46  8/23/2017 3:56  ERA600241   \n",
      "1     ERR1351808             2017  8/21/2017 23:46  8/23/2017 3:57  ERA600241   \n",
      "2     ERR1351809             2017  8/21/2017 23:46  8/23/2017 3:57  ERA600241   \n",
      "3     ERR1351810             2017  8/21/2017 23:46  8/23/2017 3:57  ERA600241   \n",
      "4     ERR1351811             2017  8/21/2017 23:46  8/23/2017 3:57  ERA600241   \n",
      "...          ...              ...              ...             ...        ...   \n",
      "5941  SRR9660388             2020    8/8/2020 0:00  7/9/2019 14:28  SRA919014   \n",
      "5942  SRR9660389             2020    8/8/2020 0:00  7/9/2019 14:28  SRA919014   \n",
      "5943  SRR9660390             2020    8/8/2020 0:00  7/9/2019 14:28  SRA919014   \n",
      "5944  SRR9660391             2020    8/8/2020 0:00  7/9/2019 14:29  SRA919014   \n",
      "5945  SRR9660392             2020    8/8/2020 0:00  7/9/2019 14:29  SRA919014   \n",
      "\n",
      "       Spots     Bases  Spots_with_mates  AvgLength  Size_MB  ...  \\\n",
      "0      97735  59813820             97735        612       34  ...   \n",
      "1      56514  34586568             56514        612       23  ...   \n",
      "2     135207  82746684            135207        612       46  ...   \n",
      "3      55972  34254864             55972        612       23  ...   \n",
      "4      66241  40539492             66241        612       27  ...   \n",
      "...      ...       ...               ...        ...      ...  ...   \n",
      "5941   44076  26447983             44076        600       18  ...   \n",
      "5942    7989   4795371              7989        600        3  ...   \n",
      "5943   82153  49325707             82153        600       34  ...   \n",
      "5944   53103  31867597             53103        600       22  ...   \n",
      "5945   42679  25616587             42679        600       17  ...   \n",
      "\n",
      "     Scopus_Research_Subject_5 Scopus_Research_Subject_6 Medicine_journal  \\\n",
      "0                          NaN                       NaN              Yes   \n",
      "1                          NaN                       NaN              Yes   \n",
      "2                          NaN                       NaN              Yes   \n",
      "3                          NaN                       NaN              Yes   \n",
      "4                          NaN                       NaN              Yes   \n",
      "...                        ...                       ...              ...   \n",
      "5941                       NaN                       NaN               No   \n",
      "5942                       NaN                       NaN               No   \n",
      "5943                       NaN                       NaN               No   \n",
      "5944                       NaN                       NaN               No   \n",
      "5945                       NaN                       NaN               No   \n",
      "\n",
      "                                      Study_description  \\\n",
      "0     Oral retinoids and tetracyclines have a major ...   \n",
      "1     Oral retinoids and tetracyclines have a major ...   \n",
      "2     Oral retinoids and tetracyclines have a major ...   \n",
      "3     Oral retinoids and tetracyclines have a major ...   \n",
      "4     Oral retinoids and tetracyclines have a major ...   \n",
      "...                                                 ...   \n",
      "5941                                                NaN   \n",
      "5942                                                NaN   \n",
      "5943                                                NaN   \n",
      "5944                                                NaN   \n",
      "5945                                                NaN   \n",
      "\n",
      "                                         Study_abstract Notes  \\\n",
      "0     Oral retinoids and tetracyclines have a major ...   NaN   \n",
      "1     Oral retinoids and tetracyclines have a major ...   NaN   \n",
      "2     Oral retinoids and tetracyclines have a major ...   NaN   \n",
      "3     Oral retinoids and tetracyclines have a major ...   NaN   \n",
      "4     Oral retinoids and tetracyclines have a major ...   NaN   \n",
      "...                                                 ...   ...   \n",
      "5941                                                NaN   NaN   \n",
      "5942                                                NaN   NaN   \n",
      "5943                                                NaN   NaN   \n",
      "5944                                                NaN   NaN   \n",
      "5945                                                NaN   NaN   \n",
      "\n",
      "       Manual_Validation                            Scopus_Research_Subject  \\\n",
      "0     Manually_Validated                       Medicine; NA; NA; NA; NA; NA   \n",
      "1     Manually_Validated                       Medicine; NA; NA; NA; NA; NA   \n",
      "2     Manually_Validated                       Medicine; NA; NA; NA; NA; NA   \n",
      "3     Manually_Validated                       Medicine; NA; NA; NA; NA; NA   \n",
      "4     Manually_Validated                       Medicine; NA; NA; NA; NA; NA   \n",
      "...                  ...                                                ...   \n",
      "5941  Manually_Validated  Agricultural and Biological Sciences; Biochemi...   \n",
      "5942  Manually_Validated  Agricultural and Biological Sciences; Biochemi...   \n",
      "5943  Manually_Validated  Agricultural and Biological Sciences; Biochemi...   \n",
      "5944  Manually_Validated  Agricultural and Biological Sciences; Biochemi...   \n",
      "5945  Manually_Validated  Agricultural and Biological Sciences; Biochemi...   \n",
      "\n",
      "                              forward-absolute-filepath  \\\n",
      "0     /home/miloscola/augene/training_data/paired_un...   \n",
      "1                                                    na   \n",
      "2                                                    na   \n",
      "3                                                    na   \n",
      "4                                                    na   \n",
      "...                                                 ...   \n",
      "5941                                                 na   \n",
      "5942                                                 na   \n",
      "5943                                                 na   \n",
      "5944                                                 na   \n",
      "5945                                                 na   \n",
      "\n",
      "                              reverse-absolute-filepath  \n",
      "0     /home/miloscola/augene/training_data/paired_un...  \n",
      "1                                                    na  \n",
      "2                                                    na  \n",
      "3                                                    na  \n",
      "4                                                    na  \n",
      "...                                                 ...  \n",
      "5941                                                 na  \n",
      "5942                                                 na  \n",
      "5943                                                 na  \n",
      "5944                                                 na  \n",
      "5945                                                 na  \n",
      "\n",
      "[5946 rows x 64 columns]>\n",
      "<bound method NDFrame.head of            id  Year_of_release     Release_Date       Load_Date Submission  \\\n",
      "0  ERR1351807             2017  8/21/2017 23:46  8/23/2017 3:56  ERA600241   \n",
      "\n",
      "   Spots     Bases  Spots_with_mates  AvgLength  Size_MB  ...  \\\n",
      "0  97735  59813820             97735        612       34  ...   \n",
      "\n",
      "  Scopus_Research_Subject_5 Scopus_Research_Subject_6 Medicine_journal  \\\n",
      "0                       NaN                       NaN              Yes   \n",
      "\n",
      "                                   Study_description  \\\n",
      "0  Oral retinoids and tetracyclines have a major ...   \n",
      "\n",
      "                                      Study_abstract Notes  \\\n",
      "0  Oral retinoids and tetracyclines have a major ...   NaN   \n",
      "\n",
      "    Manual_Validation       Scopus_Research_Subject  \\\n",
      "0  Manually_Validated  Medicine; NA; NA; NA; NA; NA   \n",
      "\n",
      "                           forward-absolute-filepath  \\\n",
      "0  /home/miloscola/augene/training_data/paired_un...   \n",
      "\n",
      "                           reverse-absolute-filepath  \n",
      "0  /home/miloscola/augene/training_data/paired_un...  \n",
      "\n",
      "[1 rows x 64 columns]>\n"
     ]
    }
   ],
   "source": [
    "#Remove old manifest and generate new manifest #OUTDATED\n",
    "\n",
    "import os\n",
    "import shutil\n",
    "import pandas as pd\n",
    "\n",
    "#TODO: make manifest for unpaired\n",
    "\n",
    "# Define paths\n",
    "current_dir = os.getcwd()\n",
    "paired_unproc = os.path.join(current_dir, 'paired_unproc')\n",
    "unpaired_unproc = os.path.join(current_dir, 'unpaired_unproc')\n",
    "paired_master_manifest = os.path.join(current_dir, 'DataFrame_2_Pruned_paired.tsv')\n",
    "unpaired_master_manifest = os.path.join(current_dir, 'DataFrame_2_Pruned_paired.tsv')\n",
    "\n",
    "# Remove old manifest files if they exist\n",
    "paired_manifest = os.path.join(paired_unproc, 'manifest.tsv')\n",
    "unpaired_manifest = os.path.join(unpaired_unproc, 'manifest.tsv')\n",
    "\n",
    "if os.path.exists(paired_manifest):\n",
    "    os.remove(paired_manifest)\n",
    "\n",
    "if os.path.exists(unpaired_manifest):\n",
    "    os.remove(unpaired_manifest)\n",
    "\n",
    "### Generate manifest in paired folder ####\n",
    "samples = [f for f in os.listdir(paired_unproc) if os.path.isdir(os.path.join(paired_unproc, f))]\n",
    "\n",
    "# Read the manifest file\n",
    "df = pd.read_csv(paired_master_manifest, sep='\\t', encoding='latin1')\n",
    "df.rename(columns={'Run': 'id'}, inplace=True)\n",
    "\n",
    "\n",
    "# Update the paired manifest\n",
    "for sample in samples:\n",
    "    sample_path = os.path.join(paired_unproc, sample)\n",
    "    reads = os.listdir(sample_path)\n",
    "    \n",
    "    print(sample)\n",
    "    \n",
    "    df.loc[df['id'] == sample, 'forward-absolute-filepath'] = os.path.join(sample_path, reads[0])\n",
    "    df.loc[df['id'] == sample, 'reverse-absolute-filepath'] = os.path.join(sample_path, reads[1])\n",
    "\n",
    "print(df.head)\n",
    "\n",
    "#remove all rows of df without sample\n",
    "df = df[df['forward-absolute-filepath'] != 'na']\n",
    "\n",
    "print(df.head)\n",
    "\n",
    "# Save the modified TSV\n",
    "df.to_csv(paired_manifest, sep='\\t', index=False)\n",
    "    \n",
    "# You can add similar code to handle the unpaired manifest if needed\n",
    "# For example:\n",
    "# shutil.copy(os.path.join(current_dir, 'DataFrame_2_Pruned_unpaired.tsv'), unpaired_manifest)\n",
    "# for sample in unpaired_samples:\n",
    "#     sample_path = os.path.join(unpaired_unproc, sample)\n",
    "#     update_manifest(unpaired_manifest, sample_path, sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Imported /home/miloscola/augene/training_data/paired_unproc/manifest.tsv as PairedEndFastqManifestPhred33V2 to /home/miloscola/augene/training_data/qza_files/paired-end-demux.qza\n",
      "initial processing complete\n",
      "Saved FeatureTable[Frequency] to: /home/miloscola/augene/training_data/qza_files/table-dada2.qza\n",
      "Saved FeatureData[Sequence] to: /home/miloscola/augene/training_data/qza_files/rep-seqs-dada2.qza\n",
      "Saved SampleData[DADA2Stats] to: /home/miloscola/augene/training_data/qza_files/stats-dada2.qza\n",
      "denoising complete\n",
      "Saved Visualization to: /home/miloscola/augene/training_data/qza_files/table-dada2.qzv\n",
      "feature table complete\n",
      "bar plot complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Usage: qiime taxa barplot [OPTIONS]\n",
      "\n",
      "  This visualizer produces an interactive barplot visualization of\n",
      "  taxonomies. Interactive features include multi-level sorting, plot\n",
      "  recoloring, sample relabeling, and SVG figure export.\n",
      "\n",
      "Inputs:\n",
      "  --i-table ARTIFACT FeatureTable[Frequency]\n",
      "                         Feature table to visualize at various taxonomic\n",
      "                         levels.                                    [required]\n",
      "  --i-taxonomy ARTIFACT FeatureData[Taxonomy]\n",
      "                         Taxonomic annotations for features in the provided\n",
      "                         feature table. All features in the feature table must\n",
      "                         have a corresponding taxonomic annotation. Taxonomic\n",
      "                         annotations that are not present in the feature table\n",
      "                         will be ignored.                           [required]\n",
      "Parameters:\n",
      "  --m-metadata-file METADATA...\n",
      "    (multiple            The sample metadata.\n",
      "     arguments will be   \n",
      "     merged)                                                        [required]\n",
      "Outputs:\n",
      "  --o-visualization VISUALIZATION\n",
      "                                                                    [required]\n",
      "Miscellaneous:\n",
      "  --output-dir PATH      Output unspecified results to a directory\n",
      "  --verbose / --quiet    Display verbose output to stdout and/or stderr\n",
      "                         during execution of this action. Or silence output if\n",
      "                         execution is successful (silence is golden).\n",
      "  --examples             Show usage examples and exit.\n",
      "  --citations            Show citations and exit.\n",
      "  --help                 Show this message and exit.\n",
      "\n",
      "                    There was a problem with the command:                     \n",
      " (1/1) Invalid value for '--i-taxonomy':\n",
      "  '/home/miloscola/augene/training_data/qza_files/taxonomy.qza' is not a valid\n",
      "  filepath\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#run qiime2\n",
    "\n",
    "#different sequencing areas\n",
    "#V4\n",
    "#V1-V3\n",
    "#V1-V2\n",
    "#V3-V4\n",
    "#V1-V2 and v4\n",
    "#V4-V5\n",
    "#V1-V4\n",
    "\n",
    "current_dir=$(pwd)\n",
    "paired_unproc=\"$current_dir/paired_unproc\"\n",
    "paired_proc=\"$current_dir/paired_proc\"\n",
    "\n",
    "qiime tools import \\\n",
    "    --type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "    --input-path \"$paired_unproc/manifest.tsv\" \\\n",
    "    --output-path \"$current_dir/qza_files/paired-end-demux.qza\" \\\n",
    "    --input-format PairedEndFastqManifestPhred33V2 \n",
    "\n",
    "mv $paired_unproc/manifest.tsv $current_dir/qza_files/sample-metadata.tsv\n",
    "\n",
    "#mv $paired_unproc/* $paired_proc\n",
    "        \n",
    "echo \"initial processing complete\"\n",
    "        \n",
    "qiime dada2 denoise-paired \\\n",
    "    --i-demultiplexed-seqs qza_files/paired-end-demux.qza \\\n",
    "    --p-trim-left-f 13 \\\n",
    "    --p-trim-left-r 13 \\\n",
    "    --p-trunc-len-f 150 \\\n",
    "    --p-trunc-len-r 150 \\\n",
    "    --o-representative-sequences \"$current_dir/qza_files/rep-seqs-dada2.qza\" \\\n",
    "    --o-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "    --o-denoising-stats \"$current_dir/qza_files/stats-dada2.qza\"\n",
    "        \n",
    "echo \"denoising complete\"\n",
    "        \n",
    "qiime feature-table summarize \\\n",
    "    --i-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "    --o-visualization \"$current_dir/qza_files/table-dada2.qzv\" \\\n",
    "    --m-sample-metadata-file \"$current_dir/qza_files/sample-metadata.tsv\"\n",
    "        \n",
    "echo \"feature table complete\"\n",
    "\n",
    "#FIX COMMAND INPUTS\n",
    "qiime feature-classifier classify-sklearn \\\n",
    "    --i-classifier \"$current_dir/classifier/gg-13-8-99-515-806-nb-classifier.qza\" \\\n",
    "    --i-reads \"$current_dir/qza_files/rep-seqs-dada2.qza\" \\\n",
    "    --o-classification \"$current_dir/qza_files/taxonomy.qza\"\n",
    "    \n",
    "echo \"classiication complete\"\n",
    "    \n",
    "#FIX COMMAND INPUTS\n",
    "qiime metadata tabulate \\\n",
    "    --m-input-file \"$current_dir/qza_files/taxonomy.qza\" \\\n",
    "    --o-visualization \"$current_dir/qza_files/taxonomy.qzv\"\n",
    "    \n",
    "echo \"meta data tabulation complete\"\n",
    "        \n",
    "qiime taxa barplot \\\n",
    "    --i-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "    --i-taxonomy \"$current_dir/qza_files/taxonomy.qza\" \\\n",
    "    --m-metadata-file \"$current_dir/qza_files/sample-metadata.tsv\" \\\n",
    "    --o-visualization \"$current_dir/qza_files/taxa-bar-plots.qzv\"\n",
    "        \n",
    "echo \"bar plot complete\"\n",
    "\n",
    "rm $current_dir/qza_files/sample-metadata.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/home/miloscola/augene/aggit2/SKINOME_training_data/seq_reads/ERR1514335_1.fastq.gz', '/home/miloscola/augene/aggit2/SKINOME_training_data/seq_reads/SRR5869529_1.fastq.gz']\n"
     ]
    }
   ],
   "source": [
    "#SORT RAW DATA INTO SAMPLE FILES #OUTDATED\n",
    "import os\n",
    "import shutil\n",
    "import glob\n",
    "\n",
    "# Enable extended pattern matching\n",
    "# (In Python, glob.glob() handles pattern matching)\n",
    "\n",
    "current_dir = os.getcwd()\n",
    "seq_reads = os.path.join(current_dir, \"seq_reads\")\n",
    "\n",
    "# TEST WITHOUT FOR LOOP\n",
    "files = sorted(glob.glob(os.path.join(seq_reads, \"*\"))[:2])\n",
    "print(files)\n",
    "\n",
    "# Check if files list is not empty and contains expected files\n",
    "if len(files) >= 2 and files[0].endswith(\"_1.fastq.gz\") and files[1].endswith(\"_2.fastq.gz\"):\n",
    "    print(\"IF CONDITION TRIGGERED\")\n",
    "    # Move both files to the processing directory\n",
    "    filename = os.path.basename(files[0])\n",
    "    print(filename)\n",
    "    sample_id = filename[:-11]  # Remove '_1.fastq.gz' from the filename\n",
    "\n",
    "    paired_unproc_dir = os.path.join(current_dir, \"paired_unproc\", sample_id)\n",
    "    os.makedirs(paired_unproc_dir)\n",
    "    shutil.move(files[0], paired_unproc_dir)\n",
    "    shutil.move(files[1], paired_unproc_dir)\n",
    "\n",
    "else:\n",
    "    # Move only the top file to the processing directory\n",
    "    if len(files) > 0:\n",
    "        filename = os.path.basename(files[0])\n",
    "        sample_id = filename[:-11]  # Remove '_1.fastq.gz' from the filename\n",
    "\n",
    "        unpaired_unproc_dir = os.path.join(current_dir, \"unpaired_unproc\", sample_id)\n",
    "        os.makedirs(unpaired_unproc_dir)\n",
    "        shutil.move(files[0], unpaired_unproc_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#NO FOR LOOP TEST #OUTDATED\n",
    "#Sort Raw Data\n",
    "\n",
    "# Enable extended pattern matching\n",
    "shopt -s extglob\n",
    "\n",
    "current_dir=$(pwd)\n",
    "seq_reads=\"seq_reads\"\n",
    "seq_reads=\"$current_dir/$seq_reads\"\n",
    "\n",
    "\n",
    "#TEST WITHOUT FOR LOOP\n",
    "files=($(ls -1 \"$seq_reads\" | head -n 2 | sort))\n",
    "    \n",
    "#echo \"Moved ${files[0]} and ${files[1]} to processing directory.\"\n",
    "\n",
    "if [[ \"${files[0]}\" == *_1.fastq.gz && \"${files[1]}\" == *_2.fastq.gz ]]; then\n",
    "    # Move both files to the processing directory\n",
    "    filename=${files[0]}\n",
    "    sample_id=\"${filename%_1.fastq.gz}\"\n",
    "    mkdir \"$current_dir/paired_unproc/$sample_id\" \n",
    "    mv \"$seq_reads/${files[0]}\" \"$seq_reads/${files[1]}\" \"$current_dir/paired_unproc/$sample_id\"\n",
    "\n",
    "else\n",
    "    # Move only the top file to the processing directory\n",
    "    filename=${files[0]}\n",
    "    sample_id=\"${filename%_1.fastq.gz}\"\n",
    "    mkdir \"$current_dir/unpaired_unproc/$sample_id\" \n",
    "    mv \"$seq_reads/${files[0]}\" \"$current_dir/unpaired_unproc/$sample_id\"\n",
    "fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process is interrupted.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#Sort Raw Data #REPLACE WITH CODE ABOVE WHEN DONE TESTING #OUTDATED\n",
    "\n",
    "# Enable extended pattern matching\n",
    "shopt -s extglob\n",
    "\n",
    "current_dir=$(pwd)\n",
    "seq_reads=\"seq_reads\"\n",
    "seq_reads=\"$current_dir/$seq_reads\"\n",
    "\n",
    "# Enter the loop\n",
    "while [ \"$(ls -A $seq_reads)\" ]; do\n",
    "    files=($(ls -1 \"$seq_reads\" | head -n 2 | sort))\n",
    "    \n",
    "    #echo \"Moved ${files[0]} and ${files[1]} to processing directory.\"\n",
    "\n",
    "    if [[ \"${files[0]}\" == *_1.fastq.gz && \"${files[1]}\" == *_2.fastq.gz ]]; then\n",
    "        # Move both files to the processing directory\n",
    "        filename=${files[0]}\n",
    "        sample_id=\"${filename%_1.fastq.gz}\"\n",
    "        mkdir \"$seq_reads/$sample_id\" \n",
    "        mv \"$seq_reads/${files[0]}\" \"$seq_reads/${files[1]}\" \"$current_dir/paired_unproc/$sample_id\"\n",
    "\n",
    "    else\n",
    "        # Move only the top file to the processing directory\n",
    "        mv \"$seq_reads/${files[0]}\" \"$current_dir/unpaired_unproc/\"\n",
    "    fi\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "paired_unproc\n",
      "ERR1351807\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "rm: cannot remove '/home/miloscola/augene/training_data/paired_unproc/manifest.tsv': No such file or directory\n",
      "rm: cannot remove '/home/miloscola/augene/training_data/unpaired_unproc/manifest.tsv': No such file or directory\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#Remove old manifest and Generate new Manifests\n",
    "#OUTDATED\n",
    "shopt -s extglob\n",
    "\n",
    "current_dir=$(pwd)\n",
    "paired_unproc=\"$current_dir/paired_unproc\"\n",
    "unpaired_unproc=\"$current_dir/unpaired_unproc\"\n",
    "\n",
    "# Enable extended pattern matching\n",
    "rm \"${paired_unproc}/manifest.tsv\"\n",
    "rm \"${unpaired_unproc}/manifest.tsv\"\n",
    "\n",
    "#generate manifest in paired folder\n",
    "\n",
    "# Get the list of folder names in the specified directory and store in an array\n",
    "samples=($(find \"$paired_unproc\" -maxdepth 1 -type d -exec basename {} \\;))\n",
    "\n",
    "#generate new manifest\n",
    "cp \"${current_dir}/DataFrame_2_Pruned_paired.tsv\" \"${paired_unproc}/manifest.tsv\"\n",
    "\n",
    "for sample in \"${samples[@]}\"; do\n",
    "    sample_path=\"${paired_unproc}/${sample}\"\n",
    "    reads=($(ls -1 \"$sample_path\"))\n",
    "    \n",
    "    awk -v row=\"$row_index\" -v col=\"$column_index\" -v val=\"$new_value\" -F'\\t' 'BEGIN {OFS=FS} { if (NR == row) $col = val; print }' \"$tsv_file\" > modified.tsv\n",
    "done\n",
    "\n",
    "#generate manifest in unpaired folder\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AlignedDNAFASTAFormat\n",
      "AlignedDNASequencesDirectoryFormat\n",
      "AlphaDiversityDirectoryFormat\n",
      "AlphaDiversityFormat\n",
      "BIOMV100DirFmt\n",
      "BIOMV100Format\n",
      "BIOMV210DirFmt\n",
      "BIOMV210Format\n",
      "BooleanSeriesDirectoryFormat\n",
      "BooleanSeriesFormat\n",
      "Bowtie2IndexDirFmt\n",
      "CasavaOneEightLanelessPerSampleDirFmt\n",
      "CasavaOneEightSingleLanePerSampleDirFmt\n",
      "DADA2StatsDirFmt\n",
      "DADA2StatsFormat\n",
      "DNAFASTAFormat\n",
      "DNASequencesDirectoryFormat\n",
      "DeblurStatsDirFmt\n",
      "DeblurStatsFmt\n",
      "DifferentialDirectoryFormat\n",
      "DifferentialFormat\n",
      "DistanceMatrixDirectoryFormat\n",
      "EMPPairedEndCasavaDirFmt\n",
      "EMPPairedEndDirFmt\n",
      "EMPSingleEndCasavaDirFmt\n",
      "EMPSingleEndDirFmt\n",
      "ErrorCorrectionDetailsDirFmt\n",
      "FastqGzFormat\n",
      "FirstDifferencesDirectoryFormat\n",
      "FirstDifferencesFormat\n",
      "HeaderlessTSVTaxonomyDirectoryFormat\n",
      "HeaderlessTSVTaxonomyFormat\n",
      "ImportanceDirectoryFormat\n",
      "ImportanceFormat\n",
      "LSMatFormat\n",
      "MultiplexedPairedEndBarcodeInSequenceDirFmt\n",
      "MultiplexedSingleEndBarcodeInSequenceDirFmt\n",
      "NewickDirectoryFormat\n",
      "NewickFormat\n",
      "OrdinationDirectoryFormat\n",
      "OrdinationFormat\n",
      "PairedDNASequencesDirectoryFormat\n",
      "PairedEndFastqManifestPhred33\n",
      "PairedEndFastqManifestPhred33V2\n",
      "PairedEndFastqManifestPhred64\n",
      "PairedEndFastqManifestPhred64V2\n",
      "PlacementsDirFmt\n",
      "PlacementsFormat\n",
      "PredictionsDirectoryFormat\n",
      "PredictionsFormat\n",
      "ProbabilitiesDirectoryFormat\n",
      "ProbabilitiesFormat\n",
      "QIIME1DemuxDirFmt\n",
      "QIIME1DemuxFormat\n",
      "QualityFilterStatsDirFmt\n",
      "QualityFilterStatsFmt\n",
      "SampleEstimatorDirFmt\n",
      "SeppReferenceDirFmt\n",
      "SingleEndFastqManifestPhred33\n",
      "SingleEndFastqManifestPhred33V2\n",
      "SingleEndFastqManifestPhred64\n",
      "SingleEndFastqManifestPhred64V2\n",
      "SingleLanePerSamplePairedEndFastqDirFmt\n",
      "SingleLanePerSampleSingleEndFastqDirFmt\n",
      "TSVTaxonomyDirectoryFormat\n",
      "TSVTaxonomyFormat\n",
      "TaxonomicClassiferTemporaryPickleDirFmt\n",
      "UchimeStatsDirFmt\n",
      "UchimeStatsFmt\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#import qiime tools\n",
    "qiime tools import --show-importable-formats\n",
    "\n",
    "#qiime tools import --show-importable-types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "#single for loop run test\n",
    "\n",
    "# Enable extended pattern matching\n",
    "shopt -s extglob\n",
    "\n",
    "current_dir=$(pwd)\n",
    "seq_reads=\"seq_reads\"\n",
    "seq_reads=\"$current_dir/$seq_reads\"\n",
    "\n",
    "files=($(ls -1 \"$seq_reads\" | head -n 2 | sort))\n",
    "    \n",
    "# Move both files to the processing directory\n",
    "mv \"$seq_reads/${files[0]}\" \"$seq_reads/${files[1]}\" \"$current_dir/processing/Sample1/\"\n",
    "echo \"Moved ${files[0]} and ${files[1]} to processing directory.\"\n",
    "\n",
    "if [[ \"${files[0]}\" == *_1.fastq.gz && \"${files[1]}\" == *_2.fastq.gz ]]; then\n",
    "    # Move both files to the processing directory\n",
    "    #mv \"$seq_reads/${files[0]}\" \"$seq_reads/${files[1]}\" \"$current_dir/processing/\"\n",
    "        \n",
    "    #echo \"Moved ${files[0]} and ${files[1]} to processing directory.\"\n",
    "        \n",
    "    echo \"if condition triggered\"\n",
    "        \n",
    "    #for gzip_file in $current_dir/processing/*.gz; do\n",
    "    #    # Unzip each file\n",
    "    #    gunzip -f \"$gzip_file\"\n",
    "    #done\n",
    "    \n",
    "#    T=$(printf '\\t')\n",
    "    \n",
    "#    cat > $current_dir/processing/manifest/manifest.tsv << EOF\n",
    "#sample-id $T forward-absolute-filepath $T reverse-absolute-filepath\n",
    "#sample-1 $T $current_dir/processing/Sample1/${files[0]} $T $current_dir/processing/Sample1/${files[1]}\n",
    "#EOF\n",
    "    \n",
    "#    cat <<EOF > \"$current_dir/processing/manifest/manifest.tsv\"\n",
    "#sample-id\"\\t\"forward-absolute-filepath\"\\t\"reverse-absolute-filepath\n",
    "#sample-1\"\\t\"\"$current_dir/processing/Sample1/${files[0]}\"\"\\t\"\"$current_dir/processing/Sample1/${files[1]}\"\n",
    "#EOF\n",
    "        \n",
    "    qiime tools import \\\n",
    "      --type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "      --input-path \"$current_dir/processing/manifest/manifest.tsv\" \\\n",
    "      --output-path \"$current_dir/qza_files/paired-end-demux.qza\" \\\n",
    "      --input-format PairedEndFastqManifestPhred33V2 \n",
    "        \n",
    "    #rm $current_dir/processing/manifest/manifest.tsv\n",
    "          \n",
    "    #re-zip files\n",
    "    #for file in $current_dir/processing/*; do\n",
    "    #   # Re-zip each file\n",
    "    #    gzip -f \"$file\"\n",
    "    #done\n",
    "        \n",
    "    mv $current_dir/processing/Sample1/* $current_dir/processed/\n",
    "        \n",
    "    echo \"initial processing complete\"\n",
    "        \n",
    "    qiime dada2 denoise-paired \\\n",
    "      --i-demultiplexed-seqs qza_files/paired-end-demux.qza \\\n",
    "      --p-trim-left-f 13 \\\n",
    "      --p-trim-left-r 13 \\\n",
    "      --p-trunc-len-f 150 \\\n",
    "      --p-trunc-len-r 150 \\\n",
    "      --o-representative-sequences \"$current_dir/qza_files/rep-seqs-dada2.qza\" \\\n",
    "      --o-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "      --o-denoising-stats \"$current_dir/qza_files/stats-dada2.qza\"\n",
    "        \n",
    "    echo \"denoising complete\"\n",
    "        \n",
    "    qiime feature-table summarize \\\n",
    "      --i-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "      --o-visualization \"$current_dir/qza_files/table-dada2.qzv\" \\\n",
    "      --m-sample-metadata-file \"$current_dir/qza_files/sample-metadata.tsv\"\n",
    "        \n",
    "    echo \"feature table complete\"\n",
    "    \n",
    "    #FIX COMMAND INPUTS\n",
    "    qiime feature-classifier classify-sklearn \\\n",
    "      --i-classifier \"$current_dir/classifiers/gg-13-8-99-515-806-nb-classifier.qza\" \\\n",
    "      --i-reads \"$current_dir/qza_files/rep-seqs-dada2.qza\" \\\n",
    "      --o-classification \"$current_dir/qza_files/taxonomy.qza\"\n",
    "    \n",
    "    #FIX COMMAND INPUTS\n",
    "    #qiime metadata tabulate \\\n",
    "    #--m-input-file \"$current_dir/qza_files/taxonomy.qza\" \\\n",
    "    #--o-visualization \"$current_dir/qza_files/taxonomy.qzv\"\n",
    "\n",
    "        \n",
    "    #qiime taxa barplot \\\n",
    "    #  --i-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "    #  --i-taxonomy \"$current_dir/qza_files/taxonomy.qza\" \\\n",
    "    #  --m-metadata-file \"$current_dir/qza_files/sample-metadata.tsv\" \\\n",
    "    #  --o-visualization \"$current_dir/qza_files/taxa-bar-plots.qzv\"\n",
    "        \n",
    "    #echo \"bar plot complete\"\n",
    "        \n",
    "else\n",
    "    # Move only the top file to the processing directory\n",
    "    echo \"else condition triggered\"\n",
    "    break \n",
    "    #TODO: add single read seq code\n",
    "    #mv \"${files[0]}\" processing/\n",
    "    #echo \"Moved ${files[0]} to processing directory.\"\n",
    "fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process is interrupted.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Enable extended pattern matching\n",
    "shopt -s extglob\n",
    "\n",
    "current_dir=$(pwd)\n",
    "seq_reads=\"seq_reads\"\n",
    "seq_reads=\"$current_dir/$seq_reads\"\n",
    "\n",
    "# Enter the loop\n",
    "while [ \"$(ls -A $seq_reads)\" ]; do\n",
    "    files=($(ls -1 \"$seq_reads\" | head -n 2 | sort))\n",
    "    \n",
    "    # Move both files to the processing directory\n",
    "    mv \"$seq_reads/${files[0]}\" \"$seq_reads/${files[1]}\" \"$current_dir/processing/\"\n",
    "    echo \"Moved ${files[0]} and ${files[1]} to processing directory.\"\n",
    "\n",
    "    if [[ \"${files[0]}\" == *_1.fastq.gz && \"${files[1]}\" == *_2.fastq.gz ]]; then\n",
    "        # Move both files to the processing directory\n",
    "        #mv \"$seq_reads/${files[0]}\" \"$seq_reads/${files[1]}\" \"$current_dir/processing/\"\n",
    "        \n",
    "        #echo \"Moved ${files[0]} and ${files[1]} to processing directory.\"\n",
    "        \n",
    "        echo \"if condition triggered\"\n",
    "        \n",
    "        for gzip_file in $current_dir/processing/*.gz; do\n",
    "            # Unzip each file\n",
    "            gunzip -f \"$gzip_file\"\n",
    "            # Remove the original gzip file\n",
    "            rm \"$gzip_file\"\n",
    "        done\n",
    "        \n",
    "        qiime tools import \\\n",
    "          --type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "          --input-path \"$current_dir/processing/\" \\\n",
    "          --output-path \"$current_dir/qza_files/paired-end-demux.qza\" \\\n",
    "          --input-format CasavaOneEightSingleLanePerSampleDirFmt\n",
    "          \n",
    "        #re-zip files\n",
    "        for file in $current_dir/processing/*; do\n",
    "            # Re-zip each file\n",
    "            gzip -f \"$file\"\n",
    "            #remove unzipped file\n",
    "            rm \"$file\"\n",
    "        done\n",
    "        \n",
    "        mv \"$current_dir/processing/*\" \"$current_dir/processed/\"\n",
    "        \n",
    "        echo \"initial processing complete\"\n",
    "        \n",
    "        qiime dada2 denoise-paired \\\n",
    "          --i-demultiplexed-seqs qza_files/paired-end-demux.qza \\\n",
    "          --p-trim-left-f 13 \\\n",
    "          --p-trim-left-r 13 \\\n",
    "          --p-trunc-len-f 150 \\\n",
    "          --p-trunc-len-r 150 \\\n",
    "          --o-representative-sequences \"$current_dir/qza_files/rep-seqs-dada2.qza\" \\\n",
    "          --o-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "          --o-denoising-stats \"$current_dir/qza_files/stats-dada2.qza\"\n",
    "        \n",
    "        echo \"denoising complete\"\n",
    "        \n",
    "        qiime feature-table summarize \\\n",
    "          --i-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "          --o-visualization \"$current_dir/qza_files/table-dada2.qzv\" \\\n",
    "          --m-sample-metadata-file \"$current_dir/qza_files/sample-metadata.tsv\"\n",
    "        \n",
    "        echo \"feature table complete\"\n",
    "        \n",
    "        qiime taxa barplot \\\n",
    "          --i-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "          --i-taxonomy \"$current_dir/qza_files/taxonomy.qza\" \\\n",
    "          --m-metadata-file \"$current_dir/qza_files/sample-metadata.tsv\" \\\n",
    "          --o-visualization \"$current_dir/qza_files/taxa-bar-plots.qzv\"\n",
    "        \n",
    "        echo \"bar plot complete\"\n",
    "        \n",
    "    else\n",
    "        # Move only the top file to the processing directory\n",
    "        echo \"else condition triggered\"\n",
    "        break \n",
    "        #TODO: add single read seq code\n",
    "        #mv \"${files[0]}\" processing/\n",
    "        #echo \"Moved ${files[0]} to processing directory.\"\n",
    "    fi\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "System versions\n",
      "Python version: 3.6.10\n",
      "QIIME 2 release: 2020.8\n",
      "QIIME 2 version: 2020.8.0\n",
      "q2cli version: 2020.8.0\n",
      "\n",
      "Installed plugins\n",
      "alignment: 2020.8.0\n",
      "composition: 2020.8.0\n",
      "cutadapt: 2020.8.0\n",
      "dada2: 2020.8.0\n",
      "deblur: 2020.8.0\n",
      "demux: 2020.8.0\n",
      "diversity: 2020.8.0\n",
      "diversity-lib: 2020.8.0\n",
      "emperor: 2020.8.0\n",
      "feature-classifier: 2020.8.0\n",
      "feature-table: 2020.8.0\n",
      "fragment-insertion: 2020.8.0\n",
      "gneiss: 2020.8.0\n",
      "longitudinal: 2020.8.0\n",
      "metadata: 2020.8.0\n",
      "phylogeny: 2020.8.0\n",
      "quality-control: 2020.8.0\n",
      "quality-filter: 2020.8.0\n",
      "sample-classifier: 2020.8.0\n",
      "taxa: 2020.8.0\n",
      "types: 2020.8.0\n",
      "vsearch: 2020.8.0\n",
      "\n",
      "Application config directory\n",
      "/home/miloscola/anaconda3/envs/qiime2/var/q2cli\n",
      "\n",
      "Getting help\n",
      "To get help with QIIME 2, visit https://qiime2.org\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "qiime info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Moved ERR1351843_R1_001.fastq.gz and ERR1351843_R2_001.fastq.gz to processing directory.\n",
      "if condition triggered\n",
      "Imported /home/miloscola/augene/training_data/processing/manifest/manifest.tsv as PairedEndFastqManifestPhred33V2 to /home/miloscola/augene/training_data/qza_files/paired-end-demux.qza\n",
      "initial processing complete\n",
      "denoising complete\n",
      "feature table complete\n",
      "bar plot complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plugin error from dada2:\n",
      "\n",
      "  An error was encountered while running DADA2 in R (return code 1), please inspect stdout and stderr to learn more.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-zwl7r5vc.log\n",
      "There was an issue with loading the file /home/miloscola/augene/training_data/qza_files/sample-metadata.tsv as metadata:\n",
      "\n",
      "  Metadata file path doesn't exist, or the path points to something other than a file. Please check that the path exists, has read permissions, and points to a regular file (not a directory): /home/miloscola/augene/training_data/qza_files/sample-metadata.tsv\n",
      "\n",
      "  There may be more errors present in the metadata file. To get a full report, sample/feature metadata files can be validated with Keemei: https://keemei.qiime2.org\n",
      "\n",
      "  Find details on QIIME 2 metadata requirements here: https://docs.qiime2.org/2020.8/tutorials/metadata/\n",
      "\n",
      "There was an issue with loading the file /home/miloscola/augene/training_data/qza_files/sample-metadata.tsv as metadata:\n",
      "\n",
      "  Metadata file path doesn't exist, or the path points to something other than a file. Please check that the path exists, has read permissions, and points to a regular file (not a directory): /home/miloscola/augene/training_data/qza_files/sample-metadata.tsv\n",
      "\n",
      "  There may be more errors present in the metadata file. To get a full report, sample/feature metadata files can be validated with Keemei: https://keemei.qiime2.org\n",
      "\n",
      "  Find details on QIIME 2 metadata requirements here: https://docs.qiime2.org/2020.8/tutorials/metadata/\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# Enable extended pattern matching\n",
    "shopt -s extglob\n",
    "\n",
    "current_dir=$(pwd)\n",
    "seq_reads=\"seq_reads\"\n",
    "seq_reads=\"$current_dir/$seq_reads\"\n",
    "\n",
    "files=($(ls -1 \"$seq_reads\" | head -n 2 | sort))\n",
    "    \n",
    "# Move both files to the processing directory\n",
    "mv \"$seq_reads/${files[0]}\" \"$seq_reads/${files[1]}\" \"$current_dir/processing/Sample1/\"\n",
    "echo \"Moved ${files[0]} and ${files[1]} to processing directory.\"\n",
    "\n",
    "if [[ \"${files[0]}\" == *_1.fastq.gz && \"${files[1]}\" == *_2.fastq.gz ]]; then\n",
    "    # Move both files to the processing directory\n",
    "    #mv \"$seq_reads/${files[0]}\" \"$seq_reads/${files[1]}\" \"$current_dir/processing/\"\n",
    "        \n",
    "    #echo \"Moved ${files[0]} and ${files[1]} to processing directory.\"\n",
    "        \n",
    "    echo \"if condition triggered\"\n",
    "        \n",
    "    #for gzip_file in $current_dir/processing/*.gz; do\n",
    "    #    # Unzip each file\n",
    "    #    gunzip -f \"$gzip_file\"\n",
    "    #done\n",
    "    \n",
    "    T=$(printf '\\t')\n",
    "    \n",
    "    cat > $current_dir/processing/manifest/manifest.tsv << EOF\n",
    "sample-id $T forward-absolute-filepath $T reverse-absolute-filepath\n",
    "sample-1 $T $current_dir/processing/Sample1/${files[0]} $T $current_dir/processing/Sample1/${files[1]}\n",
    "EOF\n",
    "    \n",
    "#    cat <<EOF > \"$current_dir/processing/manifest/manifest.tsv\"\n",
    "#sample-id\"\\t\"forward-absolute-filepath\"\\t\"reverse-absolute-filepath\n",
    "#sample-1\"\\t\"\"$current_dir/processing/Sample1/${files[0]}\"\"\\t\"\"$current_dir/processing/Sample1/${files[1]}\"\n",
    "#EOF\n",
    "        \n",
    "    qiime tools import \\\n",
    "        --type 'SampleData[PairedEndSequencesWithQuality]' \\\n",
    "        --input-path \"$current_dir/processing/manifest/manifest.tsv\" \\\n",
    "        --output-path \"$current_dir/qza_files/paired-end-demux.qza\" \\\n",
    "        --input-format PairedEndFastqManifestPhred33V2 \n",
    "        \n",
    "    #rm $current_dir/processing/manifest/manifest.tsv\n",
    "          \n",
    "    #re-zip files\n",
    "    #for file in $current_dir/processing/*; do\n",
    "    #   # Re-zip each file\n",
    "    #    gzip -f \"$file\"\n",
    "    #done\n",
    "        \n",
    "    mv $current_dir/processing/Sample1/* $current_dir/processed/\n",
    "        \n",
    "    echo \"initial processing complete\"\n",
    "        \n",
    "    qiime dada2 denoise-paired \\\n",
    "      --i-demultiplexed-seqs qza_files/paired-end-demux.qza \\\n",
    "      --p-trim-left-f 13 \\\n",
    "      --p-trim-left-r 13 \\\n",
    "      --p-trunc-len-f 150 \\\n",
    "      --p-trunc-len-r 150 \\\n",
    "      --o-representative-sequences \"$current_dir/qza_files/rep-seqs-dada2.qza\" \\\n",
    "      --o-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "      --o-denoising-stats \"$current_dir/qza_files/stats-dada2.qza\"\n",
    "        \n",
    "    echo \"denoising complete\"\n",
    "        \n",
    "    qiime feature-table summarize \\\n",
    "      --i-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "      --o-visualization \"$current_dir/qza_files/table-dada2.qzv\" \\\n",
    "      --m-sample-metadata-file \"$current_dir/qza_files/sample-metadata.tsv\"\n",
    "        \n",
    "    echo \"feature table complete\"\n",
    "        \n",
    "    qiime taxa barplot \\\n",
    "      --i-table \"$current_dir/qza_files/table-dada2.qza\" \\\n",
    "      --i-taxonomy \"$current_dir/qza_files/taxonomy.qza\" \\\n",
    "      --m-metadata-file \"$current_dir/qza_files/sample-metadata.tsv\" \\\n",
    "      --o-visualization \"$current_dir/qza_files/taxa-bar-plots.qzv\"\n",
    "        \n",
    "    echo \"bar plot complete\"\n",
    "        \n",
    "else\n",
    "    # Move only the top file to the processing directory\n",
    "    echo \"else condition triggered\"\n",
    "    break \n",
    "    #TODO: add single read seq code\n",
    "    #mv \"${files[0]}\" processing/\n",
    "    #echo \"Moved ${files[0]} to processing directory.\"\n",
    "fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO:\n",
    "#0) standardize file names -DONE\n",
    "#1) make seprate folders for each sample **account for single and paired reads\n",
    "#2) make a manifest with sample names corresponding \n",
    "#   to the sample names in the metadata TSV for all files -DONE\n",
    "#3) run pipeline with all files all at once"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Greengeens 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing /home/miloscola/.cache/pip/wheels/ad/64/10/9d66020ccd16f0754e1059f75c0fb0162dba4792f4fdf7a8bf/q2_greengenes2-2024.1-py3-none-any.whl\n",
      "Requirement already satisfied: redbiom in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from q2-greengenes2) (0.3.9)\n",
      "Collecting iow\n",
      "  Using cached iow-1.0.7.tar.gz (1.2 MB)\n",
      "Requirement already satisfied: biom-format in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from q2-greengenes2) (2.1.8)\n",
      "Requirement already satisfied: scikit-bio in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from q2-greengenes2) (0.5.5)\n",
      "Requirement already satisfied: msgpack in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom->q2-greengenes2) (1.0.0)\n",
      "Requirement already satisfied: requests in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom->q2-greengenes2) (2.24.0)\n",
      "Requirement already satisfied: pandas in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom->q2-greengenes2) (0.25.3)\n",
      "Requirement already satisfied: nltk in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom->q2-greengenes2) (3.6.7)\n",
      "Requirement already satisfied: h5py in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom->q2-greengenes2) (2.10.0)\n",
      "Requirement already satisfied: click>=6.7 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom->q2-greengenes2) (7.1.2)\n",
      "Requirement already satisfied: joblib in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom->q2-greengenes2) (0.16.0)\n",
      "Requirement already satisfied: numpy>=1.9.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from iow->q2-greengenes2) (1.19.1)\n",
      "Requirement already satisfied: nose>=1.3.7 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from iow->q2-greengenes2) (1.3.7)\n",
      "Requirement already satisfied: cython<1.0.0,>=0.24.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from iow->q2-greengenes2) (0.29.21)\n",
      "Requirement already satisfied: future>=0.16.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from biom-format->q2-greengenes2) (0.18.2)\n",
      "Requirement already satisfied: six>=1.10.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from biom-format->q2-greengenes2) (1.15.0)\n",
      "Requirement already satisfied: scipy>=1.3.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from biom-format->q2-greengenes2) (1.5.2)\n",
      "Requirement already satisfied: scikit-learn>=0.19.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio->q2-greengenes2) (0.23.1)\n",
      "Requirement already satisfied: IPython>=3.2.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio->q2-greengenes2) (7.16.1)\n",
      "Requirement already satisfied: decorator>=3.4.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio->q2-greengenes2) (4.4.2)\n",
      "Requirement already satisfied: natsort>=4.0.3 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio->q2-greengenes2) (7.0.1)\n",
      "Requirement already satisfied: lockfile>=0.10.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio->q2-greengenes2) (0.12.2)\n",
      "Requirement already satisfied: CacheControl>=0.11.5 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio->q2-greengenes2) (0.12.6)\n",
      "Requirement already satisfied: matplotlib>=1.4.3 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio->q2-greengenes2) (3.2.2)\n",
      "Requirement already satisfied: hdmedians>=0.13 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio->q2-greengenes2) (0.13)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from requests->redbiom->q2-greengenes2) (1.25.10)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from requests->redbiom->q2-greengenes2) (2.10)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from requests->redbiom->q2-greengenes2) (2021.5.30)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from requests->redbiom->q2-greengenes2) (3.0.4)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from pandas->redbiom->q2-greengenes2) (2020.1)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from pandas->redbiom->q2-greengenes2) (2.8.1)\n",
      "Requirement already satisfied: tqdm in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from nltk->redbiom->q2-greengenes2) (4.64.1)\n",
      "Requirement already satisfied: regex>=2021.8.3 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from nltk->redbiom->q2-greengenes2) (2023.8.8)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-learn>=0.19.1->scikit-bio->q2-greengenes2) (2.1.0)\n",
      "Requirement already satisfied: jedi>=0.10 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio->q2-greengenes2) (0.17.2)\n",
      "Requirement already satisfied: pygments in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio->q2-greengenes2) (2.6.1)\n",
      "Requirement already satisfied: pexpect; sys_platform != \"win32\" in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio->q2-greengenes2) (4.8.0)\n",
      "Requirement already satisfied: setuptools>=18.5 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio->q2-greengenes2) (49.6.0.post20210108)\n",
      "Requirement already satisfied: backcall in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio->q2-greengenes2) (0.2.0)\n",
      "Requirement already satisfied: pickleshare in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio->q2-greengenes2) (0.7.5)\n",
      "Requirement already satisfied: traitlets>=4.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio->q2-greengenes2) (4.3.3)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio->q2-greengenes2) (3.0.6)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from matplotlib>=1.4.3->scikit-bio->q2-greengenes2) (2.4.7)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from matplotlib>=1.4.3->scikit-bio->q2-greengenes2) (0.10.0)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from matplotlib>=1.4.3->scikit-bio->q2-greengenes2) (1.2.0)\n",
      "Requirement already satisfied: importlib-resources; python_version < \"3.7\" in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from tqdm->nltk->redbiom->q2-greengenes2) (5.4.0)\n",
      "Requirement already satisfied: parso<0.8.0,>=0.7.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from jedi>=0.10->IPython>=3.2.0->scikit-bio->q2-greengenes2) (0.7.1)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from pexpect; sys_platform != \"win32\"->IPython>=3.2.0->scikit-bio->q2-greengenes2) (0.6.0)\n",
      "Requirement already satisfied: ipython-genutils in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from traitlets>=4.2->IPython>=3.2.0->scikit-bio->q2-greengenes2) (0.2.0)\n",
      "Requirement already satisfied: wcwidth in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->IPython>=3.2.0->scikit-bio->q2-greengenes2) (0.2.5)\n",
      "Requirement already satisfied: zipp>=3.1.0; python_version < \"3.10\" in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from importlib-resources; python_version < \"3.7\"->tqdm->nltk->redbiom->q2-greengenes2) (3.1.0)\n",
      "Building wheels for collected packages: iow\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  Building wheel for iow (setup.py) ... \u001b[?25lerror\n",
      "\u001b[31m  ERROR: Command errored out with exit status 1:\n",
      "   command: /home/miloscola/anaconda3/envs/qiime2/bin/python3.6 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '\"'\"'/tmp/pip-install-r9m06ags/iow/setup.py'\"'\"'; __file__='\"'\"'/tmp/pip-install-r9m06ags/iow/setup.py'\"'\"';f=getattr(tokenize, '\"'\"'open'\"'\"', open)(__file__);code=f.read().replace('\"'\"'\\r\\n'\"'\"', '\"'\"'\\n'\"'\"');f.close();exec(compile(code, __file__, '\"'\"'exec'\"'\"'))' bdist_wheel -d /tmp/pip-wheel-m6go0kak\n",
      "       cwd: /tmp/pip-install-r9m06ags/iow/\n",
      "  Complete output (36 lines):\n",
      "  running bdist_wheel\n",
      "  running build\n",
      "  running build_py\n",
      "  make: Entering directory '/tmp/pip-install-r9m06ags/iow/bp/BitArray'\n",
      "  make: 'libbitarr.a' is up to date.\n",
      "  make: Leaving directory '/tmp/pip-install-r9m06ags/iow/bp/BitArray'\n",
      "  creating build\n",
      "  creating build/lib.linux-x86_64-3.6\n",
      "  creating build/lib.linux-x86_64-3.6/bp\n",
      "  copying bp/_cli.py -> build/lib.linux-x86_64-3.6/bp\n",
      "  copying bp/__init__.py -> build/lib.linux-x86_64-3.6/bp\n",
      "  copying bp/_version.py -> build/lib.linux-x86_64-3.6/bp\n",
      "  creating build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "  copying bp/BitArray/bit_macros.h -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "  copying bp/BitArray/README.md -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "  copying bp/BitArray/AUTHORS -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "  copying bp/BitArray/Makefile -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "  copying bp/BitArray/bit_array.h -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "  copying bp/BitArray/bit_array.c -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "  copying bp/BitArray/libbitarr.a -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "  copying bp/BitArray/LICENSE -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "  copying bp/BitArray/bit_array.o -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "  copying bp/BitArray/bar.h -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "  UPDATING build/lib.linux-x86_64-3.6/bp/_version.py\n",
      "  set build/lib.linux-x86_64-3.6/bp/_version.py to '1.0.7'\n",
      "  running build_ext\n",
      "  building 'bp._bp' extension\n",
      "  creating build/temp.linux-x86_64-3.6\n",
      "  creating build/temp.linux-x86_64-3.6/bp\n",
      "  /home/miloscola/anaconda3/envs/qiime2/bin/x86_64-conda-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/miloscola/anaconda3/envs/qiime2/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/miloscola/anaconda3/envs/qiime2/include -fPIC -I/tmp/pip-install-r9m06ags/iow/bp/BitArray -I/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/numpy/core/include -I/tmp/pip-install-r9m06ags/iow/bp/BitArray -I/home/miloscola/anaconda3/envs/qiime2/include/python3.6m -c bp/_bp.c -o build/temp.linux-x86_64-3.6/bp/_bp.o\n",
      "  In file included from bp/_bp.c:40:0:\n",
      "  /home/miloscola/anaconda3/envs/qiime2/include/python3.6m/Python.h:39:10: fatal error: crypt.h: No such file or directory\n",
      "   #include <crypt.h>\n",
      "            ^~~~~~~~~\n",
      "  compilation terminated.\n",
      "  error: command '/home/miloscola/anaconda3/envs/qiime2/bin/x86_64-conda-linux-gnu-cc' failed with exit status 1\n",
      "  ----------------------------------------\u001b[0m\n",
      "\u001b[31m  ERROR: Failed building wheel for iow\u001b[0m\n",
      "\u001b[?25h  Running setup.py clean for iow\n",
      "Failed to build iow\n",
      "\u001b[33mDEPRECATION: Could not build wheels for iow which do not use PEP 517. pip will fall back to legacy 'setup.py install' for these. pip 21.0 will remove support for this functionality. A possible replacement is to fix the wheel build issue reported above. You can find discussion regarding this at https://github.com/pypa/pip/issues/8368.\u001b[0m\n",
      "Installing collected packages: iow, q2-greengenes2\n",
      "    Running setup.py install for iow ... \u001b[?25lerror\n",
      "\u001b[31m    ERROR: Command errored out with exit status 1:\n",
      "     command: /home/miloscola/anaconda3/envs/qiime2/bin/python3.6 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '\"'\"'/tmp/pip-install-r9m06ags/iow/setup.py'\"'\"'; __file__='\"'\"'/tmp/pip-install-r9m06ags/iow/setup.py'\"'\"';f=getattr(tokenize, '\"'\"'open'\"'\"', open)(__file__);code=f.read().replace('\"'\"'\\r\\n'\"'\"', '\"'\"'\\n'\"'\"');f.close();exec(compile(code, __file__, '\"'\"'exec'\"'\"'))' install --record /tmp/pip-record-odt_yt0i/install-record.txt --single-version-externally-managed --compile --install-headers /home/miloscola/anaconda3/envs/qiime2/include/python3.6m/iow\n",
      "         cwd: /tmp/pip-install-r9m06ags/iow/\n",
      "    Complete output (39 lines):\n",
      "    running install\n",
      "    make: Entering directory '/tmp/pip-install-r9m06ags/iow/bp/BitArray'\n",
      "    make: 'libbitarr.a' is up to date.\n",
      "    make: Leaving directory '/tmp/pip-install-r9m06ags/iow/bp/BitArray'\n",
      "    running build\n",
      "    running build_py\n",
      "    make: Entering directory '/tmp/pip-install-r9m06ags/iow/bp/BitArray'\n",
      "    make: 'libbitarr.a' is up to date.\n",
      "    make: Leaving directory '/tmp/pip-install-r9m06ags/iow/bp/BitArray'\n",
      "    creating build\n",
      "    creating build/lib.linux-x86_64-3.6\n",
      "    creating build/lib.linux-x86_64-3.6/bp\n",
      "    copying bp/_cli.py -> build/lib.linux-x86_64-3.6/bp\n",
      "    copying bp/__init__.py -> build/lib.linux-x86_64-3.6/bp\n",
      "    copying bp/_version.py -> build/lib.linux-x86_64-3.6/bp\n",
      "    creating build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "    copying bp/BitArray/bit_macros.h -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "    copying bp/BitArray/README.md -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "    copying bp/BitArray/AUTHORS -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "    copying bp/BitArray/Makefile -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "    copying bp/BitArray/bit_array.h -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "    copying bp/BitArray/bit_array.c -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "    copying bp/BitArray/libbitarr.a -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "    copying bp/BitArray/LICENSE -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "    copying bp/BitArray/bit_array.o -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "    copying bp/BitArray/bar.h -> build/lib.linux-x86_64-3.6/bp/BitArray\n",
      "    UPDATING build/lib.linux-x86_64-3.6/bp/_version.py\n",
      "    set build/lib.linux-x86_64-3.6/bp/_version.py to '1.0.7'\n",
      "    running build_ext\n",
      "    building 'bp._bp' extension\n",
      "    creating build/temp.linux-x86_64-3.6\n",
      "    creating build/temp.linux-x86_64-3.6/bp\n",
      "    /home/miloscola/anaconda3/envs/qiime2/bin/x86_64-conda-linux-gnu-cc -DNDEBUG -fwrapv -O2 -Wall -Wstrict-prototypes -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/miloscola/anaconda3/envs/qiime2/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/miloscola/anaconda3/envs/qiime2/include -fPIC -I/tmp/pip-install-r9m06ags/iow/bp/BitArray -I/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/numpy/core/include -I/tmp/pip-install-r9m06ags/iow/bp/BitArray -I/home/miloscola/anaconda3/envs/qiime2/include/python3.6m -c bp/_bp.c -o build/temp.linux-x86_64-3.6/bp/_bp.o\n",
      "    In file included from bp/_bp.c:40:0:\n",
      "    /home/miloscola/anaconda3/envs/qiime2/include/python3.6m/Python.h:39:10: fatal error: crypt.h: No such file or directory\n",
      "     #include <crypt.h>\n",
      "              ^~~~~~~~~\n",
      "    compilation terminated.\n",
      "    error: command '/home/miloscola/anaconda3/envs/qiime2/bin/x86_64-conda-linux-gnu-cc' failed with exit status 1\n",
      "    ----------------------------------------\u001b[0m\n",
      "\u001b[31mERROR: Command errored out with exit status 1: /home/miloscola/anaconda3/envs/qiime2/bin/python3.6 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '\"'\"'/tmp/pip-install-r9m06ags/iow/setup.py'\"'\"'; __file__='\"'\"'/tmp/pip-install-r9m06ags/iow/setup.py'\"'\"';f=getattr(tokenize, '\"'\"'open'\"'\"', open)(__file__);code=f.read().replace('\"'\"'\\r\\n'\"'\"', '\"'\"'\\n'\"'\"');f.close();exec(compile(code, __file__, '\"'\"'exec'\"'\"'))' install --record /tmp/pip-record-odt_yt0i/install-record.txt --single-version-externally-managed --compile --install-headers /home/miloscola/anaconda3/envs/qiime2/include/python3.6m/iow Check the logs for full command output.\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: redbiom in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (0.3.9)\n",
      "Requirement already satisfied: joblib in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom) (0.16.0)\n",
      "Requirement already satisfied: msgpack in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom) (1.0.0)\n",
      "Requirement already satisfied: click>=6.7 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom) (7.1.2)\n",
      "Requirement already satisfied: biom-format>=2.1.5 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom) (2.1.8)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: requests in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom) (2.24.0)\n",
      "Requirement already satisfied: nltk in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom) (3.6.7)\n",
      "Requirement already satisfied: scikit-bio>=0.4.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom) (0.5.5)\n",
      "Requirement already satisfied: h5py in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom) (2.10.0)\n",
      "Requirement already satisfied: pandas in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from redbiom) (0.25.3)\n",
      "Requirement already satisfied: future>=0.16.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from biom-format>=2.1.5->redbiom) (0.18.2)\n",
      "Requirement already satisfied: cython>=0.29 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from biom-format>=2.1.5->redbiom) (0.29.21)\n",
      "Requirement already satisfied: numpy>=1.9.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from biom-format>=2.1.5->redbiom) (1.19.1)\n",
      "Requirement already satisfied: scipy>=1.3.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from biom-format>=2.1.5->redbiom) (1.5.2)\n",
      "Requirement already satisfied: six>=1.10.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from biom-format>=2.1.5->redbiom) (1.15.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from requests->redbiom) (2021.5.30)\n",
      "Requirement already satisfied: idna<3,>=2.5 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from requests->redbiom) (2.10)\n",
      "Requirement already satisfied: chardet<4,>=3.0.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from requests->redbiom) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from requests->redbiom) (1.25.10)\n",
      "Requirement already satisfied: regex>=2021.8.3 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from nltk->redbiom) (2023.8.8)\n",
      "Requirement already satisfied: tqdm in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from nltk->redbiom) (4.64.1)\n",
      "Requirement already satisfied: lockfile>=0.10.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio>=0.4.2->redbiom) (0.12.2)\n",
      "Requirement already satisfied: matplotlib>=1.4.3 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio>=0.4.2->redbiom) (3.2.2)\n",
      "Requirement already satisfied: decorator>=3.4.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio>=0.4.2->redbiom) (4.4.2)\n",
      "Requirement already satisfied: CacheControl>=0.11.5 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio>=0.4.2->redbiom) (0.12.6)\n",
      "Requirement already satisfied: scikit-learn>=0.19.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio>=0.4.2->redbiom) (0.23.1)\n",
      "Requirement already satisfied: hdmedians>=0.13 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio>=0.4.2->redbiom) (0.13)\n",
      "Requirement already satisfied: natsort>=4.0.3 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio>=0.4.2->redbiom) (7.0.1)\n",
      "Requirement already satisfied: IPython>=3.2.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-bio>=0.4.2->redbiom) (7.16.1)\n",
      "Requirement already satisfied: pytz>=2017.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from pandas->redbiom) (2020.1)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from pandas->redbiom) (2.8.1)\n",
      "Requirement already satisfied: importlib-resources; python_version < \"3.7\" in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from tqdm->nltk->redbiom) (5.4.0)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from matplotlib>=1.4.3->scikit-bio>=0.4.2->redbiom) (2.4.7)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from matplotlib>=1.4.3->scikit-bio>=0.4.2->redbiom) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from matplotlib>=1.4.3->scikit-bio>=0.4.2->redbiom) (0.10.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from scikit-learn>=0.19.1->scikit-bio>=0.4.2->redbiom) (2.1.0)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (3.0.6)\n",
      "Requirement already satisfied: backcall in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (0.2.0)\n",
      "Requirement already satisfied: pexpect; sys_platform != \"win32\" in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (4.8.0)\n",
      "Requirement already satisfied: pygments in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (2.6.1)\n",
      "Requirement already satisfied: jedi>=0.10 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (0.17.2)\n",
      "Requirement already satisfied: pickleshare in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (0.7.5)\n",
      "Requirement already satisfied: traitlets>=4.2 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (4.3.3)\n",
      "Requirement already satisfied: setuptools>=18.5 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (49.6.0.post20210108)\n",
      "Requirement already satisfied: zipp>=3.1.0; python_version < \"3.10\" in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from importlib-resources; python_version < \"3.7\"->tqdm->nltk->redbiom) (3.1.0)\n",
      "Requirement already satisfied: wcwidth in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0->IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (0.2.5)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from pexpect; sys_platform != \"win32\"->IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (0.6.0)\n",
      "Requirement already satisfied: parso<0.8.0,>=0.7.0 in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from jedi>=0.10->IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (0.7.1)\n",
      "Requirement already satisfied: ipython-genutils in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from traitlets>=4.2->IPython>=3.2.0->scikit-bio>=0.4.2->redbiom) (0.2.0)\n"
     ]
    }
   ],
   "source": [
    "!pip install q2-greengenes2\n",
    "!pip install redbiom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def qiime2_profile(sub_num, sub_path, master_manifest_path, profile_o_path, summary_o_folder, classifier_map, trim_left_f=13, trim_left_r=13, trunc_len_f=150, trunc_len_r=150, n_threads=8):\n",
    "    \"\"\"\n",
    "    calculates microbiome profiles from a submission of reads and stores the results in a .qza file in the specified folder\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    sub_num : str\n",
    "        ID of the submission containing the reads.\n",
    "    sub_path : str\n",
    "        path to the submission.\n",
    "    master_manifest_path : str\n",
    "        path to the master meta data file (should .tsv file) Manifest must \n",
    "        have the following columns: Run or id, Library_Layout, Region_16S, \n",
    "        Submission\n",
    "    profile_o_path : str\n",
    "        path to the output folder for microbiome profile\n",
    "    summary_o_path : str:\n",
    "        path to the output folder for read summary table\n",
    "    classifier_map : dict\n",
    "        map of 16s rRNA region to classifier built to classify that region\n",
    "    trim_left_f : int\n",
    "        number of nucleotides to trim from the 5' end of the forward read before denoising\n",
    "    trim_left_r : int\n",
    "        number of nucleotides to trim from the 5' end of the reverse read before denoising\n",
    "    trim_left_f : int\n",
    "        specifies the length at which the forward read will be truncated\n",
    "    trim_left_r : int\n",
    "        specifies the length at which the reverse read will be truncated\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    none\n",
    "    \"\"\"\n",
    "    \n",
    "    #define variables\n",
    "    current_dir = os.getcwd()\n",
    "    summary_o_path = os.path.join(summary_o_folder, sub_num)\n",
    "    manifest_path = os.path.join(summary_o_path, f\"{sub_num}-manifest.tsv\")\n",
    "    \n",
    "    #string representing weather reads are paired or single\n",
    "    p_s = None           \n",
    "    #string representing what 16s regions were used\n",
    "    region_16s = None           \n",
    "\n",
    "    #validate inputs\n",
    "    if not type(sub_num) == str:\n",
    "        raise TypeEror(\"sub_num must be a string\")\n",
    "    if not os.path.exists(sub_path):\n",
    "        raise FileNotFoundError(f\"{sub_num} submission path is invalid\")\n",
    "    if not os.path.exists(master_manifest_path):\n",
    "        raise FileNotFoundError(\"master manifest path is invalid\")\n",
    "    if not type(classifier_map) == dict:\n",
    "        raise TypeError(\"classifier_map must be a dict\")\n",
    "    for classifier in classifier_map.values():\n",
    "        if not os.path.exists(classifier):\n",
    "            raise FileNotFoundError(f\"classifier map contains invalid paths. {classifier} is not a valid path\")\n",
    "    \n",
    "    #if output folders do not exist, create them\n",
    "    os.makedirs(profile_o_path, exist_ok=True)\n",
    "    os.makedirs(summary_o_folder, exist_ok=True)\n",
    "    os.makedirs(summary_o_path, exist_ok=True)\n",
    "    \n",
    "    #generate manifest\n",
    "    manifest = pd.read_csv(master_manifest_path, sep='\\t', encoding='latin1')\n",
    "    \n",
    "    #Rename \"Run\" to 'id' to support compatability with qiime2\n",
    "    if 'Run' in manifest.columns:\n",
    "        manifest.rename(columns={'Run': 'id'}, inplace=True)\n",
    "    \n",
    "    #prune manifest\n",
    "    manifest = manifest[manifest['Submission'] == sub_num]\n",
    "    reads = os.listdir(sub_path)\n",
    "    len_manifest = len(manifest)\n",
    "    len_reads = len(reads)\n",
    "    manifest = manifest[manifest['id'].isin(reads)]\n",
    "    len_filtered_manifest = len(manifest)\n",
    "    \n",
    "    #check that names and number of samples match on manifest and in folder\n",
    "    if manifest.empty:\n",
    "        raise ValueError(\"Pruned manifest is empty. Submission number may not be in manifest or reads may be improperly named.\")\n",
    "    if len_manifest > len_reads:\n",
    "        print(\"Warning: more entries in manifest than in folder\")\n",
    "    if len_manifest < len_reads:\n",
    "        print(\"Warning: more entries in folder than in manifest\")\n",
    "    if len_manifest == len_reads and len_filtered_manifest < len_reads:\n",
    "        print(\"Warning: mismatch between some names on manifest and in folder\")\n",
    "        \n",
    "    #check that the submission does not have mixed single and paired reads\n",
    "    unique_layouts = manifest['Library_Layout'].unique().tolist()\n",
    "    \n",
    "    if len(unique_layouts) > 1:\n",
    "        raise ValueError(\"Multiple layouts detected. Submission may contain single and paired reads.\")\n",
    "    p_s = unique_layouts[0]\n",
    "    \n",
    "    #check that the submission does not have samples with different 16s regions\n",
    "    unique_16s = manifest['Region_16S'].unique().tolist()\n",
    "    if len(unique_layouts) > 1:\n",
    "        raise ValueError(\"Multiple regions detected. Submission may contain reads from different 16s regions.\")\n",
    "    region_16s = unique_16s[0]\n",
    "    \n",
    "    #add file paths to manifest\n",
    "    if p_s == 'PAIRED':\n",
    "        manifest['forward-absolute-filepath'] = 0\n",
    "        manifest['reverse-absolute-filepath'] = 0\n",
    "        for sample in manifest['id'].tolist():\n",
    "            sample_path = os.path.join(sub_path, sample)\n",
    "            reads = os.listdir(sample_path)\n",
    "            #check for correct number of reads\n",
    "            if len(reads) != 2:\n",
    "                raise ValueError(f\"Incorrect number of files for {sample}. Should be 2. Got {len(reads)}.\")\n",
    "            #add read paths to manifest\n",
    "            manifest.loc[manifest['id'] == sample, 'forward-absolute-filepath'] = os.path.join(sample_path, reads[0])\n",
    "            manifest.loc[manifest['id'] == sample, 'reverse-absolute-filepath'] = os.path.join(sample_path, reads[1])\n",
    "    elif p_s == 'SINGLE':\n",
    "        manifest['absolute-filepath'] = 0\n",
    "        for sample in manifest['id'].tolist():\n",
    "            sample_path = os.path.join(sub_path, sample)\n",
    "            reads = os.listdir(sample_path)\n",
    "            #check for correct number of reads\n",
    "            if len(reads) != 1:\n",
    "                raise ValueError(f\"incorrect number of files for {sample}. Should be 1. Got {len(reads)}.\")\n",
    "            #add read path to manifest\n",
    "            manifest.loc[manifest['id'] == sample, 'absolute-filepath'] = os.path.join(sample_path, reads[0])\n",
    "    else:\n",
    "        raise ValueError(\"invalid value for library layout\")\n",
    "        \n",
    "    #save the manifest\n",
    "    manifest.to_csv(manifest_path, sep='\\t', index=False)\n",
    "\n",
    "    ## execute qiime2 pipeline ##\n",
    "    \n",
    "    print(f\"Processing {sub_num}\")\n",
    "       \n",
    "    # Execute import command\n",
    "    if p_s == 'PAIRED':\n",
    "        demux = qiime2.Artifact.import_data(\n",
    "            'SampleData[PairedEndSequencesWithQuality]', \n",
    "            manifest_path, \n",
    "            view_type='PairedEndFastqManifestPhred33V2'\n",
    "        )\n",
    "    elif p_s == 'SINGLE':\n",
    "        demux = qiime2.Artifact.import_data(\n",
    "            'SampleData[SequencesWithQuality]', \n",
    "            manifest_path, \n",
    "            view_type='SingleEndFastqManifestPhred33V2'\n",
    "        )\n",
    "\n",
    "    # Execute denoise command\n",
    "    denoising_stats_output = os.path.join(summary_o_path, f\"{sub_num}-denoise-stats.qza\") \n",
    "\n",
    "    if p_s == 'PAIRED':\n",
    "        denoise_results = denoise_paired(\n",
    "            demultiplexed_seqs=demux,\n",
    "            trim_left_f=trim_left_f,\n",
    "            trim_left_r=trim_left_r,\n",
    "            trunc_len_f=trunc_len_f,\n",
    "            trunc_len_r=trunc_len_r, \n",
    "            n_threads=n_threads\n",
    "        )\n",
    "    elif p_s == 'SINGLE':\n",
    "        denoise_results = denoise_single(\n",
    "            demultiplexed_seqs=demux,\n",
    "            trim_left=trim_left_f,\n",
    "            trunc_len=trunc_len_f, \n",
    "            n_threads=n_threads\n",
    "        )\n",
    "\n",
    "    denoise_results.denoising_stats.save(denoising_stats_output)\n",
    "    \n",
    "    #temporay? Keep for longer?\n",
    "    denoise_results.representative_sequences.save(os.path.join(summary_o_path, f\"{sub_num}-rep_seqs\"))\n",
    "    denoise_results.table.save(os.path.join(summary_o_path, f\"{sub_num}-freq_table\"))\n",
    "    denoise_results.denoising_stats.save(denoising_stats_output)\n",
    "    print(f\"{sub_num} denoising complete\")\n",
    "\n",
    "    # Execute summary command\n",
    "    visualization_output = os.path.join(summary_o_path, f\"{sub_num}-summary\")\n",
    "    metadata = qiime2.Metadata.load(manifest_path)\n",
    "\n",
    "    summary_result = summarize(\n",
    "        table=denoise_results.table,\n",
    "        sample_metadata=metadata\n",
    "    )\n",
    "    summary_result.visualization.save(visualization_output)\n",
    "    print(f\"{sub_num} summary table complete\")\n",
    "\n",
    "    # Execute classify command\n",
    "    classifier = qiime2.Artifact.load(classifier_map[region_16s])\n",
    "    taxonomy_output = os.path.join(summary_o_path, f\"{sub_num}-taxonomy\")\n",
    "\n",
    "    taxonomy = classify_sklearn(\n",
    "        reads=denoise_results.representative_sequences,\n",
    "        classifier=classifier, \n",
    "        n_jobs=-1,\n",
    "    )\n",
    "    taxonomy.classification.save(taxonomy_output)\n",
    "    \n",
    "    print(f\"{sub_num} taxonomy classification complete\")\n",
    "    \n",
    "    \"\"\"from qiime2 import Artifact\n",
    "    current_dir = os.getcwd()\n",
    "    profile_o_path = os.path.join(current_dir, \"profile_o\")\n",
    "    taxonomy_path = os.path.join(profile_o_path, f\"{sub_num}-taxonomy.qza\")\n",
    "    temp_path = os.path.join(current_dir, \"temp\")\n",
    "    denoise_path = os.path.join(temp_path, \"denoise.qza\")\n",
    "    manifest=os.path.join(temp_path, \"manifest.tsv\")\n",
    "    taxonomy = Artifact.load(taxonomy_path)\n",
    "    denoise_results = Artifact.load(denoise_path)\"\"\"\n",
    "    \n",
    "    #get assignments from BLAST\n",
    "    \n",
    "    \n",
    "    #Generate Relitive Frequencies\n",
    "    rel_freq = relative_frequency(denoise_results.table)[0] #add .table for denoise results\n",
    "    print(f\"{sub_num} relitive frequency table complete\")\n",
    "    \n",
    "    #Generate Taxonomy Bar Graph\n",
    "    barplot_path = os.path.join(summary_o_path, f\"{sub_num}-barplot\") \n",
    "    metadata = Metadata.load(manifest_path)\n",
    "    \n",
    "    barplot_vis = barplot(\n",
    "        table=denoise_results.table, \n",
    "        taxonomy=taxonomy.classification, \n",
    "        metadata=metadata\n",
    "    )\n",
    "    barplot_vis.visualization.save(barplot_path)\n",
    "    print(f\"{sub_num} barplot complete\")\n",
    "    \n",
    "    #generate final output data\n",
    "    rel_freq_df = rel_freq.view(pd.DataFrame)\n",
    "    rel_freq_df = rel_freq_df.transpose()\n",
    "    taxonomy_df = taxonomy.classification.view(pd.DataFrame)\n",
    "    \n",
    "    #display_table(rel_freq_df)\n",
    "    #display_table(taxonomy_df)\n",
    "    \n",
    "    merged_df = pd.merge(rel_freq_df, taxonomy_df, left_index=True, right_index=True, how='inner')\n",
    "    merged_df = merged_df.drop(columns=['Confidence']) #used to also drop feature ID, will see if it works\n",
    "    #sum accross multiple OTUs that are the same species\n",
    "    merged_df = merged_df.groupby('Taxon').sum()\n",
    "    \n",
    "    #DEBUG: \n",
    "    #display_table(merged_df)\n",
    "    \n",
    "    \n",
    "    #print(\"\\n========== fel_freq_df ==========\\n\")\n",
    "    #print(set(rel_freq_df.index))\n",
    "    #print(\"\\n========== taxon_df ==========\\n\")\n",
    "    #print(set(taxonomy_df['Feature ID']))\n",
    "    #print(\"\\n========== join ==========\\n\")\n",
    "    #print(set(rel_freq_df.index) & set(taxonomy_df['Feature ID']))\n",
    "    \n",
    "    #save final output data\n",
    "    profile_output = os.path.join(profile_o_path, f\"{sub_num}-profile.csv\")\n",
    "    merged_df.to_csv(profile_output, index=True)\n",
    "    print(f\"==================== {sub_num} processing complete ====================\")\n",
    "\n",
    "\n",
    "#TODO: double check sequence assignments with blast\n",
    "def display_table(table):\n",
    "    display(table.head())\n",
    "    num_rows, num_cols = table.shape\n",
    "    print(\"Number of rows:\", num_rows)\n",
    "    print(\"Number of columns:\", num_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run Pipeline\n",
    "#generate file paths\n",
    "current_dir = os.getcwd()\n",
    "classifier_folder = os.path.join(current_dir, \"classifier\")\n",
    "master_manifest_path = os.path.join(current_dir, \"DataFrame_2_Pruned.tsv\")\n",
    "profile_o_path = os.path.join(current_dir, \"profile_o\")\n",
    "summary_o_path = os.path.join(current_dir, \"summary_o\")\n",
    "gg_nb = os.path.join(classifier_folder, \"gg-13-8-99-nb-classifier.qza\") \n",
    "\n",
    "#only run the pipeline on unprocessed reads\n",
    "sub_nums = set(os.listdir(os.path.join(current_dir, \"sorted_reads\")))\n",
    "completed = set(os.listdir(profile_o_path))\n",
    "completed = [x[:-len(\"-profile.csv\")] for x in completed]\n",
    "sub_nums.difference_update(completed)\n",
    "\n",
    "\n",
    "classifier_map = {\n",
    "    \"V4\":gg_nb, \n",
    "    \"V1-V3\":gg_nb, \n",
    "    \"V1-V2\":gg_nb, \n",
    "    \"V3-V4\":gg_nb, \n",
    "    \"V1-V2 and v4\":gg_nb, \n",
    "    \"V4-V5\":gg_nb, \n",
    "    \"V1-V4\":gg_nb\n",
    "}\n",
    "\n",
    "#BUG: wrong set being used to join taxon df and freq df\n",
    "\n",
    "print(f\"Processing submissions {sub_nums}\\n\")\n",
    "for sub_num in sub_nums:\n",
    "    sub_path = os.path.join(current_dir, \"sorted_reads\", sub_num)\n",
    "    try:\n",
    "        qiime2_profile(sub_num, sub_path, master_manifest_path, profile_o_path, summary_o_path, classifier_map, trim_left_f=0, trim_left_r=0, trunc_len_f=0, trunc_len_r=0, )\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing {sub_num}: {e}\")\n",
    "        print(f\"==================== {sub_num} processing failed ====================\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2",
   "language": "python",
   "name": "qiime2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
