{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "#make files if don't already exist\n",
    "os.makedirs(\"profile_o\", exist_ok=True)\n",
    "os.makedirs(\"sorted_reads\", exist_ok=True)\n",
    "os.makedirs(\"seq_reads\", exist_ok=True)\n",
    "os.makedirs(\"summary_o\", exist_ok=True)\n",
    "os.makedirs(\"classifier\", exist_ok=True)\n",
    "os.makedirs(\"classifier_training_data\", exist_ok=True)\n",
    "\n",
    "#downstream analysis: https://www.youtube.com/watch?v=9AmM-BHYxJo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (1.79)\r\n",
      "Requirement already satisfied: numpy in /home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages (from biopython) (1.19.1)\r\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## SORT RAW SKINOME READS ##\n",
    "\n",
    "import os\n",
    "import shutil\n",
    "import pandas as pd\n",
    "import glob\n",
    "\n",
    "# Define paths\n",
    "current_dir = os.getcwd()\n",
    "sorted_reads = os.path.join(current_dir, 'sorted_reads')\n",
    "seq_reads = os.path.join(current_dir, 'seq_reads')\n",
    "master_manifest = os.path.join(current_dir, 'DataFrame_2_Pruned.tsv')\n",
    "\n",
    "# load in df\n",
    "df = pd.read_csv(master_manifest, sep='\\t', encoding='latin1')\n",
    "df = df.dropna(subset=['Submission'])\n",
    "submissions = df['Submission'].unique()\n",
    "submissions = submissions.tolist()\n",
    "\n",
    "#create seprate file for each submission\n",
    "for sub in submissions: \n",
    "    new_direc = os.path.join(sorted_reads, sub)\n",
    "    os.makedirs(new_direc, exist_ok=True)\n",
    "    pruned_df = df[df['Submission'] == sub]\n",
    "    p_s = pruned_df['Library_Layout'].unique().tolist()[0]\n",
    "    if p_s == 'PAIRED':\n",
    "        for run in pruned_df['Run'].tolist():\n",
    "            sample_direc = os.path.join(new_direc, run)\n",
    "            sample_1_path = os.path.join(seq_reads, run + '_1.fastq.gz')\n",
    "            sample_2_path = os.path.join(seq_reads, run + '_2.fastq.gz')\n",
    "            \n",
    "            if (not os.path.exists(sample_1_path)) or  (not os.path.exists(sample_2_path)):\n",
    "                \n",
    "                if not os.path.exists(sample_1_path):\n",
    "                    print('file path ' + sample_1_path + ' was not found for paired sample ' + run)\n",
    "                    \n",
    "                if not os.path.exists(sample_1_path):\n",
    "                    print('file path ' + sample_1_path + ' was not found for paired sample ' + run)\n",
    "                \n",
    "                continue\n",
    "                \n",
    "            os.makedirs(sample_direc, exist_ok=True)\n",
    "            shutil.move(sample_1_path, sample_direc)\n",
    "            shutil.move(sample_2_path, sample_direc)\n",
    "    elif p_s == 'SINGLE':\n",
    "        for run in pruned_df['Run'].tolist():\n",
    "            sample_direc = os.path.join(new_direc, run)\n",
    "            sample_1_path = os.path.join(seq_reads, run + '_1.fastq.gz')\n",
    "            \n",
    "            if not os.path.exists(sample_1_path):\n",
    "                print('file path ' + sample_1_path + ' was not found for single sample ' + run)\n",
    "                continue\n",
    "            \n",
    "            os.makedirs(sample_direc, exist_ok=True)\n",
    "            shutil.move(sample_1_path, sample_direc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Run Qiime2 ##\n",
    "\n",
    "import os\n",
    "import shutil\n",
    "import pandas as pd\n",
    "import glob\n",
    "import subprocess\n",
    "import re\n",
    "import qiime2\n",
    "from qiime2.plugins.dada2.methods import denoise_single, denoise_paired\n",
    "from qiime2.plugins.feature_table.visualizers import summarize\n",
    "from qiime2.plugins.feature_classifier.methods import classify_sklearn\n",
    "from qiime2.plugins.taxa.visualizers import barplot\n",
    "from Bio import SeqIO\n",
    "from qiime2.plugins.feature_table.methods import relative_frequency\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from qiime2 import Metadata\n",
    "\n",
    "\n",
    "def qiime2_profile(sub_num, sub_path, master_manifest_path, profile_o_path, summary_o_folder, classifier_map, trim_left_f=13, trim_left_r=13, trunc_len_f=150, trunc_len_r=150, n_threads=8):\n",
    "    \"\"\"\n",
    "    calculates microbiome profiles from a submission of reads and stores the results in a .qza file in the specified folder\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    sub_num : str\n",
    "        ID of the submission containing the reads.\n",
    "    sub_path : str\n",
    "        path to the submission.\n",
    "    master_manifest_path : str\n",
    "        path to the master meta data file (should .tsv file) Manifest must \n",
    "        have the following columns: Run or id, Library_Layout, Region_16S, \n",
    "        Submission\n",
    "    profile_o_path : str\n",
    "        path to the output folder for microbiome profile\n",
    "    summary_o_path : str:\n",
    "        path to the output folder for read summary table\n",
    "    classifier_map : dict\n",
    "        map of 16s rRNA region to classifier built to classify that region\n",
    "    trim_left_f : int\n",
    "        number of nucleotides to trim from the 5' end of the forward read before denoising\n",
    "    trim_left_r : int\n",
    "        number of nucleotides to trim from the 5' end of the reverse read before denoising\n",
    "    trim_left_f : int\n",
    "        specifies the length at which the forward read will be truncated\n",
    "    trim_left_r : int\n",
    "        specifies the length at which the reverse read will be truncated\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    none\n",
    "    \"\"\"\n",
    "    \n",
    "    #define variables\n",
    "    current_dir = os.getcwd()\n",
    "    summary_o_path = os.path.join(summary_o_folder, sub_num)\n",
    "    manifest_path = os.path.join(summary_o_path, f\"{sub_num}-manifest.tsv\")\n",
    "    \n",
    "    #string representing weather reads are paired or single\n",
    "    p_s = None           \n",
    "    #string representing what 16s regions were used\n",
    "    region_16s = None           \n",
    "\n",
    "    #validate inputs\n",
    "    if not type(sub_num) == str:\n",
    "        raise TypeEror(\"sub_num must be a string\")\n",
    "    if not os.path.exists(sub_path):\n",
    "        raise FileNotFoundError(f\"{sub_num} submission path is invalid\")\n",
    "    if not os.path.exists(master_manifest_path):\n",
    "        raise FileNotFoundError(\"master manifest path is invalid\")\n",
    "    if not type(classifier_map) == dict:\n",
    "        raise TypeError(\"classifier_map must be a dict\")\n",
    "    for classifier in classifier_map.values():\n",
    "        if not os.path.exists(classifier):\n",
    "            raise FileNotFoundError(f\"classifier map contains invalid paths. {classifier} is not a valid path\")\n",
    "    \n",
    "    #if output folders do not exist, create them\n",
    "    os.makedirs(profile_o_path, exist_ok=True)\n",
    "    os.makedirs(summary_o_folder, exist_ok=True)\n",
    "    os.makedirs(summary_o_path, exist_ok=True)\n",
    "    \n",
    "    #generate manifest\n",
    "    manifest = pd.read_csv(master_manifest_path, sep='\\t', encoding='latin1')\n",
    "    \n",
    "    #Rename \"Run\" to 'id' to support compatability with qiime2\n",
    "    if 'Run' in manifest.columns:\n",
    "        manifest.rename(columns={'Run': 'id'}, inplace=True)\n",
    "    \n",
    "    #prune manifest\n",
    "    manifest = manifest[manifest['Submission'] == sub_num]\n",
    "    reads = os.listdir(sub_path)\n",
    "    len_manifest = len(manifest)\n",
    "    len_reads = len(reads)\n",
    "    manifest = manifest[manifest['id'].isin(reads)]\n",
    "    len_filtered_manifest = len(manifest)\n",
    "    \n",
    "    #check that names and number of samples match on manifest and in folder\n",
    "    if manifest.empty:\n",
    "        raise ValueError(\"Pruned manifest is empty. Submission number may not be in manifest or reads may be improperly named.\")\n",
    "    if len_manifest > len_reads:\n",
    "        print(\"Warning: more entries in manifest than in folder\")\n",
    "    if len_manifest < len_reads:\n",
    "        print(\"Warning: more entries in folder than in manifest\")\n",
    "    if len_manifest == len_reads and len_filtered_manifest < len_reads:\n",
    "        print(\"Warning: mismatch between some names on manifest and in folder\")\n",
    "        \n",
    "    #check that the submission does not have mixed single and paired reads\n",
    "    unique_layouts = manifest['Library_Layout'].unique().tolist()\n",
    "    \n",
    "    if len(unique_layouts) > 1:\n",
    "        raise ValueError(\"Multiple layouts detected. Submission may contain single and paired reads.\")\n",
    "    p_s = unique_layouts[0]\n",
    "    \n",
    "    #check that the submission does not have samples with different 16s regions\n",
    "    unique_16s = manifest['Region_16S'].unique().tolist()\n",
    "    if len(unique_layouts) > 1:\n",
    "        raise ValueError(\"Multiple regions detected. Submission may contain reads from different 16s regions.\")\n",
    "    region_16s = unique_16s[0]\n",
    "    \n",
    "    #add file paths to manifest\n",
    "    if p_s == 'PAIRED':\n",
    "        manifest['forward-absolute-filepath'] = 0\n",
    "        manifest['reverse-absolute-filepath'] = 0\n",
    "        for sample in manifest['id'].tolist():\n",
    "            sample_path = os.path.join(sub_path, sample)\n",
    "            reads = os.listdir(sample_path)\n",
    "            #check for correct number of reads\n",
    "            if len(reads) != 2:\n",
    "                raise ValueError(f\"Incorrect number of files for {sample}. Should be 2. Got {len(reads)}.\")\n",
    "            #add read paths to manifest\n",
    "            manifest.loc[manifest['id'] == sample, 'forward-absolute-filepath'] = os.path.join(sample_path, reads[0])\n",
    "            manifest.loc[manifest['id'] == sample, 'reverse-absolute-filepath'] = os.path.join(sample_path, reads[1])\n",
    "    elif p_s == 'SINGLE':\n",
    "        manifest['absolute-filepath'] = 0\n",
    "        for sample in manifest['id'].tolist():\n",
    "            sample_path = os.path.join(sub_path, sample)\n",
    "            reads = os.listdir(sample_path)\n",
    "            #check for correct number of reads\n",
    "            if len(reads) != 1:\n",
    "                raise ValueError(f\"incorrect number of files for {sample}. Should be 1. Got {len(reads)}.\")\n",
    "            #add read path to manifest\n",
    "            manifest.loc[manifest['id'] == sample, 'absolute-filepath'] = os.path.join(sample_path, reads[0])\n",
    "    else:\n",
    "        raise ValueError(\"invalid value for library layout\")\n",
    "        \n",
    "    #save the manifest\n",
    "    manifest.to_csv(manifest_path, sep='\\t', index=False)\n",
    "\n",
    "    ## execute qiime2 pipeline ##\n",
    "    \n",
    "    print(f\"Processing {sub_num}\")\n",
    "       \n",
    "    # Execute import command\n",
    "    if p_s == 'PAIRED':\n",
    "        demux = qiime2.Artifact.import_data(\n",
    "            'SampleData[PairedEndSequencesWithQuality]', \n",
    "            manifest_path, \n",
    "            view_type='PairedEndFastqManifestPhred33V2'\n",
    "        )\n",
    "    elif p_s == 'SINGLE':\n",
    "        demux = qiime2.Artifact.import_data(\n",
    "            'SampleData[SequencesWithQuality]', \n",
    "            manifest_path, \n",
    "            view_type='SingleEndFastqManifestPhred33V2'\n",
    "        )\n",
    "\n",
    "    # Execute denoise command\n",
    "    denoising_stats_output = os.path.join(summary_o_path, f\"{sub_num}-denoise-stats.qza\") \n",
    "\n",
    "    if p_s == 'PAIRED':\n",
    "        denoise_results = denoise_paired(\n",
    "            demultiplexed_seqs=demux,\n",
    "            trim_left_f=trim_left_f,\n",
    "            trim_left_r=trim_left_r,\n",
    "            trunc_len_f=trunc_len_f,\n",
    "            trunc_len_r=trunc_len_r, \n",
    "            n_threads=n_threads\n",
    "        )\n",
    "    elif p_s == 'SINGLE':\n",
    "        denoise_results = denoise_single(\n",
    "            demultiplexed_seqs=demux,\n",
    "            trim_left=trim_left_f,\n",
    "            trunc_len=trunc_len_f, \n",
    "            n_threads=n_threads\n",
    "        )\n",
    "\n",
    "    denoise_results.denoising_stats.save(denoising_stats_output)\n",
    "    \n",
    "    #temporay? Keep for longer?\n",
    "    denoise_results.representative_sequences.save(os.path.join(summary_o_path, f\"{sub_num}-rep_seqs\"))\n",
    "    denoise_results.table.save(os.path.join(summary_o_path, f\"{sub_num}-freq_table\"))\n",
    "    denoise_results.denoising_stats.save(denoising_stats_output)\n",
    "    print(f\"{sub_num} denoising complete\")\n",
    "\n",
    "    # Execute summary command\n",
    "    visualization_output = os.path.join(summary_o_path, f\"{sub_num}-summary\")\n",
    "    metadata = qiime2.Metadata.load(manifest_path)\n",
    "\n",
    "    summary_result = summarize(\n",
    "        table=denoise_results.table,\n",
    "        sample_metadata=metadata\n",
    "    )\n",
    "    summary_result.visualization.save(visualization_output)\n",
    "    print(f\"{sub_num} summary table complete\")\n",
    "\n",
    "    # Execute classify command\n",
    "    classifier = qiime2.Artifact.load(classifier_map[region_16s])\n",
    "    taxonomy_output = os.path.join(summary_o_path, f\"{sub_num}-taxonomy\")\n",
    "\n",
    "    taxonomy = classify_sklearn(\n",
    "        reads=denoise_results.representative_sequences,\n",
    "        classifier=classifier, \n",
    "        n_jobs=-1,\n",
    "    )\n",
    "    taxonomy.classification.save(taxonomy_output)\n",
    "    \n",
    "    print(f\"{sub_num} taxonomy classification complete\")   \n",
    "    \n",
    "    #Generate Relitive Frequencies\n",
    "    rel_freq = relative_frequency(denoise_results.table)[0] #add .table for denoise results\n",
    "    print(f\"{sub_num} relitive frequency table complete\")\n",
    "    \n",
    "    #Generate Taxonomy Bar Graph\n",
    "    barplot_path = os.path.join(summary_o_path, f\"{sub_num}-barplot\") \n",
    "    metadata = Metadata.load(manifest_path)\n",
    "    \n",
    "    barplot_vis = barplot(\n",
    "        table=denoise_results.table, \n",
    "        taxonomy=taxonomy.classification, \n",
    "        metadata=metadata\n",
    "    )\n",
    "    barplot_vis.visualization.save(barplot_path)\n",
    "    print(f\"{sub_num} barplot complete\")\n",
    "    \n",
    "    #generate final output data\n",
    "    rel_freq_df = rel_freq.view(pd.DataFrame)\n",
    "    rel_freq_df = rel_freq_df.transpose()\n",
    "    taxonomy_df = taxonomy.classification.view(pd.DataFrame)\n",
    "    \n",
    "    merged_df = pd.merge(rel_freq_df, taxonomy_df, left_index=True, right_index=True, how='inner')\n",
    "    merged_df = merged_df.drop(columns=['Confidence']) #used to also drop feature ID, will see if it works\n",
    "    #sum accross multiple OTUs that are the same species\n",
    "    merged_df = merged_df.groupby('Taxon').sum()\n",
    "    \n",
    "    #save final output data\n",
    "    profile_output = os.path.join(profile_o_path, f\"{sub_num}-profile.csv\")\n",
    "    merged_df.to_csv(profile_output, index=True)\n",
    "    print(f\"==================== {sub_num} processing complete ====================\")\n",
    "\n",
    "\n",
    "#TODO: double check sequence assignments with blast\n",
    "def display_table(table):\n",
    "    display(table.head())\n",
    "    num_rows, num_cols = table.shape\n",
    "    print(\"Number of rows:\", num_rows)\n",
    "    print(\"Number of columns:\", num_cols)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing submissions {'ERA1111396', 'ERA292152', 'SRA578693', 'SRA1038019', 'SRA298410', 'SRA860003', 'SRA575681', 'SRA996410'}\n",
      "\n",
      "Warning: more entries in manifest than in folder\n",
      "Processing ERA1111396\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-fz0pn7ls/2d82f39b-0744-4caa-8965-e42a868b27f3/data /tmp/tmp6_w64ewo/output.tsv.biom /tmp/tmp6_w64ewo/track.tsv /tmp/tmp6_w64ewo 0 0 2.0 2 Inf independent consensus 1.0 8 1000000 NULL 16\n",
      "\n",
      "ERA1111396 denoising complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:122: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'sample-frequency-detail.csv'))\n",
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:124: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'feature-frequency-detail.csv'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERA1111396 summary table complete\n",
      "ERA1111396 taxonomy classification complete\n",
      "ERA1111396 relitive frequency table complete\n",
      "ERA1111396 barplot complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ERR2162452</th>\n",
       "      <th>ERR2162453</th>\n",
       "      <th>ERR2162454</th>\n",
       "      <th>ERR2162455</th>\n",
       "      <th>ERR2162456</th>\n",
       "      <th>ERR2162457</th>\n",
       "      <th>ERR2162458</th>\n",
       "      <th>ERR2162459</th>\n",
       "      <th>ERR2162460</th>\n",
       "      <th>ERR2162461</th>\n",
       "      <th>...</th>\n",
       "      <th>ERR2163209</th>\n",
       "      <th>ERR2163210</th>\n",
       "      <th>ERR2163211</th>\n",
       "      <th>ERR2163212</th>\n",
       "      <th>ERR2163213</th>\n",
       "      <th>ERR2163214</th>\n",
       "      <th>ERR2163215</th>\n",
       "      <th>ERR2163216</th>\n",
       "      <th>ERR2163217</th>\n",
       "      <th>ERR2163218</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Unassigned</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.052964</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.001969</td>\n",
       "      <td>0.000192</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.029153</td>\n",
       "      <td>0.000133</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Archaea</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Archaea; p__Crenarchaeota; c__Thaumarchaeota; o__Cenarchaeales; f__Cenarchaeaceae; g__Nitrosopumilus; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Archaea; p__Crenarchaeota; c__Thaumarchaeota; o__Nitrososphaerales; f__Nitrososphaeraceae</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Archaea; p__Crenarchaeota; c__Thaumarchaeota; o__Nitrososphaerales; f__Nitrososphaeraceae; g__Candidatus Nitrososphaera; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 761 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    ERR2162452  ERR2162453  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                                 0.0    0.000261   \n",
       "k__Archaea                                                 0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "\n",
       "                                                    ERR2162454  ERR2162455  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                            0.052964         0.0   \n",
       "k__Archaea                                            0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "\n",
       "                                                    ERR2162456  ERR2162457  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                                 0.0         0.0   \n",
       "k__Archaea                                                 0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "\n",
       "                                                    ERR2162458  ERR2162459  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                                 0.0    0.001969   \n",
       "k__Archaea                                                 0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "\n",
       "                                                    ERR2162460  ERR2162461  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                            0.000192         0.0   \n",
       "k__Archaea                                            0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "\n",
       "                                                    ...  ERR2163209  \\\n",
       "Taxon                                               ...               \n",
       "Unassigned                                          ...         0.0   \n",
       "k__Archaea                                          ...         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...  ...         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...  ...         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...  ...         0.0   \n",
       "\n",
       "                                                    ERR2163210  ERR2163211  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                                 0.0    0.029153   \n",
       "k__Archaea                                                 0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0    0.000000   \n",
       "\n",
       "                                                    ERR2163212  ERR2163213  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                            0.000133    0.000081   \n",
       "k__Archaea                                            0.000000    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000    0.000000   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000    0.000000   \n",
       "\n",
       "                                                    ERR2163214  ERR2163215  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                            0.000034         0.0   \n",
       "k__Archaea                                            0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...    0.000000         0.0   \n",
       "\n",
       "                                                    ERR2163216  ERR2163217  \\\n",
       "Taxon                                                                        \n",
       "Unassigned                                                 0.0         0.0   \n",
       "k__Archaea                                                 0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0         0.0   \n",
       "\n",
       "                                                    ERR2163218  \n",
       "Taxon                                                           \n",
       "Unassigned                                                 0.0  \n",
       "k__Archaea                                                 0.0  \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0  \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0  \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...         0.0  \n",
       "\n",
       "[5 rows x 761 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 2181\n",
      "Number of columns: 761\n",
      "==================== ERA1111396 processing complete ====================\n",
      "Warning: more entries in manifest than in folder\n",
      "Processing ERA292152\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-3ffhmlxu/b1dea680-b371-4f96-9a85-3c2f07afbf8f/data /tmp/tmpcgfdg1sd/output.tsv.biom /tmp/tmpcgfdg1sd/track.tsv /tmp/tmpcgfdg1sd 0 0 2.0 2 Inf independent consensus 1.0 8 1000000 NULL 16\n",
      "\n",
      "Error processing ERA292152: An error was encountered while running DADA2 in R (return code 1), please inspect stdout and stderr to learn more.\n",
      "==================== ERA292152 processing failed ====================\n",
      "Error processing SRA578693: Pruned manifest is empty. Submission number may not be in manifest or reads may be improperly named.\n",
      "==================== SRA578693 processing failed ====================\n",
      "Processing SRA1038019\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_paired.R /tmp/tmpjbvblg_6/forward /tmp/tmpjbvblg_6/reverse /tmp/tmpjbvblg_6/output.tsv.biom /tmp/tmpjbvblg_6/track.tsv /tmp/tmpjbvblg_6/filt_f /tmp/tmpjbvblg_6/filt_r 0 0 0 0 2.0 2.0 2 independent consensus 1.0 8 1000000\n",
      "\n",
      "SRA1038019 denoising complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:122: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'sample-frequency-detail.csv'))\n",
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:124: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'feature-frequency-detail.csv'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRA1038019 summary table complete\n",
      "SRA1038019 taxonomy classification complete\n",
      "SRA1038019 relitive frequency table complete\n",
      "SRA1038019 barplot complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRR11027332</th>\n",
       "      <th>SRR11027333</th>\n",
       "      <th>SRR11027334</th>\n",
       "      <th>SRR11027335</th>\n",
       "      <th>SRR11027336</th>\n",
       "      <th>SRR11027337</th>\n",
       "      <th>SRR11027338</th>\n",
       "      <th>SRR11027339</th>\n",
       "      <th>SRR11027340</th>\n",
       "      <th>SRR11027341</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR11027344</th>\n",
       "      <th>SRR11027345</th>\n",
       "      <th>SRR11027346</th>\n",
       "      <th>SRR11027347</th>\n",
       "      <th>SRR11027348</th>\n",
       "      <th>SRR11027349</th>\n",
       "      <th>SRR11027350</th>\n",
       "      <th>SRR11027351</th>\n",
       "      <th>SRR11027352</th>\n",
       "      <th>SRR11027353</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>k__Bacteria</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.901408</td>\n",
       "      <td>0.34127</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.111842</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales; f__Porphyromonadaceae; g__Porphyromonas; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.412214</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Firmicutes; c__Bacilli; o__Bacillales; f__Staphylococcaceae; g__Staphylococcus</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.916667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__Clostridiaceae; g__Clostridium; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.156489</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.425</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Firmicutes; c__Clostridia; o__Clostridiales; f__[Tissierellaceae]; g__Anaerococcus; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.526316</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.148855</td>\n",
       "      <td>...</td>\n",
       "      <td>0.888158</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    SRR11027332  SRR11027333  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "\n",
       "                                                    SRR11027334  SRR11027335  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                            0.000000          0.0   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...     0.333333          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...     0.000000          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...     0.000000          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...     0.500000          0.0   \n",
       "\n",
       "                                                    SRR11027336  SRR11027337  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                            0.901408      0.34127   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...     0.000000      0.00000   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...     0.000000      0.00000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...     0.000000      0.00000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...     0.000000      0.00000   \n",
       "\n",
       "                                                    SRR11027338  SRR11027339  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0     0.000000   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.526316   \n",
       "\n",
       "                                                    SRR11027340  SRR11027341  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0     0.000000   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0     0.412214   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.156489   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.148855   \n",
       "\n",
       "                                                    ...  SRR11027344  \\\n",
       "Taxon                                               ...                \n",
       "k__Bacteria                                         ...     0.111842   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...  ...     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...  ...     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...  ...     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...  ...     0.888158   \n",
       "\n",
       "                                                    SRR11027345  SRR11027346  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0        0.000   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0        0.000   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0        0.000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0        0.425   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0        0.000   \n",
       "\n",
       "                                                    SRR11027347  SRR11027348  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0     0.083333   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0     0.916667   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.000000   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0     0.000000   \n",
       "\n",
       "                                                    SRR11027349  SRR11027350  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          1.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "\n",
       "                                                    SRR11027351  SRR11027352  \\\n",
       "Taxon                                                                          \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0          0.0   \n",
       "\n",
       "                                                    SRR11027353  \n",
       "Taxon                                                            \n",
       "k__Bacteria                                                 0.0  \n",
       "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; ...          0.0  \n",
       "k__Bacteria; p__Firmicutes; c__Bacilli; o__Baci...          0.0  \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0  \n",
       "k__Bacteria; p__Firmicutes; c__Clostridia; o__C...          0.0  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 15\n",
      "Number of columns: 22\n",
      "==================== SRA1038019 processing complete ====================\n",
      "Warning: more entries in manifest than in folder\n",
      "Processing SRA298410\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_paired.R /tmp/tmpscm0387u/forward /tmp/tmpscm0387u/reverse /tmp/tmpscm0387u/output.tsv.biom /tmp/tmpscm0387u/track.tsv /tmp/tmpscm0387u/filt_f /tmp/tmpscm0387u/filt_r 0 0 0 0 2.0 2.0 2 independent consensus 1.0 8 1000000\n",
      "\n",
      "Error processing SRA298410: An error was encountered while running DADA2 in R (return code -11), please inspect stdout and stderr to learn more.\n",
      "==================== SRA298410 processing failed ====================\n",
      "Processing SRA860003\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_paired.R /tmp/tmprx5v08qh/forward /tmp/tmprx5v08qh/reverse /tmp/tmprx5v08qh/output.tsv.biom /tmp/tmprx5v08qh/track.tsv /tmp/tmprx5v08qh/filt_f /tmp/tmprx5v08qh/filt_r 0 0 0 0 2.0 2.0 2 independent consensus 1.0 8 1000000\n",
      "\n",
      "SRA860003 denoising complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:122: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'sample-frequency-detail.csv'))\n",
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:124: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'feature-frequency-detail.csv'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRA860003 summary table complete\n",
      "SRA860003 taxonomy classification complete\n",
      "SRA860003 relitive frequency table complete\n",
      "SRA860003 barplot complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRR8728265</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Corynebacteriaceae; g__Corynebacterium; s__</th>\n",
       "      <td>0.082248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Dermabacteraceae; g__Brachybacterium</th>\n",
       "      <td>0.020562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Dietziaceae; g__Dietzia; s__</th>\n",
       "      <td>0.002443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Micrococcaceae</th>\n",
       "      <td>0.002850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Actinobacteria; c__Actinobacteria; o__Actinomycetales; f__Micrococcaceae; g__Kocuria; s__palustris</th>\n",
       "      <td>0.036849</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    SRR8728265\n",
       "Taxon                                                         \n",
       "k__Bacteria; p__Actinobacteria; c__Actinobacter...    0.082248\n",
       "k__Bacteria; p__Actinobacteria; c__Actinobacter...    0.020562\n",
       "k__Bacteria; p__Actinobacteria; c__Actinobacter...    0.002443\n",
       "k__Bacteria; p__Actinobacteria; c__Actinobacter...    0.002850\n",
       "k__Bacteria; p__Actinobacteria; c__Actinobacter...    0.036849"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 49\n",
      "Number of columns: 1\n",
      "==================== SRA860003 processing complete ====================\n",
      "Error processing SRA575681: Pruned manifest is empty. Submission number may not be in manifest or reads may be improperly named.\n",
      "==================== SRA575681 processing failed ====================\n",
      "Processing SRA996410\n",
      "Running external command line application(s). This may print messages to stdout and/or stderr.\n",
      "The command(s) being run are below. These commands cannot be manually re-run as they will depend on temporary files that no longer exist.\n",
      "\n",
      "Command: run_dada_single.R /tmp/qiime2-archive-uoy8c3cn/d6fc8df8-2285-46bb-a7e6-d4210da98c38/data /tmp/tmp9eye8jzr/output.tsv.biom /tmp/tmp9eye8jzr/track.tsv /tmp/tmp9eye8jzr 0 0 2.0 2 Inf independent consensus 1.0 8 1000000 NULL 16\n",
      "\n",
      "SRA996410 denoising complete\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:122: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'sample-frequency-detail.csv'))\n",
      "/home/miloscola/anaconda3/envs/qiime2/lib/python3.6/site-packages/q2_feature_table/_summarize/_visualizer.py:124: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  os.path.join(output_dir, 'feature-frequency-detail.csv'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRA996410 summary table complete\n",
      "SRA996410 taxonomy classification complete\n",
      "SRA996410 relitive frequency table complete\n",
      "SRA996410 barplot complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRR10447626</th>\n",
       "      <th>SRR10447627</th>\n",
       "      <th>SRR10447628</th>\n",
       "      <th>SRR10447629</th>\n",
       "      <th>SRR10447630</th>\n",
       "      <th>SRR10447631</th>\n",
       "      <th>SRR10447632</th>\n",
       "      <th>SRR10447633</th>\n",
       "      <th>SRR10447634</th>\n",
       "      <th>SRR10447635</th>\n",
       "      <th>...</th>\n",
       "      <th>SRR10447734</th>\n",
       "      <th>SRR10447735</th>\n",
       "      <th>SRR10447736</th>\n",
       "      <th>SRR10447737</th>\n",
       "      <th>SRR10447738</th>\n",
       "      <th>SRR10447739</th>\n",
       "      <th>SRR10447740</th>\n",
       "      <th>SRR10447741</th>\n",
       "      <th>SRR10447742</th>\n",
       "      <th>SRR10447743</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Taxon</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>k__Archaea; p__Crenarchaeota; c__Thaumarchaeota; o__Nitrososphaerales; f__Nitrososphaeraceae; g__Candidatus Nitrososphaera; s__SCA1170</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006458</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.003238</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__; c__; o__; f__; g__; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Acidobacteria</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>k__Bacteria; p__Acidobacteria; c__Acidobacteria-5; o__; f__; g__; s__</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 118 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    SRR10447626  SRR10447627  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447628  SRR10447629  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447630  SRR10447631  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447632  SRR10447633  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447634  SRR10447635  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    ...  SRR10447734  \\\n",
       "Taxon                                               ...                \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...  ...          0.0   \n",
       "k__Bacteria                                         ...          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__           ...          0.0   \n",
       "k__Bacteria; p__Acidobacteria                       ...          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...  ...          0.0   \n",
       "\n",
       "                                                    SRR10447735  SRR10447736  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0     0.000000   \n",
       "k__Bacteria                                                 0.0     0.006458   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0     0.000000   \n",
       "k__Bacteria; p__Acidobacteria                               0.0     0.000000   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0     0.000000   \n",
       "\n",
       "                                                    SRR10447737  SRR10447738  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0     0.000000   \n",
       "k__Bacteria                                                 0.0     0.003238   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0     0.000000   \n",
       "k__Bacteria; p__Acidobacteria                               0.0     0.000000   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0     0.000000   \n",
       "\n",
       "                                                    SRR10447739  SRR10447740  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447741  SRR10447742  \\\n",
       "Taxon                                                                          \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0          0.0   \n",
       "k__Bacteria                                                 0.0          0.0   \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria                               0.0          0.0   \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0          0.0   \n",
       "\n",
       "                                                    SRR10447743  \n",
       "Taxon                                                            \n",
       "k__Archaea; p__Crenarchaeota; c__Thaumarchaeota...          0.0  \n",
       "k__Bacteria                                                 0.0  \n",
       "k__Bacteria; p__; c__; o__; f__; g__; s__                   0.0  \n",
       "k__Bacteria; p__Acidobacteria                               0.0  \n",
       "k__Bacteria; p__Acidobacteria; c__Acidobacteria...          0.0  \n",
       "\n",
       "[5 rows x 118 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 994\n",
      "Number of columns: 118\n",
      "==================== SRA996410 processing complete ====================\n"
     ]
    }
   ],
   "source": [
    "#Run Pipeline\n",
    "#generate file paths\n",
    "current_dir = os.getcwd()\n",
    "classifier_folder = os.path.join(current_dir, \"classifier\")\n",
    "master_manifest_path = os.path.join(current_dir, \"DataFrame_2_Pruned.tsv\")\n",
    "profile_o_path = os.path.join(current_dir, \"profile_o\")\n",
    "summary_o_path = os.path.join(current_dir, \"summary_o\")\n",
    "\n",
    "#classifiers\n",
    "gg_nb = os.path.join(classifier_folder, \"gg-13-8-99-nb-classifier.qza\") \n",
    "\n",
    "#only run the pipeline on unprocessed reads\n",
    "sub_nums = set(os.listdir(os.path.join(current_dir, \"sorted_reads\")))\n",
    "completed = set(os.listdir(profile_o_path))\n",
    "completed = [x[:-len(\"-profile.csv\")] for x in completed]\n",
    "sub_nums.difference_update(completed)\n",
    "\n",
    "\n",
    "classifier_map = {\n",
    "    \"V4\":gg_nb, \n",
    "    \"V1-V3\":gg_nb, \n",
    "    \"V1-V2\":gg_nb, \n",
    "    \"V3-V4\":gg_nb, \n",
    "    \"V1-V2 and v4\":gg_nb, \n",
    "    \"V4-V5\":gg_nb, \n",
    "    \"V1-V4\":gg_nb\n",
    "}\n",
    "\n",
    "#BUG: wrong set being used to join taxon df and freq df\n",
    "\n",
    "print(f\"Processing submissions {sub_nums}\\n\")\n",
    "for sub_num in sub_nums:\n",
    "    sub_path = os.path.join(current_dir, \"sorted_reads\", sub_num)\n",
    "    try:\n",
    "        qiime2_profile(sub_num, sub_path, master_manifest_path, profile_o_path, summary_o_path, classifier_map, trim_left_f=0, trim_left_r=0, trunc_len_f=0, trunc_len_r=0, )\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing {sub_num}: {e}\")\n",
    "        print(f\"==================== {sub_num} processing failed ====================\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-zun6d_x8')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 36cba8a6-82f7-49a6-b05a-c112ed7a271c>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#visuialize bar plot graph\n",
    "from qiime2 import Visualization\n",
    "from qiime2 import Artifact\n",
    "from qiime2.plugins import demux, dada2, feature_classifier, feature_table\n",
    "from qiime2 import Metadata\n",
    "import os\n",
    "from qiime2.plugins.taxa.visualizers import barplot\n",
    "\n",
    "sub_num = \"ERA1522711\"\n",
    "current_dir = os.getcwd()\n",
    "summary_path = os.path.join(current_dir, \"summary_o\", sub_num, f\"{sub_num}-summary.qzv\")\n",
    "\n",
    "# Load the summary visualization\n",
    "summary_viz = Visualization.load(summary_path)\n",
    "\n",
    "# Visualize the barplot\n",
    "summary_viz\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-yeoxx3u7')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 6f280066-cf4c-45ff-bd4b-2c34a4859684>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#visuialize bar plot graph\n",
    "from qiime2 import Visualization\n",
    "from qiime2 import Artifact\n",
    "from qiime2.plugins import demux, dada2, feature_classifier, feature_table\n",
    "from qiime2 import Metadata\n",
    "import os\n",
    "from qiime2.plugins.taxa.visualizers import barplot\n",
    "\n",
    "sub_num = \"ERA1522711\"\n",
    "current_dir = os.getcwd()\n",
    "barplot_path = os.path.join(current_dir, \"summary_o\", sub_num, f\"{sub_num}-barplot.qzv\")\n",
    "\n",
    "# Load the barplot visualization\n",
    "barplot_viz = Visualization.load(barplot_path)\n",
    "\n",
    "# Visualize the barplot\n",
    "barplot_viz\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Licenzing and Citations:\n",
    "QIIME2 overall:\n",
    "\n",
    "\n",
    "QIIME2 pluging:\n",
    "\n",
    "\n",
    "classifiers:\n",
    "\n",
    "Silva:\n",
    "Michael S Robeson II, Devon R O’Rourke, Benjamin D Kaehler, Michal Ziemski, Matthew R Dillon, Jeffrey T Foster, Nicholas A Bokulich. RESCRIPt: Reproducible sequence taxonomy reference database management for the masses. bioRxiv 2020.10.05.326504; doi: https://doi.org/10.1101/2020.10.05.326504\n",
    "\n",
    "See the SILVA website for the latest citation information for SILVA. https://www.arb-silva.de/\n",
    "\n",
    "GTDB:\n",
    "UUID: 5d5461cc-6a51-434b-90ab-040f388e4221\n",
    "SHA256: 07aadcf7472d9cc6f853f6b4615348619f1a3eceb56c1fb1b6d8dbb20554765f\n",
    "Sklearn Version: 1.4.2\n",
    "Date Trained: 2024-05-30\n",
    "Citations: Parks et al. (2021), Parks et al. (2020), Parks et al. (2018), Rinke et al. (2021)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2",
   "language": "python",
   "name": "qiime2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
