{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from qiime2 import Artifact\n",
    "import pandas as pd\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Taxon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...</td>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>d1438cbf44c49ecae37c506c03115689</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>d346c8a02aff2fa769c2e94e95447488</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>29841a48cdb3345f69e10a0bbb13f91f</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>7f6d5a06aee44bcfc8e166fa2a8ed03d</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ca349b86996171347f6b5afb7b9b5e6c</td>\n",
       "      <td>(((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...</td>\n",
       "      <td>k__Bacteria; p__OD1; c__; o__; f__; g__; s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>bc2eec26f20121d03c3ff6790b97fcbb</td>\n",
       "      <td>(((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...</td>\n",
       "      <td>k__Bacteria; p__OD1; c__; o__; f__; g__; s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>faef44dd4e01f7c4519a0b3a8979e6ad</td>\n",
       "      <td>(((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...</td>\n",
       "      <td>k__Bacteria; p__OD1; c__; o__; f__; g__; s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>72909c55c824fa230fa38e17d57ff46e</td>\n",
       "      <td>(((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...</td>\n",
       "      <td>k__Bacteria; p__OD1; c__; o__; f__; g__; s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>cec58968560141b8c1f19ff3cd3dbc32</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>e80dd587b56f01005f1449880db7c024</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((T...</td>\n",
       "      <td>k__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>c0daf916a4fca792dbd956ba1377b536</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((T...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>e7ee63fee0af657e48429aacc96113bd</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...</td>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>12fe9bc4b3c216f4b6bb7163e226582b</td>\n",
       "      <td>(((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...</td>\n",
       "      <td>k__Bacteria; p__OD1; c__; o__; f__; g__; s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>d6588d9cc8688d706ce164d37a89c2c7</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((T...</td>\n",
       "      <td>k__Bacteria; p__Proteobacteria; c__Gammaproteo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>5c7eee06342c71e0c3119eebdf1fa0d4</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>8efb26fcf4cd6c6fc0ee2e581eaedf2e</td>\n",
       "      <td>(((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...</td>\n",
       "      <td>k__Bacteria; p__OD1; c__; o__; f__; g__; s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>8826dffa3e49fc78e4d235b5cbf995ab</td>\n",
       "      <td>(((T)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...</td>\n",
       "      <td>k__Bacteria; p__OD1; c__; o__; f__; g__; s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1f6adeb643bf0bf9ff5770df8c8a58b9</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((T...</td>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>136797f7a299de78fb2c0dac4dc756c5</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...</td>\n",
       "      <td>k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>4a2738a8f5e31068f68a08ec167589d6</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...</td>\n",
       "      <td>k__Bacteria; p__Proteobacteria; c__Epsilonprot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>3862cdba0e52f6b9e73fb73aabcee8fc</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>77c0ee7d06bf3015aaf5b3dbb4e5cf47</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>39dfd641e980997eb09e0fbfa32515bc</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>8bfb7db560fbb32ad2911b65634a5c6f</td>\n",
       "      <td>(((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...</td>\n",
       "      <td>k__Bacteria; p__OD1; c__; o__; f__; g__; s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>10ef72bab177aa520342b617f1ee2186</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>51b367f222f3c99e8170666c53928e1d</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>2ee4deff7b5c7d3798c0cae18d13632a</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>95fcb2d33bd2a46e98e51e5cb5e4d924</td>\n",
       "      <td>(((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...</td>\n",
       "      <td>k__Bacteria; p__OD1; c__; o__; f__; g__; s__</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>(((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...</td>\n",
       "      <td>k__Bacteria; p__Firmicutes; c__Clostridia; o__...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Feature ID  \\\n",
       "3   4900b14ef37cec718bf18a0d92210a03   \n",
       "4   d1438cbf44c49ecae37c506c03115689   \n",
       "5   d346c8a02aff2fa769c2e94e95447488   \n",
       "11  29841a48cdb3345f69e10a0bbb13f91f   \n",
       "12  7f6d5a06aee44bcfc8e166fa2a8ed03d   \n",
       "13  ca349b86996171347f6b5afb7b9b5e6c   \n",
       "14  bc2eec26f20121d03c3ff6790b97fcbb   \n",
       "18  faef44dd4e01f7c4519a0b3a8979e6ad   \n",
       "20  72909c55c824fa230fa38e17d57ff46e   \n",
       "21  cec58968560141b8c1f19ff3cd3dbc32   \n",
       "22  e80dd587b56f01005f1449880db7c024   \n",
       "23  c0daf916a4fca792dbd956ba1377b536   \n",
       "25  e7ee63fee0af657e48429aacc96113bd   \n",
       "26  12fe9bc4b3c216f4b6bb7163e226582b   \n",
       "28  d6588d9cc8688d706ce164d37a89c2c7   \n",
       "30  5c7eee06342c71e0c3119eebdf1fa0d4   \n",
       "32  8efb26fcf4cd6c6fc0ee2e581eaedf2e   \n",
       "34  8826dffa3e49fc78e4d235b5cbf995ab   \n",
       "36  1f6adeb643bf0bf9ff5770df8c8a58b9   \n",
       "37  136797f7a299de78fb2c0dac4dc756c5   \n",
       "39  4a2738a8f5e31068f68a08ec167589d6   \n",
       "40  3862cdba0e52f6b9e73fb73aabcee8fc   \n",
       "42  77c0ee7d06bf3015aaf5b3dbb4e5cf47   \n",
       "44  39dfd641e980997eb09e0fbfa32515bc   \n",
       "48  8bfb7db560fbb32ad2911b65634a5c6f   \n",
       "49  10ef72bab177aa520342b617f1ee2186   \n",
       "50  51b367f222f3c99e8170666c53928e1d   \n",
       "51  2ee4deff7b5c7d3798c0cae18d13632a   \n",
       "52  95fcb2d33bd2a46e98e51e5cb5e4d924   \n",
       "53  9186e853f994fd8285d68ccd223cd34d   \n",
       "\n",
       "                                             Sequence  \\\n",
       "3   (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...   \n",
       "4   (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...   \n",
       "5   (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...   \n",
       "11  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...   \n",
       "12  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...   \n",
       "13  (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...   \n",
       "14  (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...   \n",
       "18  (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...   \n",
       "20  (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...   \n",
       "21  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...   \n",
       "22  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((T...   \n",
       "23  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((T...   \n",
       "25  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...   \n",
       "26  (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...   \n",
       "28  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((T...   \n",
       "30  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...   \n",
       "32  (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...   \n",
       "34  (((T)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...   \n",
       "36  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((T...   \n",
       "37  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...   \n",
       "39  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...   \n",
       "40  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...   \n",
       "42  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...   \n",
       "44  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...   \n",
       "48  (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...   \n",
       "49  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((C...   \n",
       "50  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...   \n",
       "51  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...   \n",
       "52  (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...   \n",
       "53  (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...   \n",
       "\n",
       "                                                Taxon  \n",
       "3   k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...  \n",
       "4   k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "5   k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "11  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "12  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "13       k__Bacteria; p__OD1; c__; o__; f__; g__; s__  \n",
       "14       k__Bacteria; p__OD1; c__; o__; f__; g__; s__  \n",
       "18       k__Bacteria; p__OD1; c__; o__; f__; g__; s__  \n",
       "20       k__Bacteria; p__OD1; c__; o__; f__; g__; s__  \n",
       "21  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "22  k__Bacteria; p__Proteobacteria; c__Gammaproteo...  \n",
       "23  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "25  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...  \n",
       "26       k__Bacteria; p__OD1; c__; o__; f__; g__; s__  \n",
       "28  k__Bacteria; p__Proteobacteria; c__Gammaproteo...  \n",
       "30  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "32       k__Bacteria; p__OD1; c__; o__; f__; g__; s__  \n",
       "34       k__Bacteria; p__OD1; c__; o__; f__; g__; s__  \n",
       "36  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...  \n",
       "37  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...  \n",
       "39  k__Bacteria; p__Proteobacteria; c__Epsilonprot...  \n",
       "40  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "42  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "44  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "48       k__Bacteria; p__OD1; c__; o__; f__; g__; s__  \n",
       "49  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "50  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "51  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  \n",
       "52       k__Bacteria; p__OD1; c__; o__; f__; g__; s__  \n",
       "53  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rep_seqs_qza = Artifact.load('qiime2_blast_data/SRA1038019-rep_seqs.qza') \n",
    "taxonomy_qza = Artifact.load('qiime2_blast_data/SRA1038019-taxonomy.qza')\n",
    "\n",
    "rep_seqs = rep_seqs_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "rep_seqs.index.name = 'Feature ID'\n",
    "tax = taxonomy_qza.view(pd.Series)\n",
    "tax.index.name = 'Feature ID' \n",
    "\n",
    "merged = pd.merge(rep_seqs, tax, left_index=True, right_index=True)\n",
    "merged.reset_index(inplace=True)\n",
    "unassigned = merged[merged['Taxon'].str.endswith('s__')]\n",
    "unique_values_count = unassigned.iloc[:, 0].nunique()\n",
    "unassigned\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def blast_classifier(rep_seqs_path, taxonomy_path, top_n=10):\n",
    "    '''\n",
    "    This classifier classifies the sequences by performing BLAST search on the unassigned sequences.\n",
    "    \n",
    "    Parameters:\n",
    "    rep_seqs_path (str): path to the representative sequences file\n",
    "    taxonomy_path (str): path to the taxonomy file\n",
    "    top_n (int): number of top alignments to consider each BLAST search\n",
    "     max_results (int): number of unassigned sequences to process\n",
    "      \n",
    "    Returns:\n",
    "    pd.DataFrame: DataFrame containing BLAST results.  \n",
    "    '''\n",
    "    # load in files\n",
    "    rep_seqs_qza = Artifact.load(rep_seqs_path) \n",
    "    taxonomy_qza = Artifact.load(taxonomy_path)\n",
    "\n",
    "    # convert taxonomy to df\n",
    "    taxonomy_df = taxonomy_qza.view(pd.DataFrame)\n",
    "\n",
    "    # convert req seqs to df\n",
    "    rep_seqs = rep_seqs_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "    rep_seqs.index.name = 'Feature ID'\n",
    "\n",
    "    # merge dfs on Feature ID\n",
    "    merged = pd.merge(rep_seqs, taxonomy_df, left_index=True, right_index=True)\n",
    "    merged.reset_index(inplace=True)\n",
    "\n",
    "    # filter for entries without species assignment\n",
    "    unassigned = merged[merged['Taxon'].str.endswith('s__')]\n",
    "\n",
    "    # empty list for BLAST results\n",
    "    blast_results = []\n",
    "\n",
    "    # perform BLAST for each unassigned sequencec\n",
    "    for feature_id, row in unassigned.iterrows():\n",
    "        sequence = row['Sequence']\n",
    "        \n",
    "        # BLAST search\n",
    "        result_handle = NCBIWWW.qblast('blastn', 'nt', sequence)\n",
    "        blast_record = NCBIXML.read(result_handle)\n",
    "        if blast_record.alignments:\n",
    "            for alignment in blast_record.alignments[:top_n]:\n",
    "                for hsp in alignment.hsps:\n",
    "                    #taxonomy = row['Taxon']\n",
    "                    blast_results.append({\n",
    "                        'Feature ID': feature_id,\n",
    "                        'expect': hsp.expect,\n",
    "                        'score': hsp.score,\n",
    "                        'bits': hsp.bits, \n",
    "                        'align_length': hsp.align_length,\n",
    "                        'identities': hsp.identities, \n",
    "                        'positives': hsp.positives,\n",
    "                        'gaps': hsp.gaps,\n",
    "                        'query': hsp.query,\n",
    "                        'sbjct': hsp.sbjct,\n",
    "                        'match': hsp.match,\n",
    "                        #'taxonomy': taxonomy  \n",
    "                    })\n",
    "                time.sleep(1)\n",
    "\n",
    "    blast_results_df = pd.DataFrame(blast_results)\n",
    "\n",
    "    return blast_results_df\n",
    "\n",
    "# file paths\n",
    "rep_seqs_path = 'qiime2_blast_data/SRA1038019-rep_seqs.qza'\n",
    "taxonomy_path = 'qiime2_blast_data/SRA1038019-taxonomy.qza'\n",
    "\n",
    "# Load the taxonomy artifact\n",
    "#taxonomy_qza_path = 'qiime2_blast_data/ERA1115542-taxonomy.qza'\n",
    "#taxonomy_qza = Artifact.load(taxonomy_qza_path)\n",
    "\n",
    "# Convert the taxonomy artifact to a DataFrame\n",
    "#taxonomy_df = taxonomy_qza.view(pd.DataFrame)\n",
    "\n",
    "# Display the first few rows of the taxonomy DataFrame\n",
    "#taxonomy_df.head()\n",
    "\n",
    "# Display the first few rows of the taxonomy DataFrame\n",
    "#print(taxonomy_df.head())\n",
    "\n",
    "# Save the DataFrame to a CSV file for further inspection if needed\n",
    "#taxonomy_df.to_csv('taxonomy.csv', index=False)\n",
    "\n",
    "# Perform BLAST classification and get results\n",
    "results_df = blast_classifier(rep_seqs_path, taxonomy_path) \n",
    "\n",
    "# Display the results\n",
    "results_df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_blast_taxonomy(sequence, expect_threshold=0.001):\n",
    "    result_handle = NCBIWWW.qblast('blastn', 'nt', sequence)\n",
    "    blast_record = NCBIXML.read(result_handle)\n",
    "    for alignment in blast_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            if hsp.expect < expect_threshold:\n",
    "                taxonomy = f\"k__{alignment.hit_def.split(';')[0]}; p__{alignment.hit_def.split(';')[1]}; \" \\\n",
    "                           f\"c__{alignment.hit_def.split(';')[2]}; o__{alignment.hit_def.split(';')[3]}; \" \\\n",
    "                           f\"f__{alignment.hit_def.split(';')[4]}; g__{alignment.hit_def.split(';')[5]}; \" \\\n",
    "                           f\"s__{alignment.hit_def.split(';')[6]}\"\n",
    "                return taxonomy\n",
    "    return None\n",
    "\n",
    "# perform BLAST for each sequence and create reference database files\n",
    "otu_list = []\n",
    "taxonomy_list = []\n",
    "\n",
    "for feature_id, sequence in rep_seqs.items():\n",
    "    taxonomy = get_blast_taxonomy(sequence)\n",
    "    if taxonomy:\n",
    "        otu_list.append(feature_id)\n",
    "        taxonomy_list.append(taxonomy)\n",
    "\n",
    "# df for OTUs and Taxonomy\n",
    "otu_df = pd.DataFrame({'Feature ID': otu_list, 'Taxonomy': taxonomy_list})\n",
    "otu_artifact = Artifact.import_data('FeatureData[Sequence]', otu_df)\n",
    "taxonomy_artifact = Artifact.import_data('FeatureData[Taxonomy]', otu_df)\n",
    "\n",
    "# save artifacts\n",
    "otu_artifact.save('otus.qza')\n",
    "taxonomy_artifact.save('taxonomy.qza')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#naive bayes\n",
    "# import necessary plugins\n",
    "qiime tools import \\\n",
    "  --type 'FeatureData[Sequence]' \\\n",
    "  --input-path otus.qza \\\n",
    "  --output-path ref-seqs.qza\n",
    "\n",
    "qiime tools import \\\n",
    "  --type 'FeatureData[Taxonomy]' \\\n",
    "  --input-path taxonomy.qza \\\n",
    "  --output-path ref-taxonomy.qza\n",
    "\n",
    "# Train the classifier\n",
    "qiime feature-classifier fit-classifier-naive-bayes \\\n",
    "  --i-reference-reads ref-seqs.qza \\\n",
    "  --i-reference-taxonomy ref-taxonomy.qza \\\n",
    "  --o-classifier classifier.qza\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2",
   "language": "python",
   "name": "qiime2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
