{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "811b7910",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: line 1: activate: command not found\n"
     ]
    }
   ],
   "source": [
    "#dependencies\n",
    "#MIDAS2 setup: https://midas2.readthedocs.io/en/latest/installation.html\n",
    "#!conda activate midas2.0\n",
    "#!conda install -c bioconda fastqc --yes\n",
    "#!conda install -c bioconda multiqc --yes\n",
    "#!conda install -c bioconda trimmomatic --yes\n",
    "#!pip install ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e38470cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "null\n",
      "null\n",
      "application/gzip\n",
      "application/gzip\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of c2671_R1.fastq\n",
      "Approx 5% complete for c2671_R1.fastq\n",
      "Approx 10% complete for c2671_R1.fastq\n",
      "Approx 15% complete for c2671_R1.fastq\n",
      "Approx 20% complete for c2671_R1.fastq\n",
      "Approx 25% complete for c2671_R1.fastq\n",
      "Approx 30% complete for c2671_R1.fastq\n",
      "Approx 35% complete for c2671_R1.fastq\n",
      "Approx 40% complete for c2671_R1.fastq\n",
      "Approx 45% complete for c2671_R1.fastq\n",
      "Approx 50% complete for c2671_R1.fastq\n",
      "Approx 55% complete for c2671_R1.fastq\n",
      "Approx 60% complete for c2671_R1.fastq\n",
      "Approx 65% complete for c2671_R1.fastq\n",
      "Approx 70% complete for c2671_R1.fastq\n",
      "Approx 75% complete for c2671_R1.fastq\n",
      "Approx 80% complete for c2671_R1.fastq\n",
      "Approx 85% complete for c2671_R1.fastq\n",
      "Approx 90% complete for c2671_R1.fastq\n",
      "Approx 95% complete for c2671_R1.fastq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for c2671_R1.fastq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of c2671_R2.fastq\n",
      "Approx 5% complete for c2671_R2.fastq\n",
      "Approx 10% complete for c2671_R2.fastq\n",
      "Approx 15% complete for c2671_R2.fastq\n",
      "Approx 20% complete for c2671_R2.fastq\n",
      "Approx 25% complete for c2671_R2.fastq\n",
      "Approx 30% complete for c2671_R2.fastq\n",
      "Approx 35% complete for c2671_R2.fastq\n",
      "Approx 40% complete for c2671_R2.fastq\n",
      "Approx 45% complete for c2671_R2.fastq\n",
      "Approx 50% complete for c2671_R2.fastq\n",
      "Approx 55% complete for c2671_R2.fastq\n",
      "Approx 60% complete for c2671_R2.fastq\n",
      "Approx 65% complete for c2671_R2.fastq\n",
      "Approx 70% complete for c2671_R2.fastq\n",
      "Approx 75% complete for c2671_R2.fastq\n",
      "Approx 80% complete for c2671_R2.fastq\n",
      "Approx 85% complete for c2671_R2.fastq\n",
      "Approx 90% complete for c2671_R2.fastq\n",
      "Approx 95% complete for c2671_R2.fastq\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for c2671_R2.fastq\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of sample1_R1.fastq.gz\n",
      "Approx 5% complete for sample1_R1.fastq.gz\n",
      "Approx 10% complete for sample1_R1.fastq.gz\n",
      "Approx 15% complete for sample1_R1.fastq.gz\n",
      "Approx 20% complete for sample1_R1.fastq.gz\n",
      "Approx 25% complete for sample1_R1.fastq.gz\n",
      "Approx 30% complete for sample1_R1.fastq.gz\n",
      "Approx 35% complete for sample1_R1.fastq.gz\n",
      "Approx 40% complete for sample1_R1.fastq.gz\n",
      "Approx 45% complete for sample1_R1.fastq.gz\n",
      "Approx 50% complete for sample1_R1.fastq.gz\n",
      "Approx 55% complete for sample1_R1.fastq.gz\n",
      "Approx 60% complete for sample1_R1.fastq.gz\n",
      "Approx 65% complete for sample1_R1.fastq.gz\n",
      "Approx 70% complete for sample1_R1.fastq.gz\n",
      "Approx 75% complete for sample1_R1.fastq.gz\n",
      "Approx 80% complete for sample1_R1.fastq.gz\n",
      "Approx 85% complete for sample1_R1.fastq.gz\n",
      "Approx 90% complete for sample1_R1.fastq.gz\n",
      "Approx 95% complete for sample1_R1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for sample1_R1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Started analysis of sample2_R1.fastq.gz\n",
      "Approx 5% complete for sample2_R1.fastq.gz\n",
      "Approx 10% complete for sample2_R1.fastq.gz\n",
      "Approx 15% complete for sample2_R1.fastq.gz\n",
      "Approx 20% complete for sample2_R1.fastq.gz\n",
      "Approx 25% complete for sample2_R1.fastq.gz\n",
      "Approx 30% complete for sample2_R1.fastq.gz\n",
      "Approx 35% complete for sample2_R1.fastq.gz\n",
      "Approx 40% complete for sample2_R1.fastq.gz\n",
      "Approx 45% complete for sample2_R1.fastq.gz\n",
      "Approx 50% complete for sample2_R1.fastq.gz\n",
      "Approx 55% complete for sample2_R1.fastq.gz\n",
      "Approx 60% complete for sample2_R1.fastq.gz\n",
      "Approx 65% complete for sample2_R1.fastq.gz\n",
      "Approx 70% complete for sample2_R1.fastq.gz\n",
      "Approx 75% complete for sample2_R1.fastq.gz\n",
      "Approx 80% complete for sample2_R1.fastq.gz\n",
      "Approx 85% complete for sample2_R1.fastq.gz\n",
      "Approx 90% complete for sample2_R1.fastq.gz\n",
      "Approx 95% complete for sample2_R1.fastq.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis complete for sample2_R1.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=884562;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ üîç \u001b[2m| v1.16\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /home/miloscola/ab/pipe/qual_rep_temp/c2671_R1_fastqc.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /home/miloscola/ab/pipe/qual_rep_temp/c2671_R1_fastqc.zip\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /home/miloscola/ab/pipe/qual_rep_temp/c2671_R2_fastqc.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /home/miloscola/ab/pipe/qual_rep_temp/c2671_R2_fastqc.zip\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /home/miloscola/ab/pipe/qual_rep_temp/sample1_R1_fastqc.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /home/miloscola/ab/pipe/qual_rep_temp/sample1_R1_fastqc.zip\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /home/miloscola/ab/pipe/qual_rep_temp/sample2_R1_fastqc.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /home/miloscola/ab/pipe/qual_rep_temp/sample2_R1_fastqc.zip\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m8/8\u001b[0m  0/8\u001b[0m  \n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 4 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : combined_quality_reports/multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : combined_quality_reports/multiqc_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#generate quality report\n",
    "fastqc untrimmed_fastq/* -o qual_rep_temp\n",
    "multiqc qual_rep_temp/* -o combined_quality_reports\n",
    "rm qual_rep_temp/*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a5c8b627",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "TrimmomaticSE: Started with arguments:\n",
      " c2671_R1.fastq c2671_R1.trim.fastq SLIDINGWINDOW:4:20 MINLEN:25\n",
      "Automatically using 1 threads\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 50040 Surviving: 48288 (96.50%) Dropped: 1752 (3.50%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " sample1_R1.fastq sample1_R1.trim.fastq SLIDINGWINDOW:4:20 MINLEN:25\n",
      "Automatically using 1 threads\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 353318 Surviving: 353257 (99.98%) Dropped: 61 (0.02%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "TrimmomaticSE: Started with arguments:\n",
      " sample2_R1.fastq sample2_R1.trim.fastq SLIDINGWINDOW:4:20 MINLEN:25\n",
      "Automatically using 1 threads\n",
      "Quality encoding detected as phred33\n",
      "Input Reads: 497377 Surviving: 497285 (99.98%) Dropped: 92 (0.02%)\n",
      "TrimmomaticSE: Completed successfully\n",
      "mv: cannot stat 'untrimmed_fastq/*un.trim.fastq': No such file or directory\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#clean data\n",
    "cd untrimmed_fastq\n",
    "for infile in *_R1.fastq\n",
    " do\n",
    "   base=$(basename ${infile} _R1.fastq)\n",
    "   trimmomatic SE ${infile} ${base}_R1.trim.fastq \\\n",
    "                SLIDINGWINDOW:4:20 MINLEN:25 #ILLUMINACLIP:NexteraPE-PE.fa:2:40:15 \n",
    " done\n",
    " cd ..\n",
    "#move data into proper files TODO: make folder for processed fastq files\n",
    "mv untrimmed_fastq/*.trim.fastq reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b6a68c7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "1697090441.1:  Single sample abundant species profiling in subcommand run_species with args\n",
      "1697090441.1:  {\n",
      "1697090441.1:      \"subcommand\": \"run_species\",\n",
      "1697090441.1:      \"force\": false,\n",
      "1697090441.1:      \"debug\": false,\n",
      "1697090441.1:      \"zzz_worker_mode\": false,\n",
      "1697090441.1:      \"batch_branch\": \"master\",\n",
      "1697090441.1:      \"batch_memory\": 378880,\n",
      "1697090441.1:      \"batch_vcpus\": 48,\n",
      "1697090441.1:      \"batch_queue\": \"pairani\",\n",
      "1697090441.1:      \"batch_ecr_image\": \"pairani:latest\",\n",
      "1697090441.1:      \"midas_outdir\": \"midas2_output\",\n",
      "1697090441.1:      \"sample_name\": \"c2671\",\n",
      "1697090441.1:      \"r1\": \"reads/c2671_R1.trim.fastq\",\n",
      "1697090441.1:      \"r2\": null,\n",
      "1697090441.1:      \"midasdb_name\": \"uhgg\",\n",
      "1697090441.1:      \"midasdb_dir\": \"my_midasdb_uhgg\",\n",
      "1697090441.1:      \"word_size\": 28,\n",
      "1697090441.1:      \"aln_mapid\": null,\n",
      "1697090441.1:      \"aln_cov\": 0.75,\n",
      "1697090441.1:      \"marker_reads\": 2,\n",
      "1697090441.1:      \"marker_covered\": 2,\n",
      "1697090441.1:      \"max_reads\": null,\n",
      "1697090441.1:      \"num_cores\": 4\n",
      "1697090441.1:  }\n",
      "1697090441.1:  Create OUTPUT directory for c2671.\n",
      "1697090441.1:  'rm -rf midas2_output/c2671/species'\n",
      "1697090441.1:  'mkdir -p midas2_output/c2671/species'\n",
      "1697090441.1:  Create TEMP directory for c2671.\n",
      "1697090441.1:  'rm -rf midas2_output/c2671/temp/species'\n",
      "1697090441.1:  'mkdir -p midas2_output/c2671/temp/species'\n",
      "1697090441.2:  MIDAS2::fetch_midasdb_files::start\n",
      "1697090442.7:  MIDAS2::fetch_midasdb_files::finish\n",
      "1697090442.7:  MIDAS2::map_reads_hsblastn::start\n",
      "[HS-BLASTN] Loading database.\n",
      "\tLoading /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa.sequence, size = 0.04GB\n",
      "\tLoading /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa.bwt, size = 0.4GB\n",
      "\tLoading /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa.sa, size = 0.08GB\n",
      "[HS-BLASTN] done. Time elapsed: 0.16 secs.\n",
      "\n",
      "\n",
      "[HS-BLASTN] Processing /dev/stdin.\n",
      "\tProcessing 32000 queries.\n",
      "1697090443.2:  parse_reads:: parsed 48288 reads from reads/c2671_R1.trim.fastq\n",
      "\tProcessing 16288 queries.\n",
      "\n",
      "\n",
      "\n",
      "[HS-BLASTN] done. Elpased time: 0.3920 secs.\n",
      "[HS-BLASTN] 48288 queries processed.\n",
      "\n",
      "[CMD] hs-blastn align -outfmt 6 -num_threads 4 -evalue 1e-3 -word_size 28 -query /dev/stdin -db /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa \n",
      "1697090443.3:  MIDAS2::map_reads_hsblastn::finish\n",
      "1697090443.3:  MIDAS2::read in marker information::start\n",
      "1697090444.0:  MIDAS2::read in marker information::finish\n",
      "1697090444.0:  MIDAS2::find_best_hits::start\n",
      "1697090444.0:    total alignments: 3718\n",
      "1697090444.0:  MIDAS2::find_best_hits::finish\n",
      "1697090444.0:  MIDAS2::assign_unique::start\n",
      "1697090444.0:   uniquely mapped reads: 0\n",
      "1697090444.0:   ambiguously mapped reads: 0\n",
      "1697090444.0:  MIDAS2::assign_unique::finish\n",
      "1697090444.0:  MIDAS2::assign_non_unique::start\n",
      "1697090444.0:  MIDAS2::assign_non_unique::finish\n",
      "1697090444.0:  MIDAS2::normalize_counts::start\n",
      "1697090444.0:    total marker-gene coverage 0.000\n",
      "1697090444.0:  MIDAS2::normalize_counts::finish\n",
      "1697090444.4:  Single sample abundant species profiling in subcommand run_species with args\n",
      "1697090444.4:  {\n",
      "1697090444.4:      \"subcommand\": \"run_species\",\n",
      "1697090444.4:      \"force\": false,\n",
      "1697090444.4:      \"debug\": false,\n",
      "1697090444.4:      \"zzz_worker_mode\": false,\n",
      "1697090444.4:      \"batch_branch\": \"master\",\n",
      "1697090444.4:      \"batch_memory\": 378880,\n",
      "1697090444.4:      \"batch_vcpus\": 48,\n",
      "1697090444.4:      \"batch_queue\": \"pairani\",\n",
      "1697090444.4:      \"batch_ecr_image\": \"pairani:latest\",\n",
      "1697090444.4:      \"midas_outdir\": \"midas2_output\",\n",
      "1697090444.4:      \"sample_name\": \"sample1\",\n",
      "1697090444.4:      \"r1\": \"reads/sample1_R1.trim.fastq\",\n",
      "1697090444.4:      \"r2\": null,\n",
      "1697090444.4:      \"midasdb_name\": \"uhgg\",\n",
      "1697090444.4:      \"midasdb_dir\": \"my_midasdb_uhgg\",\n",
      "1697090444.4:      \"word_size\": 28,\n",
      "1697090444.4:      \"aln_mapid\": null,\n",
      "1697090444.4:      \"aln_cov\": 0.75,\n",
      "1697090444.4:      \"marker_reads\": 2,\n",
      "1697090444.4:      \"marker_covered\": 2,\n",
      "1697090444.4:      \"max_reads\": null,\n",
      "1697090444.4:      \"num_cores\": 4\n",
      "1697090444.4:  }\n",
      "1697090444.4:  Create OUTPUT directory for sample1.\n",
      "1697090444.4:  'rm -rf midas2_output/sample1/species'\n",
      "1697090444.4:  'mkdir -p midas2_output/sample1/species'\n",
      "1697090444.4:  Create TEMP directory for sample1.\n",
      "1697090444.4:  'rm -rf midas2_output/sample1/temp/species'\n",
      "1697090444.4:  'mkdir -p midas2_output/sample1/temp/species'\n",
      "1697090444.4:  MIDAS2::fetch_midasdb_files::start\n",
      "1697090445.9:  MIDAS2::fetch_midasdb_files::finish\n",
      "1697090445.9:  MIDAS2::map_reads_hsblastn::start\n",
      "[HS-BLASTN] Loading database.\n",
      "\tLoading /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa.sequence, size = 0.04GB\n",
      "\tLoading /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa.bwt, size = 0.4GB\n",
      "\tLoading /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa.sa, size = 0.08GB\n",
      "[HS-BLASTN] done. Time elapsed: 0.16 secs.\n",
      "\n",
      "\n",
      "[HS-BLASTN] Processing /dev/stdin.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "1697090448.2:  parse_reads:: parsed 353257 reads from reads/sample1_R1.trim.fastq\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 1257 queries.\n",
      "\n",
      "\n",
      "\n",
      "[HS-BLASTN] done. Elpased time: 2.2024 secs.\n",
      "[HS-BLASTN] 353257 queries processed.\n",
      "\n",
      "[CMD] hs-blastn align -outfmt 6 -num_threads 4 -evalue 1e-3 -word_size 28 -query /dev/stdin -db /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa \n",
      "1697090448.3:  MIDAS2::map_reads_hsblastn::finish\n",
      "1697090448.3:  MIDAS2::read in marker information::start\n",
      "1697090448.9:  MIDAS2::read in marker information::finish\n",
      "1697090448.9:  MIDAS2::find_best_hits::start\n",
      "1697090449.0:    total alignments: 7084\n",
      "1697090449.0:  MIDAS2::find_best_hits::finish\n",
      "1697090449.0:  MIDAS2::assign_unique::start\n",
      "1697090449.0:   uniquely mapped reads: 1049\n",
      "1697090449.0:   ambiguously mapped reads: 55\n",
      "1697090449.0:  MIDAS2::assign_unique::finish\n",
      "1697090449.0:  MIDAS2::assign_non_unique::start\n",
      "1697090449.0:  MIDAS2::assign_non_unique::finish\n",
      "1697090449.0:  MIDAS2::normalize_counts::start\n",
      "1697090449.0:    total marker-gene coverage 12.904\n",
      "1697090449.0:  MIDAS2::normalize_counts::finish\n",
      "1697090449.3:  Single sample abundant species profiling in subcommand run_species with args\n",
      "1697090449.3:  {\n",
      "1697090449.3:      \"subcommand\": \"run_species\",\n",
      "1697090449.3:      \"force\": false,\n",
      "1697090449.3:      \"debug\": false,\n",
      "1697090449.3:      \"zzz_worker_mode\": false,\n",
      "1697090449.3:      \"batch_branch\": \"master\",\n",
      "1697090449.3:      \"batch_memory\": 378880,\n",
      "1697090449.3:      \"batch_vcpus\": 48,\n",
      "1697090449.3:      \"batch_queue\": \"pairani\",\n",
      "1697090449.3:      \"batch_ecr_image\": \"pairani:latest\",\n",
      "1697090449.3:      \"midas_outdir\": \"midas2_output\",\n",
      "1697090449.3:      \"sample_name\": \"sample2\",\n",
      "1697090449.3:      \"r1\": \"reads/sample2_R1.trim.fastq\",\n",
      "1697090449.3:      \"r2\": null,\n",
      "1697090449.3:      \"midasdb_name\": \"uhgg\",\n",
      "1697090449.3:      \"midasdb_dir\": \"my_midasdb_uhgg\",\n",
      "1697090449.3:      \"word_size\": 28,\n",
      "1697090449.3:      \"aln_mapid\": null,\n",
      "1697090449.3:      \"aln_cov\": 0.75,\n",
      "1697090449.3:      \"marker_reads\": 2,\n",
      "1697090449.3:      \"marker_covered\": 2,\n",
      "1697090449.3:      \"max_reads\": null,\n",
      "1697090449.3:      \"num_cores\": 4\n",
      "1697090449.3:  }\n",
      "1697090449.3:  Create OUTPUT directory for sample2.\n",
      "1697090449.3:  'rm -rf midas2_output/sample2/species'\n",
      "1697090449.3:  'mkdir -p midas2_output/sample2/species'\n",
      "1697090449.3:  Create TEMP directory for sample2.\n",
      "1697090449.3:  'rm -rf midas2_output/sample2/temp/species'\n",
      "1697090449.3:  'mkdir -p midas2_output/sample2/temp/species'\n",
      "1697090449.3:  MIDAS2::fetch_midasdb_files::start\n",
      "1697090450.8:  MIDAS2::fetch_midasdb_files::finish\n",
      "1697090450.8:  MIDAS2::map_reads_hsblastn::start\n",
      "[HS-BLASTN] Loading database.\n",
      "\tLoading /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa.sequence, size = 0.04GB\n",
      "\tLoading /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa.bwt, size = 0.4GB\n",
      "\tLoading /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa.sa, size = 0.08GB\n",
      "[HS-BLASTN] done. Time elapsed: 0.15 secs.\n",
      "\n",
      "\n",
      "[HS-BLASTN] Processing /dev/stdin.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "\tProcessing 32000 queries.\n",
      "1697090454.1:  parse_reads:: parsed 497285 reads from reads/sample2_R1.trim.fastq\n",
      "\tProcessing 17285 queries.\n",
      "\n",
      "\n",
      "\n",
      "[HS-BLASTN] done. Elpased time: 3.1798 secs.\n",
      "[HS-BLASTN] 497285 queries processed.\n",
      "\n",
      "[CMD] hs-blastn align -outfmt 6 -num_threads 4 -evalue 1e-3 -word_size 28 -query /dev/stdin -db /home/miloscola/ab/pipe/my_midasdb_uhgg/markers/phyeco/phyeco.fa \n",
      "1697090454.2:  MIDAS2::map_reads_hsblastn::finish\n",
      "1697090454.2:  MIDAS2::read in marker information::start\n",
      "1697090454.8:  MIDAS2::read in marker information::finish\n",
      "1697090454.8:  MIDAS2::find_best_hits::start\n",
      "1697090454.9:    total alignments: 4878\n",
      "1697090454.9:  MIDAS2::find_best_hits::finish\n",
      "1697090454.9:  MIDAS2::assign_unique::start\n",
      "1697090454.9:   uniquely mapped reads: 973\n",
      "1697090454.9:   ambiguously mapped reads: 134\n",
      "1697090454.9:  MIDAS2::assign_unique::finish\n",
      "1697090454.9:  MIDAS2::assign_non_unique::start\n",
      "1697090454.9:  MIDAS2::assign_non_unique::finish\n",
      "1697090454.9:  MIDAS2::normalize_counts::start\n",
      "1697090454.9:    total marker-gene coverage 12.612\n",
      "1697090454.9:  MIDAS2::normalize_counts::finish\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#run MIDAS2\n",
    "for sample_name in reads/*_R1.trim.fastq\n",
    " do\n",
    "  base=$(basename ${sample_name} _R1.trim.fastq)\n",
    "  midas2 run_species \\\n",
    "      --sample_name ${base} \\\n",
    "      -1 reads/${base}_R1.trim.fastq \\\n",
    "      --midasdb_name uhgg \\\n",
    "      --midasdb_dir my_midasdb_uhgg \\\n",
    "      --num_cores 4 \\\n",
    "      midas2_output\n",
    "  done\n",
    "\n",
    "#TODO: add refrence genome of face microbiome to database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b0f7c11",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (midas2.0)",
   "language": "python",
   "name": "midas2.0"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
