{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from qiime2 import Artifact\n",
    "import pandas as pd\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ab3a5be3635b079ece06a592de8b5f3f    (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...\n",
       "5dfdd3fb2dd235ff32d3eae7888f6ded    (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...\n",
       "0d96b3b4618186fc901d06b05e8744dc    (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((G...\n",
       "64452b5815ba16e252231dcd811688b2    (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...\n",
       "dd50b4ffbdca915499e738f619259aa8    (((C)), ((C)), ((T)), ((A)), ((C)), ((G)), ((G...\n",
       "                                                          ...                        \n",
       "9dc9bb57a0e20c7db8b9a3e0943dfa40    (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((T...\n",
       "deaa773c59aef232a12ac8467c1e05a0    (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((T...\n",
       "2736f6e813be1cc83bb5b594f5660b3b    (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...\n",
       "6362858494f48d48c18d38c02f8594f9    (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((A...\n",
       "ae99022c9de0af443b3d473270dd6eda    (((G)), ((A)), ((C)), ((T)), ((A)), ((C)), ((G...\n",
       "Length: 14161, dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rep_seqs_qza = Artifact.load('qiime2_blast_data/SRA482819-rep_seqs.qza') \n",
    "#taxonomy_qza = Artifact.load('qiime2_blast_data/ERA1115542-taxonomy.qza')\n",
    "\n",
    "rep_seqs_path = Artifact.load('qiime2_blast_data/SRA482819-rep_seqs.qza')\n",
    "rep_seqs = rep_seqs_qza.view(pd.Series)\n",
    "rep_seqs\n",
    "\n",
    "#rep_seqs = rep_seqs_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "#rep_seqs.index.name = 'Feature ID'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                              Taxon  \\\n",
      "Feature ID                                                                            \n",
      "25cef8e738313df56dded85a4e57dbc5  k__Bacteria; p__Firmicutes; c__Bacilli; o__Bac...   \n",
      "2c2ff5f860fa0e6a03a14152ce6ce0ce  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
      "6722fabb0f59d2d6942dffb65c8ac0c3  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
      "5299a0de540765765e98253cd86f65dd  k__Bacteria; p__Actinobacteria; c__Actinobacte...   \n",
      "0203a5fdabe513861a8936651ef4019f  k__Bacteria; p__Firmicutes; c__Bacilli; o__Lac...   \n",
      "\n",
      "                                          Confidence  \n",
      "Feature ID                                            \n",
      "25cef8e738313df56dded85a4e57dbc5  0.7605489472822277  \n",
      "2c2ff5f860fa0e6a03a14152ce6ce0ce  0.8228727992486198  \n",
      "6722fabb0f59d2d6942dffb65c8ac0c3  0.9900719420545704  \n",
      "5299a0de540765765e98253cd86f65dd  0.9332780981959647  \n",
      "0203a5fdabe513861a8936651ef4019f  0.9999994549609157  \n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-ed5f4dfc6525>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     84\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     85\u001b[0m \u001b[0;31m# Perform BLAST classification and get results\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 86\u001b[0;31m \u001b[0mresults_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mblast_classifier\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrep_seqs_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtaxonomy_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     87\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     88\u001b[0m \u001b[0;31m# Display the results\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-ed5f4dfc6525>\u001b[0m in \u001b[0;36mblast_classifier\u001b[0;34m(rep_seqs_path, taxonomy_path, top_n, max_results)\u001b[0m\n\u001b[1;32m     57\u001b[0m                         \u001b[0;34m'taxonomy'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mtaxonomy\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m                     })\n\u001b[0;32m---> 59\u001b[0;31m                 \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msleep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     60\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     61\u001b[0m     \u001b[0mblast_results_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblast_results\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "def blast_classifier(rep_seqs_path, taxonomy_path, top_n=10, max_results=1):\n",
    "    '''\n",
    "    This classifier classifies the sequences by performing BLAST search on the unassigned sequences.\n",
    "    \n",
    "    Parameters:\n",
    "    rep_seqs_path (str): path to the representative sequences file\n",
    "    taxonomy_path (str): path to the taxonomy file\n",
    "    top_n (int): number of top alignments to consider each BLAST search\n",
    "     max_results (int): number of unassigned sequences to process\n",
    "      \n",
    "    Returns:\n",
    "    pd.DataFrame: DataFrame containing BLAST results.  \n",
    "    '''\n",
    "    # load in files\n",
    "    rep_seqs_qza = Artifact.load(rep_seqs_path) \n",
    "    taxonomy_qza = Artifact.load(taxonomy_path)\n",
    "\n",
    "    # convert taxonomy to df\n",
    "    taxonomy_df = taxonomy_qza.view(pd.DataFrame)\n",
    "\n",
    "    # convert req seqs to df\n",
    "    rep_seqs = rep_seqs_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "    rep_seqs.index.name = 'Feature ID'\n",
    "\n",
    "    # merge dfs on Feature ID\n",
    "    merged = pd.merge(rep_seqs, taxonomy_df, left_index=True, right_index=True)\n",
    "\n",
    "    # filter for entries without species assignment\n",
    "    unassigned = merged[merged['Taxon'].str.endswith('s__')].head(max_results)\n",
    "\n",
    "    # empty list for BLAST results\n",
    "    blast_results = []\n",
    "\n",
    "    # perform BLAST for each unassigned sequencec\n",
    "    for feature_id, row in unassigned.iterrows():\n",
    "        sequence = row['Sequence']\n",
    "        \n",
    "        # BLAST search\n",
    "        result_handle = NCBIWWW.qblast('blastn', 'nt', sequence)\n",
    "        blast_record = NCBIXML.read(result_handle)\n",
    "        if blast_record.alignments:\n",
    "            for alignment in blast_record.alignments[:top_n]:\n",
    "                for hsp in alignment.hsps:\n",
    "                    taxonomy = row['Taxon']\n",
    "                    blast_results.append({\n",
    "                        'Feature ID': feature_id,\n",
    "                        'expect': hsp.expect,\n",
    "                        'score': hsp.score,\n",
    "                        'bits': hsp.bits, \n",
    "                        'align_length': hsp.align_length,\n",
    "                        'identities': hsp.identities, \n",
    "                        'positives': hsp.positives,\n",
    "                        'gaps': hsp.gaps,\n",
    "                        'query': hsp.query,\n",
    "                        'sbjct': hsp.sbjct,\n",
    "                        'match': hsp.match,\n",
    "                        'taxonomy': taxonomy  \n",
    "                    })\n",
    "                time.sleep(1)\n",
    "\n",
    "    blast_results_df = pd.DataFrame(blast_results)\n",
    "\n",
    "    return blast_results_df\n",
    "\n",
    "# file paths\n",
    "rep_seqs_path = 'qiime2_blast_data/ERA1115542-rep_seqs.qza'\n",
    "taxonomy_path = 'qiime2_blast_data/ERA1115542-taxonomy.qza'\n",
    "\n",
    "# Load the taxonomy artifact\n",
    "taxonomy_qza_path = 'qiime2_blast_data/ERA1115542-taxonomy.qza'\n",
    "taxonomy_qza = Artifact.load(taxonomy_qza_path)\n",
    "\n",
    "# Convert the taxonomy artifact to a DataFrame\n",
    "taxonomy_df = taxonomy_qza.view(pd.DataFrame)\n",
    "\n",
    "# Display the first few rows of the taxonomy DataFrame\n",
    "taxonomy_df.head()\n",
    "\n",
    "# Display the first few rows of the taxonomy DataFrame\n",
    "print(taxonomy_df.head())\n",
    "\n",
    "# Save the DataFrame to a CSV file for further inspection if needed\n",
    "#taxonomy_df.to_csv('taxonomy.csv', index=False)\n",
    "\n",
    "# Perform BLAST classification and get results\n",
    "results_df = blast_classifier(rep_seqs_path, taxonomy_path) \n",
    "\n",
    "# Display the results\n",
    "results_df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_blast_taxonomy(sequence, expect_threshold=0.001):\n",
    "    result_handle = NCBIWWW.qblast('blastn', 'nt', sequence)\n",
    "    blast_record = NCBIXML.read(result_handle)\n",
    "    for alignment in blast_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            if hsp.expect < expect_threshold:\n",
    "                taxonomy = f\"k__{alignment.hit_def.split(';')[0]}; p__{alignment.hit_def.split(';')[1]}; \" \\\n",
    "                           f\"c__{alignment.hit_def.split(';')[2]}; o__{alignment.hit_def.split(';')[3]}; \" \\\n",
    "                           f\"f__{alignment.hit_def.split(';')[4]}; g__{alignment.hit_def.split(';')[5]}; \" \\\n",
    "                           f\"s__{alignment.hit_def.split(';')[6]}\"\n",
    "                return taxonomy\n",
    "    return None\n",
    "\n",
    "# perform BLAST for each sequence and create reference database files\n",
    "otu_list = []\n",
    "taxonomy_list = []\n",
    "\n",
    "for feature_id, sequence in rep_seqs.items():\n",
    "    taxonomy = get_blast_taxonomy(sequence)\n",
    "    if taxonomy:\n",
    "        otu_list.append(feature_id)\n",
    "        taxonomy_list.append(taxonomy)\n",
    "\n",
    "# df for OTUs and Taxonomy\n",
    "otu_df = pd.DataFrame({'Feature ID': otu_list, 'Taxonomy': taxonomy_list})\n",
    "otu_artifact = Artifact.import_data('FeatureData[Sequence]', otu_df)\n",
    "taxonomy_artifact = Artifact.import_data('FeatureData[Taxonomy]', otu_df)\n",
    "\n",
    "# save artifacts\n",
    "otu_artifact.save('otus.qza')\n",
    "taxonomy_artifact.save('taxonomy.qza')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#naive bayes\n",
    "# import necessary plugins\n",
    "qiime tools import \\\n",
    "  --type 'FeatureData[Sequence]' \\\n",
    "  --input-path otus.qza \\\n",
    "  --output-path ref-seqs.qza\n",
    "\n",
    "qiime tools import \\\n",
    "  --type 'FeatureData[Taxonomy]' \\\n",
    "  --input-path taxonomy.qza \\\n",
    "  --output-path ref-taxonomy.qza\n",
    "\n",
    "# Train the classifier\n",
    "qiime feature-classifier fit-classifier-naive-bayes \\\n",
    "  --i-reference-reads ref-seqs.qza \\\n",
    "  --i-reference-taxonomy ref-taxonomy.qza \\\n",
    "  --o-classifier classifier.qza\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2",
   "language": "python",
   "name": "qiime2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
