{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from qiime2 import Artifact\n",
    "import pandas as pd\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GACTACAAGGGTATCTAATCCTGTTTGCTACCCACGCTTTCGTACCTCAGCGTCAGATAATGGCCAGAAAGTCGCCTTCGCCACCGGTATTCCTCCTAATATCTGCGCATTTCACCGCTACACTAGGAATTCCACTTTCCCTTCCATCTCTCAAGTTAAACAGTTTTAAAAGCTTACTATAGTTGAGCTATAGCCTTTCACTTCTAACTTTTCTAACCGCCTACGTACCCTTTACGCCCAATGATTCCGGACAACGCTCGGTCCTTACGTATTACCGCGGCTGCTGGCACGTAATTAGCCGGACCTTATTCGTATAGTACTGTCATTTTCTTCCTATACAAAAGATTTTTACAACCCGAAGGCCTTCTAAATCACGCGGCGTCGCTGCATCAGGGTTGCCCCCATTGTGCAAAATTCCCCACTGCTGCCTCCCGTAGG\n"
     ]
    }
   ],
   "source": [
    "rep_seqs_qza = Artifact.load('qiime2_blast_data/SRA1038019-rep_seqs.qza') \n",
    "taxonomy_qza = Artifact.load('qiime2_blast_data/SRA1038019-taxonomy.qza')\n",
    "\n",
    "rep_seqs = rep_seqs_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "rep_seqs.index.name = 'Feature ID'\n",
    "tax = taxonomy_qza.view(pd.Series)\n",
    "tax.index.name = 'Feature ID' \n",
    "\n",
    "merged = pd.merge(rep_seqs, tax, left_index=True, right_index=True)\n",
    "merged.reset_index(inplace=True)\n",
    "unassigned = merged[merged['Taxon'].str.endswith('s__')]\n",
    "unique_values_count = unassigned.iloc[:, 0].nunique()\n",
    "first_sequence = unassigned['Sequence'].iloc[1]\n",
    "tax = unassigned['Taxon'].iloc[1]\n",
    "print (first_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature ID</th>\n",
       "      <th>expect</th>\n",
       "      <th>score</th>\n",
       "      <th>bits</th>\n",
       "      <th>align_length</th>\n",
       "      <th>identities</th>\n",
       "      <th>positives</th>\n",
       "      <th>gaps</th>\n",
       "      <th>query</th>\n",
       "      <th>sbjct</th>\n",
       "      <th>match</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>accession</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>914.0</td>\n",
       "      <td>825.426</td>\n",
       "      <td>462</td>\n",
       "      <td>460</td>\n",
       "      <td>460</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas bennonis strain WAL 1926C 16S ri...</td>\n",
       "      <td>NR_044491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>909.0</td>\n",
       "      <td>820.917</td>\n",
       "      <td>462</td>\n",
       "      <td>459</td>\n",
       "      <td>459</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas bennonis gene for 16S ribosomal ...</td>\n",
       "      <td>NR_113228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>895.0</td>\n",
       "      <td>808.294</td>\n",
       "      <td>462</td>\n",
       "      <td>455</td>\n",
       "      <td>455</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||| ||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas bennonis strain A7 16S ribosomal...</td>\n",
       "      <td>OP269714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>2.252820e-166</td>\n",
       "      <td>665.0</td>\n",
       "      <td>600.906</td>\n",
       "      <td>462</td>\n",
       "      <td>411</td>\n",
       "      <td>411</td>\n",
       "      <td>1</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGATCCCCACGCTTTCGTGCA...</td>\n",
       "      <td>|||||| |||||||||||||||||| | ||||||||||||||||| ...</td>\n",
       "      <td>Porphyromonadaceae bacterium FC4 partial 16S r...</td>\n",
       "      <td>LN881610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>9.579250e-165</td>\n",
       "      <td>659.0</td>\n",
       "      <td>595.496</td>\n",
       "      <td>463</td>\n",
       "      <td>412</td>\n",
       "      <td>412</td>\n",
       "      <td>3</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGATACCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||| | | ||||||||||||||||...</td>\n",
       "      <td>Porphyromonas somerae strain KA00683 16S ribos...</td>\n",
       "      <td>KP192301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>875.0</td>\n",
       "      <td>790.260</td>\n",
       "      <td>440</td>\n",
       "      <td>439</td>\n",
       "      <td>439</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Bacterium N14-24 16S ribosomal RNA gene, parti...</td>\n",
       "      <td>AY880043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>871.0</td>\n",
       "      <td>786.653</td>\n",
       "      <td>440</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella asaccharolytica strain WAL 1855C ...</td>\n",
       "      <td>NR_116331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>815.0</td>\n",
       "      <td>736.159</td>\n",
       "      <td>440</td>\n",
       "      <td>427</td>\n",
       "      <td>427</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella sp. SIT12 partial 16S rRNA gene, ...</td>\n",
       "      <td>NR_148568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>813.0</td>\n",
       "      <td>734.356</td>\n",
       "      <td>439</td>\n",
       "      <td>426</td>\n",
       "      <td>426</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella massiliensis partial 16S rRNA gene</td>\n",
       "      <td>OX250391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>800.0</td>\n",
       "      <td>722.634</td>\n",
       "      <td>440</td>\n",
       "      <td>424</td>\n",
       "      <td>424</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella vaginalis isolate MGYG-HGUT-01563...</td>\n",
       "      <td>LT632322</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Feature ID         expect  score     bits  \\\n",
       "0    4900b14ef37cec718bf18a0d92210a03   0.000000e+00  914.0  825.426   \n",
       "1    4900b14ef37cec718bf18a0d92210a03   0.000000e+00  909.0  820.917   \n",
       "2    4900b14ef37cec718bf18a0d92210a03   0.000000e+00  895.0  808.294   \n",
       "3    4900b14ef37cec718bf18a0d92210a03  2.252820e-166  665.0  600.906   \n",
       "4    4900b14ef37cec718bf18a0d92210a03  9.579250e-165  659.0  595.496   \n",
       "..                                ...            ...    ...      ...   \n",
       "295  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  875.0  790.260   \n",
       "296  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  871.0  786.653   \n",
       "297  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  815.0  736.159   \n",
       "298  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  813.0  734.356   \n",
       "299  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  800.0  722.634   \n",
       "\n",
       "     align_length  identities  positives  gaps  \\\n",
       "0             462         460        460     0   \n",
       "1             462         459        459     0   \n",
       "2             462         455        455     0   \n",
       "3             462         411        411     1   \n",
       "4             463         412        412     3   \n",
       "..            ...         ...        ...   ...   \n",
       "295           440         439        439     0   \n",
       "296           440         438        438     0   \n",
       "297           440         427        427     0   \n",
       "298           439         426        426     0   \n",
       "299           440         424        424     0   \n",
       "\n",
       "                                                 query  \\\n",
       "0    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "1    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "2    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "3    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "4    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "..                                                 ...   \n",
       "295  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "296  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "297  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "298  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "299  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "\n",
       "                                                 sbjct  \\\n",
       "0    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "1    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "2    GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTGCC...   \n",
       "3    GACTACCAGGGTATCTAATCCTGTTTGATCCCCACGCTTTCGTGCA...   \n",
       "4    GACTACCAGGGTATCTAATCCTGTTTGATACCCACGCTTTCGTGCC...   \n",
       "..                                                 ...   \n",
       "295  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "296  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "297  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "298  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "299  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "\n",
       "                                                 match  \\\n",
       "0    |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "1    |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "2    |||||| |||||||||||||||||| ||||||||||||||||||||...   \n",
       "3    |||||| |||||||||||||||||| | ||||||||||||||||| ...   \n",
       "4    |||||| |||||||||||||||||| | | ||||||||||||||||...   \n",
       "..                                                 ...   \n",
       "295  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "296  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "297  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "298  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "299  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "\n",
       "                                              taxonomy  accession  \n",
       "0    Porphyromonas bennonis strain WAL 1926C 16S ri...  NR_044491  \n",
       "1    Porphyromonas bennonis gene for 16S ribosomal ...  NR_113228  \n",
       "2    Porphyromonas bennonis strain A7 16S ribosomal...   OP269714  \n",
       "3    Porphyromonadaceae bacterium FC4 partial 16S r...   LN881610  \n",
       "4    Porphyromonas somerae strain KA00683 16S ribos...   KP192301  \n",
       "..                                                 ...        ...  \n",
       "295  Bacterium N14-24 16S ribosomal RNA gene, parti...   AY880043  \n",
       "296  Murdochiella asaccharolytica strain WAL 1855C ...  NR_116331  \n",
       "297  Murdochiella sp. SIT12 partial 16S rRNA gene, ...  NR_148568  \n",
       "298    Murdochiella massiliensis partial 16S rRNA gene   OX250391  \n",
       "299  Murdochiella vaginalis isolate MGYG-HGUT-01563...   LT632322  \n",
       "\n",
       "[300 rows x 13 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import time\n",
    "import pandas as pd\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from qiime2 import Artifact\n",
    "\n",
    "def extract_lineage(taxon):\n",
    "    # Extract and clean lineage information\n",
    "    lineage = taxon.split(';')\n",
    "    cleaned_lineage = [tax.split('__')[1] for tax in lineage if '__' in tax and tax.split('__')[1] != '']\n",
    "    return cleaned_lineage\n",
    "\n",
    "def blast_classifier(rep_seqs_path, taxonomy_path, top_n=10, expect_threshold=0.01):\n",
    "    # Load in files\n",
    "    rep_seqs_qza = Artifact.load(rep_seqs_path)\n",
    "    taxonomy_qza = Artifact.load(taxonomy_path)\n",
    "\n",
    "    # Convert taxonomy to df\n",
    "    taxonomy_df = taxonomy_qza.view(pd.DataFrame)\n",
    "\n",
    "    # Convert rep seqs to df\n",
    "    rep_seqs = rep_seqs_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "    rep_seqs.index.name = 'Feature ID'\n",
    "\n",
    "    # Merge dfs on Feature ID\n",
    "    merged = pd.merge(rep_seqs, taxonomy_df, left_index=True, right_index=True)\n",
    "    merged.reset_index(inplace=True)\n",
    "\n",
    "    # Filter for entries without species assignment\n",
    "    unassigned = merged[merged['Taxon'].str.endswith('s__')]\n",
    "   \n",
    "    # Empty list for BLAST results\n",
    "    blast_results = []\n",
    "\n",
    "    # Perform BLAST for each unassigned sequence\n",
    "    for _, row in unassigned.iterrows():\n",
    "        feature_id = row['Feature ID']\n",
    "        sequence = row['Sequence']\n",
    "        lineage = extract_lineage(row['Taxon'])\n",
    "        \n",
    "        # Create a query with the lineage information\n",
    "        lineage_query = \" AND \".join(f\"{taxon}[ORGN]\" for taxon in lineage if taxon)\n",
    "        query = f\"{sequence} AND ({lineage_query})\"\n",
    "        \n",
    "        # BLAST search\n",
    "        result_handle = NCBIWWW.qblast('blastn', 'nt', query, entrez_query='all [filter] NOT(environmental samples[organism] OR metagenomes[orgn])')\n",
    "        blast_record = NCBIXML.read(result_handle)\n",
    "        \n",
    "        hit_count = 0\n",
    "        if blast_record.alignments:\n",
    "            for alignment in blast_record.alignments:\n",
    "                for hsp in alignment.hsps:\n",
    "                    if hit_count >= top_n:\n",
    "                        break\n",
    "                    if hsp.expect <= expect_threshold:  # Check if expect value is below the threshold\n",
    "                        extracted_title = alignment.title.rsplit('|', 1)[-1].strip()\n",
    "                        blast_results.append({\n",
    "                            'Feature ID': feature_id,\n",
    "                            'expect': hsp.expect,\n",
    "                            'score': hsp.score,\n",
    "                            'bits': hsp.bits, \n",
    "                            'align_length': hsp.align_length,\n",
    "                            'identities': hsp.identities, \n",
    "                            'positives': hsp.positives,\n",
    "                            'gaps': hsp.gaps,\n",
    "                            'query': hsp.query,\n",
    "                            'sbjct': hsp.sbjct,\n",
    "                            'match': hsp.match,\n",
    "                            'taxonomy': extracted_title,\n",
    "                            'accession': alignment.accession\n",
    "                        })\n",
    "                        hit_count += 1\n",
    "                if hit_count >= top_n:\n",
    "                    break\n",
    "        time.sleep(1)\n",
    "\n",
    "    blast_results_df = pd.DataFrame(blast_results)\n",
    "\n",
    "    return blast_results_df\n",
    "\n",
    "# File paths\n",
    "rep_seqs_path = 'qiime2_blast_data/SRA1038019-rep_seqs.qza'\n",
    "taxonomy_path = 'qiime2_blast_data/SRA1038019-taxonomy.qza'\n",
    "\n",
    "# Perform BLAST classification and get results\n",
    "results_df = blast_classifier(rep_seqs_path, taxonomy_path, expect_threshold=0.05) \n",
    "\n",
    "# Display the results\n",
    "results_df\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature ID</th>\n",
       "      <th>expect</th>\n",
       "      <th>score</th>\n",
       "      <th>bits</th>\n",
       "      <th>align_length</th>\n",
       "      <th>identities</th>\n",
       "      <th>positives</th>\n",
       "      <th>gaps</th>\n",
       "      <th>query</th>\n",
       "      <th>sbjct</th>\n",
       "      <th>match</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>accession</th>\n",
       "      <th>taxid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>914.0</td>\n",
       "      <td>825.426</td>\n",
       "      <td>462</td>\n",
       "      <td>460</td>\n",
       "      <td>460</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas bennonis strain WAL 1926C 16S ri...</td>\n",
       "      <td>NR_044491</td>\n",
       "      <td>501496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>909.0</td>\n",
       "      <td>820.917</td>\n",
       "      <td>462</td>\n",
       "      <td>459</td>\n",
       "      <td>459</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas bennonis gene for 16S ribosomal ...</td>\n",
       "      <td>NR_113228</td>\n",
       "      <td>501496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>895.0</td>\n",
       "      <td>808.294</td>\n",
       "      <td>462</td>\n",
       "      <td>455</td>\n",
       "      <td>455</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||| ||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas bennonis strain A7 16S ribosomal...</td>\n",
       "      <td>OP269714</td>\n",
       "      <td>501496</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>2.252820e-166</td>\n",
       "      <td>665.0</td>\n",
       "      <td>600.906</td>\n",
       "      <td>462</td>\n",
       "      <td>411</td>\n",
       "      <td>411</td>\n",
       "      <td>1</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGATCCCCACGCTTTCGTGCA...</td>\n",
       "      <td>|||||| |||||||||||||||||| | ||||||||||||||||| ...</td>\n",
       "      <td>Porphyromonadaceae bacterium FC4 partial 16S r...</td>\n",
       "      <td>LN881610</td>\n",
       "      <td>1720317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>9.579250e-165</td>\n",
       "      <td>659.0</td>\n",
       "      <td>595.496</td>\n",
       "      <td>463</td>\n",
       "      <td>412</td>\n",
       "      <td>412</td>\n",
       "      <td>3</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGATACCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||| | | ||||||||||||||||...</td>\n",
       "      <td>Porphyromonas somerae strain KA00683 16S ribos...</td>\n",
       "      <td>KP192301</td>\n",
       "      <td>322095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>875.0</td>\n",
       "      <td>790.260</td>\n",
       "      <td>440</td>\n",
       "      <td>439</td>\n",
       "      <td>439</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Bacterium N14-24 16S ribosomal RNA gene, parti...</td>\n",
       "      <td>AY880043</td>\n",
       "      <td>310935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>871.0</td>\n",
       "      <td>786.653</td>\n",
       "      <td>440</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella asaccharolytica strain WAL 1855C ...</td>\n",
       "      <td>NR_116331</td>\n",
       "      <td>507844</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>815.0</td>\n",
       "      <td>736.159</td>\n",
       "      <td>440</td>\n",
       "      <td>427</td>\n",
       "      <td>427</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella sp. SIT12 partial 16S rRNA gene, ...</td>\n",
       "      <td>NR_148568</td>\n",
       "      <td>1673723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>813.0</td>\n",
       "      <td>734.356</td>\n",
       "      <td>439</td>\n",
       "      <td>426</td>\n",
       "      <td>426</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella massiliensis partial 16S rRNA gene</td>\n",
       "      <td>OX250391</td>\n",
       "      <td>1673723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>800.0</td>\n",
       "      <td>722.634</td>\n",
       "      <td>440</td>\n",
       "      <td>424</td>\n",
       "      <td>424</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella vaginalis isolate MGYG-HGUT-01563...</td>\n",
       "      <td>LT632322</td>\n",
       "      <td>1852373</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Feature ID         expect  score     bits  \\\n",
       "0    4900b14ef37cec718bf18a0d92210a03   0.000000e+00  914.0  825.426   \n",
       "1    4900b14ef37cec718bf18a0d92210a03   0.000000e+00  909.0  820.917   \n",
       "2    4900b14ef37cec718bf18a0d92210a03   0.000000e+00  895.0  808.294   \n",
       "3    4900b14ef37cec718bf18a0d92210a03  2.252820e-166  665.0  600.906   \n",
       "4    4900b14ef37cec718bf18a0d92210a03  9.579250e-165  659.0  595.496   \n",
       "..                                ...            ...    ...      ...   \n",
       "295  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  875.0  790.260   \n",
       "296  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  871.0  786.653   \n",
       "297  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  815.0  736.159   \n",
       "298  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  813.0  734.356   \n",
       "299  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  800.0  722.634   \n",
       "\n",
       "     align_length  identities  positives  gaps  \\\n",
       "0             462         460        460     0   \n",
       "1             462         459        459     0   \n",
       "2             462         455        455     0   \n",
       "3             462         411        411     1   \n",
       "4             463         412        412     3   \n",
       "..            ...         ...        ...   ...   \n",
       "295           440         439        439     0   \n",
       "296           440         438        438     0   \n",
       "297           440         427        427     0   \n",
       "298           439         426        426     0   \n",
       "299           440         424        424     0   \n",
       "\n",
       "                                                 query  \\\n",
       "0    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "1    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "2    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "3    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "4    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "..                                                 ...   \n",
       "295  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "296  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "297  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "298  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "299  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "\n",
       "                                                 sbjct  \\\n",
       "0    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "1    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "2    GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTGCC...   \n",
       "3    GACTACCAGGGTATCTAATCCTGTTTGATCCCCACGCTTTCGTGCA...   \n",
       "4    GACTACCAGGGTATCTAATCCTGTTTGATACCCACGCTTTCGTGCC...   \n",
       "..                                                 ...   \n",
       "295  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "296  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "297  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "298  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "299  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "\n",
       "                                                 match  \\\n",
       "0    |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "1    |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "2    |||||| |||||||||||||||||| ||||||||||||||||||||...   \n",
       "3    |||||| |||||||||||||||||| | ||||||||||||||||| ...   \n",
       "4    |||||| |||||||||||||||||| | | ||||||||||||||||...   \n",
       "..                                                 ...   \n",
       "295  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "296  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "297  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "298  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "299  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "\n",
       "                                              taxonomy  accession    taxid  \n",
       "0    Porphyromonas bennonis strain WAL 1926C 16S ri...  NR_044491   501496  \n",
       "1    Porphyromonas bennonis gene for 16S ribosomal ...  NR_113228   501496  \n",
       "2    Porphyromonas bennonis strain A7 16S ribosomal...   OP269714   501496  \n",
       "3    Porphyromonadaceae bacterium FC4 partial 16S r...   LN881610  1720317  \n",
       "4    Porphyromonas somerae strain KA00683 16S ribos...   KP192301   322095  \n",
       "..                                                 ...        ...      ...  \n",
       "295  Bacterium N14-24 16S ribosomal RNA gene, parti...   AY880043   310935  \n",
       "296  Murdochiella asaccharolytica strain WAL 1855C ...  NR_116331   507844  \n",
       "297  Murdochiella sp. SIT12 partial 16S rRNA gene, ...  NR_148568  1673723  \n",
       "298    Murdochiella massiliensis partial 16S rRNA gene   OX250391  1673723  \n",
       "299  Murdochiella vaginalis isolate MGYG-HGUT-01563...   LT632322  1852373  \n",
       "\n",
       "[300 rows x 14 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "def accession_to_taxid(accessions):\n",
    "    taxid_dict = {}\n",
    "    for accession in accessions:\n",
    "        # Run the bash command\n",
    "        bash_command = f\"esummary -db nuccore -id {accession} | xtract -pattern DocumentSummary -element Caption,TaxId\"\n",
    "        process = subprocess.Popen(bash_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "        stdout, stderr = process.communicate()\n",
    "\n",
    "        # Check for errors\n",
    "        if process.returncode != 0:\n",
    "            print(f\"Error running command for accession {accession}: {stderr.decode('utf-8')}\")\n",
    "            continue\n",
    "\n",
    "        # Parse the output to get the taxid\n",
    "        output = stdout.decode('utf-8').strip()\n",
    "        if output:\n",
    "            accession_id, taxid = output.split()\n",
    "            taxid_dict[accession_id] = taxid\n",
    "\n",
    "    return taxid_dict\n",
    "\n",
    "# Example usage:\n",
    "accessions = results_df['accession']  # Add more accession numbers as needed\n",
    "taxid_dict = accession_to_taxid(accessions)\n",
    "results_df['taxid'] = results_df['accession'].map(taxid_dict)\n",
    "results_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature ID</th>\n",
       "      <th>expect</th>\n",
       "      <th>score</th>\n",
       "      <th>bits</th>\n",
       "      <th>align_length</th>\n",
       "      <th>identities</th>\n",
       "      <th>positives</th>\n",
       "      <th>gaps</th>\n",
       "      <th>query</th>\n",
       "      <th>sbjct</th>\n",
       "      <th>match</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>accession</th>\n",
       "      <th>taxid</th>\n",
       "      <th>lineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>914.0</td>\n",
       "      <td>825.426</td>\n",
       "      <td>462</td>\n",
       "      <td>460</td>\n",
       "      <td>460</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas bennonis strain WAL 1926C 16S ri...</td>\n",
       "      <td>NR_044491</td>\n",
       "      <td>501496</td>\n",
       "      <td>cellular organisms;Bacteria;FCB group;Bacteroi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>909.0</td>\n",
       "      <td>820.917</td>\n",
       "      <td>462</td>\n",
       "      <td>459</td>\n",
       "      <td>459</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas bennonis gene for 16S ribosomal ...</td>\n",
       "      <td>NR_113228</td>\n",
       "      <td>501496</td>\n",
       "      <td>cellular organisms;Bacteria;FCB group;Bacteroi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>895.0</td>\n",
       "      <td>808.294</td>\n",
       "      <td>462</td>\n",
       "      <td>455</td>\n",
       "      <td>455</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||| ||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas bennonis strain A7 16S ribosomal...</td>\n",
       "      <td>OP269714</td>\n",
       "      <td>501496</td>\n",
       "      <td>cellular organisms;Bacteria;FCB group;Bacteroi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>2.252820e-166</td>\n",
       "      <td>665.0</td>\n",
       "      <td>600.906</td>\n",
       "      <td>462</td>\n",
       "      <td>411</td>\n",
       "      <td>411</td>\n",
       "      <td>1</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGATCCCCACGCTTTCGTGCA...</td>\n",
       "      <td>|||||| |||||||||||||||||| | ||||||||||||||||| ...</td>\n",
       "      <td>Porphyromonadaceae bacterium FC4 partial 16S r...</td>\n",
       "      <td>LN881610</td>\n",
       "      <td>1720317</td>\n",
       "      <td>cellular organisms;Bacteria;FCB group;Bacteroi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>9.579250e-165</td>\n",
       "      <td>659.0</td>\n",
       "      <td>595.496</td>\n",
       "      <td>463</td>\n",
       "      <td>412</td>\n",
       "      <td>412</td>\n",
       "      <td>3</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGATACCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||| | | ||||||||||||||||...</td>\n",
       "      <td>Porphyromonas somerae strain KA00683 16S ribos...</td>\n",
       "      <td>KP192301</td>\n",
       "      <td>322095</td>\n",
       "      <td>cellular organisms;Bacteria;FCB group;Bacteroi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>875.0</td>\n",
       "      <td>790.260</td>\n",
       "      <td>440</td>\n",
       "      <td>439</td>\n",
       "      <td>439</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Bacterium N14-24 16S ribosomal RNA gene, parti...</td>\n",
       "      <td>AY880043</td>\n",
       "      <td>310935</td>\n",
       "      <td>cellular organisms;Bacteria;unclassified Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>871.0</td>\n",
       "      <td>786.653</td>\n",
       "      <td>440</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella asaccharolytica strain WAL 1855C ...</td>\n",
       "      <td>NR_116331</td>\n",
       "      <td>507844</td>\n",
       "      <td>cellular organisms;Bacteria;Terrabacteria grou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>815.0</td>\n",
       "      <td>736.159</td>\n",
       "      <td>440</td>\n",
       "      <td>427</td>\n",
       "      <td>427</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella sp. SIT12 partial 16S rRNA gene, ...</td>\n",
       "      <td>NR_148568</td>\n",
       "      <td>1673723</td>\n",
       "      <td>cellular organisms;Bacteria;Terrabacteria grou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>813.0</td>\n",
       "      <td>734.356</td>\n",
       "      <td>439</td>\n",
       "      <td>426</td>\n",
       "      <td>426</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella massiliensis partial 16S rRNA gene</td>\n",
       "      <td>OX250391</td>\n",
       "      <td>1673723</td>\n",
       "      <td>cellular organisms;Bacteria;Terrabacteria grou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>800.0</td>\n",
       "      <td>722.634</td>\n",
       "      <td>440</td>\n",
       "      <td>424</td>\n",
       "      <td>424</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Murdochiella vaginalis isolate MGYG-HGUT-01563...</td>\n",
       "      <td>LT632322</td>\n",
       "      <td>1852373</td>\n",
       "      <td>cellular organisms;Bacteria;Terrabacteria grou...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Feature ID         expect  score     bits  \\\n",
       "0    4900b14ef37cec718bf18a0d92210a03   0.000000e+00  914.0  825.426   \n",
       "1    4900b14ef37cec718bf18a0d92210a03   0.000000e+00  909.0  820.917   \n",
       "2    4900b14ef37cec718bf18a0d92210a03   0.000000e+00  895.0  808.294   \n",
       "3    4900b14ef37cec718bf18a0d92210a03  2.252820e-166  665.0  600.906   \n",
       "4    4900b14ef37cec718bf18a0d92210a03  9.579250e-165  659.0  595.496   \n",
       "..                                ...            ...    ...      ...   \n",
       "295  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  875.0  790.260   \n",
       "296  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  871.0  786.653   \n",
       "297  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  815.0  736.159   \n",
       "298  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  813.0  734.356   \n",
       "299  9186e853f994fd8285d68ccd223cd34d   0.000000e+00  800.0  722.634   \n",
       "\n",
       "     align_length  identities  positives  gaps  \\\n",
       "0             462         460        460     0   \n",
       "1             462         459        459     0   \n",
       "2             462         455        455     0   \n",
       "3             462         411        411     1   \n",
       "4             463         412        412     3   \n",
       "..            ...         ...        ...   ...   \n",
       "295           440         439        439     0   \n",
       "296           440         438        438     0   \n",
       "297           440         427        427     0   \n",
       "298           439         426        426     0   \n",
       "299           440         424        424     0   \n",
       "\n",
       "                                                 query  \\\n",
       "0    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "1    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "2    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "3    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "4    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "..                                                 ...   \n",
       "295  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "296  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "297  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "298  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "299  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "\n",
       "                                                 sbjct  \\\n",
       "0    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "1    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "2    GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTGCC...   \n",
       "3    GACTACCAGGGTATCTAATCCTGTTTGATCCCCACGCTTTCGTGCA...   \n",
       "4    GACTACCAGGGTATCTAATCCTGTTTGATACCCACGCTTTCGTGCC...   \n",
       "..                                                 ...   \n",
       "295  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "296  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "297  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "298  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "299  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "\n",
       "                                                 match  \\\n",
       "0    |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "1    |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "2    |||||| |||||||||||||||||| ||||||||||||||||||||...   \n",
       "3    |||||| |||||||||||||||||| | ||||||||||||||||| ...   \n",
       "4    |||||| |||||||||||||||||| | | ||||||||||||||||...   \n",
       "..                                                 ...   \n",
       "295  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "296  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "297  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "298  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "299  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "\n",
       "                                              taxonomy  accession    taxid  \\\n",
       "0    Porphyromonas bennonis strain WAL 1926C 16S ri...  NR_044491   501496   \n",
       "1    Porphyromonas bennonis gene for 16S ribosomal ...  NR_113228   501496   \n",
       "2    Porphyromonas bennonis strain A7 16S ribosomal...   OP269714   501496   \n",
       "3    Porphyromonadaceae bacterium FC4 partial 16S r...   LN881610  1720317   \n",
       "4    Porphyromonas somerae strain KA00683 16S ribos...   KP192301   322095   \n",
       "..                                                 ...        ...      ...   \n",
       "295  Bacterium N14-24 16S ribosomal RNA gene, parti...   AY880043   310935   \n",
       "296  Murdochiella asaccharolytica strain WAL 1855C ...  NR_116331   507844   \n",
       "297  Murdochiella sp. SIT12 partial 16S rRNA gene, ...  NR_148568  1673723   \n",
       "298    Murdochiella massiliensis partial 16S rRNA gene   OX250391  1673723   \n",
       "299  Murdochiella vaginalis isolate MGYG-HGUT-01563...   LT632322  1852373   \n",
       "\n",
       "                                               lineage  \n",
       "0    cellular organisms;Bacteria;FCB group;Bacteroi...  \n",
       "1    cellular organisms;Bacteria;FCB group;Bacteroi...  \n",
       "2    cellular organisms;Bacteria;FCB group;Bacteroi...  \n",
       "3    cellular organisms;Bacteria;FCB group;Bacteroi...  \n",
       "4    cellular organisms;Bacteria;FCB group;Bacteroi...  \n",
       "..                                                 ...  \n",
       "295  cellular organisms;Bacteria;unclassified Bacteria  \n",
       "296  cellular organisms;Bacteria;Terrabacteria grou...  \n",
       "297  cellular organisms;Bacteria;Terrabacteria grou...  \n",
       "298  cellular organisms;Bacteria;Terrabacteria grou...  \n",
       "299  cellular organisms;Bacteria;Terrabacteria grou...  \n",
       "\n",
       "[300 rows x 15 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_lineage_from_taxid(taxid):\n",
    "    # Run the bash command to fetch lineage information\n",
    "    bash_command = f\"efetch -db taxonomy -id {taxid} -format xml | xtract -pattern Taxon -block LineageEx -sep ';' -element ScientificName\"\n",
    "    process = subprocess.Popen(bash_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    stdout, stderr = process.communicate()\n",
    "\n",
    "    # Check for errors\n",
    "    if process.returncode != 0:\n",
    "        print(f\"Error running command for taxid {taxid}: {stderr.decode('utf-8')}\")\n",
    "        return None\n",
    "\n",
    "    # Parse the output to get the lineage\n",
    "    lineage = stdout.decode('utf-8').strip()\n",
    "    return lineage\n",
    "\n",
    "# Example usage with a single taxid\n",
    "taxid = results_df['taxid']  # Example taxid\n",
    "results_df['lineage'] = results_df['taxid'].apply(get_lineage_from_taxid)\n",
    "results_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature ID</th>\n",
       "      <th>expect</th>\n",
       "      <th>score</th>\n",
       "      <th>bits</th>\n",
       "      <th>align_length</th>\n",
       "      <th>identities</th>\n",
       "      <th>positives</th>\n",
       "      <th>gaps</th>\n",
       "      <th>query</th>\n",
       "      <th>sbjct</th>\n",
       "      <th>match</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>accession</th>\n",
       "      <th>taxid</th>\n",
       "      <th>lineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>7.364520e-160</td>\n",
       "      <td>641.0</td>\n",
       "      <td>579.266</td>\n",
       "      <td>466</td>\n",
       "      <td>411</td>\n",
       "      <td>411</td>\n",
       "      <td>9</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGATACCCACGCTTTCGTGCA...</td>\n",
       "      <td>|||||| |||||||||||||||||| | | ||||||||||||||| ...</td>\n",
       "      <td>Porphyromonas levii clone 407-6 16S ribosomal ...</td>\n",
       "      <td>FJ822455</td>\n",
       "      <td>28114</td>\n",
       "      <td>cellular organisms;Bacteria;FCB group;Bacteroi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>29841a48cdb3345f69e10a0bbb13f91f</td>\n",
       "      <td>7.458290e-166</td>\n",
       "      <td>663.0</td>\n",
       "      <td>599.103</td>\n",
       "      <td>444</td>\n",
       "      <td>399</td>\n",
       "      <td>399</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGCGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGCGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Bacterium XPD2007 16S ribosomal RNA gene, part...</td>\n",
       "      <td>KF698493</td>\n",
       "      <td>1452267</td>\n",
       "      <td>cellular organisms;Bacteria;unclassified Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>bc2eec26f20121d03c3ff6790b97fcbb</td>\n",
       "      <td>4.323480e-162</td>\n",
       "      <td>649.0</td>\n",
       "      <td>586.479</td>\n",
       "      <td>465</td>\n",
       "      <td>412</td>\n",
       "      <td>412</td>\n",
       "      <td>9</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...</td>\n",
       "      <td>||||||||||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas levii clone 407-6 16S ribosomal ...</td>\n",
       "      <td>FJ822455</td>\n",
       "      <td>28114</td>\n",
       "      <td>cellular organisms;Bacteria;FCB group;Bacteroi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>12fe9bc4b3c216f4b6bb7163e226582b</td>\n",
       "      <td>5.267070e-161</td>\n",
       "      <td>644.0</td>\n",
       "      <td>581.971</td>\n",
       "      <td>465</td>\n",
       "      <td>411</td>\n",
       "      <td>411</td>\n",
       "      <td>9</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...</td>\n",
       "      <td>||||||||||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas levii clone 407-6 16S ribosomal ...</td>\n",
       "      <td>FJ822455</td>\n",
       "      <td>28114</td>\n",
       "      <td>cellular organisms;Bacteria;FCB group;Bacteroi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>3862cdba0e52f6b9e73fb73aabcee8fc</td>\n",
       "      <td>9.086060e-165</td>\n",
       "      <td>659.0</td>\n",
       "      <td>595.496</td>\n",
       "      <td>442</td>\n",
       "      <td>397</td>\n",
       "      <td>397</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGCGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGCGCC...</td>\n",
       "      <td>||||||||||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Bacterium XPD2007 16S ribosomal RNA gene, part...</td>\n",
       "      <td>KF698493</td>\n",
       "      <td>1452267</td>\n",
       "      <td>cellular organisms;Bacteria;unclassified Bacteria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>95fcb2d33bd2a46e98e51e5cb5e4d924</td>\n",
       "      <td>2.239620e-159</td>\n",
       "      <td>639.0</td>\n",
       "      <td>577.463</td>\n",
       "      <td>465</td>\n",
       "      <td>410</td>\n",
       "      <td>410</td>\n",
       "      <td>9</td>\n",
       "      <td>CCTACGGGAGGCTGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...</td>\n",
       "      <td>CCTACGGGAGGCAGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...</td>\n",
       "      <td>|||||||||||| |||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Porphyromonas levii clone 407-6 16S ribosomal ...</td>\n",
       "      <td>FJ822455</td>\n",
       "      <td>28114</td>\n",
       "      <td>cellular organisms;Bacteria;FCB group;Bacteroi...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Feature ID         expect  score     bits  \\\n",
       "9    4900b14ef37cec718bf18a0d92210a03  7.364520e-160  641.0  579.266   \n",
       "38   29841a48cdb3345f69e10a0bbb13f91f  7.458290e-166  663.0  599.103   \n",
       "69   bc2eec26f20121d03c3ff6790b97fcbb  4.323480e-162  649.0  586.479   \n",
       "139  12fe9bc4b3c216f4b6bb7163e226582b  5.267070e-161  644.0  581.971   \n",
       "219  3862cdba0e52f6b9e73fb73aabcee8fc  9.086060e-165  659.0  595.496   \n",
       "289  95fcb2d33bd2a46e98e51e5cb5e4d924  2.239620e-159  639.0  577.463   \n",
       "\n",
       "     align_length  identities  positives  gaps  \\\n",
       "9             466         411        411     9   \n",
       "38            444         399        399     0   \n",
       "69            465         412        412     9   \n",
       "139           465         411        411     9   \n",
       "219           442         397        397     0   \n",
       "289           465         410        410     9   \n",
       "\n",
       "                                                 query  \\\n",
       "9    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "38   GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGCGCC...   \n",
       "69   CCTACGGGAGGCAGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...   \n",
       "139  CCTACGGGAGGCAGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...   \n",
       "219  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGCGCC...   \n",
       "289  CCTACGGGAGGCTGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...   \n",
       "\n",
       "                                                 sbjct  \\\n",
       "9    GACTACCAGGGTATCTAATCCTGTTTGATACCCACGCTTTCGTGCA...   \n",
       "38   GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGCGCC...   \n",
       "69   CCTACGGGAGGCAGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...   \n",
       "139  CCTACGGGAGGCAGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...   \n",
       "219  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGCGCC...   \n",
       "289  CCTACGGGAGGCAGCAGTGAGGAATATTGGTCAATGGAGGAAACTC...   \n",
       "\n",
       "                                                 match  \\\n",
       "9    |||||| |||||||||||||||||| | | ||||||||||||||| ...   \n",
       "38   |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "69   ||||||||||||||||||||||||||||||||||||||||||||||...   \n",
       "139  ||||||||||||||||||||||||||||||||||||||||||||||...   \n",
       "219  ||||||||||||||||||||||||||||||||||||||||||||||...   \n",
       "289  |||||||||||| |||||||||||||||||||||||||||||||||...   \n",
       "\n",
       "                                              taxonomy accession    taxid  \\\n",
       "9    Porphyromonas levii clone 407-6 16S ribosomal ...  FJ822455    28114   \n",
       "38   Bacterium XPD2007 16S ribosomal RNA gene, part...  KF698493  1452267   \n",
       "69   Porphyromonas levii clone 407-6 16S ribosomal ...  FJ822455    28114   \n",
       "139  Porphyromonas levii clone 407-6 16S ribosomal ...  FJ822455    28114   \n",
       "219  Bacterium XPD2007 16S ribosomal RNA gene, part...  KF698493  1452267   \n",
       "289  Porphyromonas levii clone 407-6 16S ribosomal ...  FJ822455    28114   \n",
       "\n",
       "                                               lineage  \n",
       "9    cellular organisms;Bacteria;FCB group;Bacteroi...  \n",
       "38   cellular organisms;Bacteria;unclassified Bacteria  \n",
       "69   cellular organisms;Bacteria;FCB group;Bacteroi...  \n",
       "139  cellular organisms;Bacteria;FCB group;Bacteroi...  \n",
       "219  cellular organisms;Bacteria;unclassified Bacteria  \n",
       "289  cellular organisms;Bacteria;FCB group;Bacteroi...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def filter_by_highest_expect_value(results_df):\n",
    "    # Group by Feature ID to process each OTU separately\n",
    "    filtered_results = []\n",
    "\n",
    "    for feature_id, group in results_df.groupby('Feature ID'):\n",
    "        # Sort the group by expect value in descending order (highest expect value first)\n",
    "        sorted_group = group.sort_values(by='expect', ascending=False)\n",
    "        \n",
    "        # Get the top expect value\n",
    "        top_expect_value = sorted_group['expect'].max()\n",
    "        \n",
    "        # Filter the group to only include rows with the top expect value\n",
    "        top_hits = sorted_group[sorted_group['expect'] == top_expect_value]\n",
    "\n",
    "        if len(top_hits) == 1:\n",
    "            # If there's only one hit with the top expect value, select it\n",
    "            filtered_results.append(top_hits.iloc[0])\n",
    "        else:\n",
    "            # If there's a tie, check if all top hits have the same species (taxonomy)\n",
    "            species_names = top_hits['taxonomy'].apply(lambda x: x.split()[-1])\n",
    "            if species_names.nunique() == 1:\n",
    "                # If all species names are the same, select one (e.g., the first one)\n",
    "                filtered_results.append(top_hits.iloc[0])\n",
    "            else:\n",
    "                # If species names differ, handle the case as needed (e.g., select the first one or mark as unresolved)\n",
    "                continue\n",
    "    # Convert the results to a DataFrame\n",
    "    filtered_results_df = pd.DataFrame(filtered_results)\n",
    "\n",
    "    return filtered_results_df\n",
    "\n",
    "\n",
    "results_df['original_index'] = results_df.index\n",
    "filtered_results_df = filter_by_highest_expect_value(results_df)\n",
    "filtered_results_df = filtered_results_df.sort_values(by='original_index')\n",
    "filtered_results_df = filtered_results_df.drop(columns=['original_index'])\n",
    "filtered_results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-08-05 02:03:33--  https://data.qiime2.org/2024.5/tutorials/training-feature-classifiers/85_otus.fasta\n",
      "Resolving data.qiime2.org (data.qiime2.org)... 54.200.1.12\n",
      "Connecting to data.qiime2.org (data.qiime2.org)|54.200.1.12|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 FOUND\n",
      "Location: https://s3-us-west-2.amazonaws.com/qiime2-data/2024.5/tutorials/training-feature-classifiers/85_otus.fasta [following]\n",
      "--2024-08-05 02:03:34--  https://s3-us-west-2.amazonaws.com/qiime2-data/2024.5/tutorials/training-feature-classifiers/85_otus.fasta\n",
      "Resolving s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)... 52.92.210.16, 52.92.242.24, 52.92.233.144, ...\n",
      "Connecting to s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)|52.92.210.16|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 7324324 (7.0M) [binary/octet-stream]\n",
      "Saving to: ‘85_otus.fasta’\n",
      "\n",
      "85_otus.fasta       100%[===================>]   6.98M  1.58MB/s    in 6.0s    \n",
      "\n",
      "2024-08-05 02:03:40 (1.17 MB/s) - ‘85_otus.fasta’ saved [7324324/7324324]\n",
      "\n",
      "--2024-08-05 02:03:40--  https://data.qiime2.org/2024.5/tutorials/training-feature-classifiers/85_otu_taxonomy.txt\n",
      "Resolving data.qiime2.org (data.qiime2.org)... 54.200.1.12\n",
      "Connecting to data.qiime2.org (data.qiime2.org)|54.200.1.12|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 FOUND\n",
      "Location: https://s3-us-west-2.amazonaws.com/qiime2-data/2024.5/tutorials/training-feature-classifiers/85_otu_taxonomy.txt [following]\n",
      "--2024-08-05 02:03:41--  https://s3-us-west-2.amazonaws.com/qiime2-data/2024.5/tutorials/training-feature-classifiers/85_otu_taxonomy.txt\n",
      "Resolving s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)... 52.92.209.40, 52.218.236.160, 52.92.229.224, ...\n",
      "Connecting to s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)|52.92.209.40|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 452277 (442K) [text/plain]\n",
      "Saving to: ‘85_otu_taxonomy.txt’\n",
      "\n",
      "85_otu_taxonomy.txt 100%[===================>] 441.68K  2.41MB/s    in 0.2s    \n",
      "\n",
      "2024-08-05 02:03:41 (2.41 MB/s) - ‘85_otu_taxonomy.txt’ saved [452277/452277]\n",
      "\n",
      "\u001b[32mImported 85_otus.fasta as DNASequencesDirectoryFormat to 85_otus.qza\u001b[0m\n",
      "\u001b[32mImported 85_otu_taxonomy.txt as HeaderlessTSVTaxonomyFormat to 85_otu_taxonomy.qza\u001b[0m\n",
      "\u001b[32mUUID\u001b[0m:        e53a995e-0eae-4190-bc03-9c713af3c817\n",
      "\u001b[32mType\u001b[0m:        FeatureData[Sequence]\n",
      "\u001b[32mData format\u001b[0m: DNASequencesDirectoryFormat\n",
      "\u001b[32mUUID\u001b[0m:        3c5c7039-a4ee-4c8e-a0d0-5cd7c8e01311\n",
      "\u001b[32mType\u001b[0m:        FeatureData[Taxonomy]\n",
      "\u001b[32mData format\u001b[0m: TSVTaxonomyDirectoryFormat\n"
     ]
    }
   ],
   "source": [
    "#downloading reference sequences and taxonomy\n",
    "!wget -O \"85_otus.fasta\" \"https://data.qiime2.org/2024.5/tutorials/training-feature-classifiers/85_otus.fasta\"\n",
    "!wget -O \"85_otu_taxonomy.txt\" \"https://data.qiime2.org/2024.5/tutorials/training-feature-classifiers/85_otu_taxonomy.txt\"\n",
    "\n",
    "#importing into QIIME 2 artifact\n",
    "!qiime tools import \\\n",
    "  --type 'FeatureData[Sequence]' \\\n",
    "  --input-path 85_otus.fasta \\\n",
    "  --output-path 85_otus.qza\n",
    "\n",
    "#mport the downloaded taxonomy into a QIIME 2 artifact\n",
    "!qiime tools import \\\n",
    "  --type 'FeatureData[Taxonomy]' \\\n",
    "  --input-format HeaderlessTSVTaxonomyFormat \\\n",
    "  --input-path 85_otu_taxonomy.txt \\\n",
    "  --output-path 85_otu_taxonomy.qza\n",
    "\n",
    "#verify the imports by viewing the artifacts\n",
    "!qiime tools peek 85_otus.qza\n",
    "!qiime tools peek 85_otu_taxonomy.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureData[Sequence] to: ref-seqs.qza\u001b[0m\r\n"
     ]
    }
   ],
   "source": [
    "!qiime feature-classifier extract-reads \\\n",
    "  --i-sequences 85_otus.qza \\\n",
    "  --p-f-primer GTGCCAGCMGCCGCGGTAA \\\n",
    "  --p-r-primer GGACTACHVGGGTWTCTAAT \\\n",
    "  --p-trunc-len 120 \\\n",
    "  --p-min-length 100 \\\n",
    "  --p-max-length 400 \\\n",
    "  --o-reads ref-seqs.qza\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: \u001b[34mqiime feature-classifier fit-classifier-naive-bayes\u001b[0m \r\n",
      "           [OPTIONS]\r\n",
      "\r\n",
      "  Create a scikit-learn naive_bayes classifier for reads\r\n",
      "\r\n",
      "\u001b[1mInputs\u001b[0m:\r\n",
      "  \u001b[34m\u001b[4m--i-reference-reads\u001b[0m ARTIFACT \u001b[32mFeatureData[Sequence]\u001b[0m\r\n",
      "                                                                    \u001b[35m[required]\u001b[0m\r\n",
      "  \u001b[34m\u001b[4m--i-reference-taxonomy\u001b[0m ARTIFACT \u001b[32mFeatureData[Taxonomy]\u001b[0m\r\n",
      "                                                                    \u001b[35m[required]\u001b[0m\r\n",
      "  \u001b[34m--i-class-weight\u001b[0m ARTIFACT \u001b[32mFeatureTable[RelativeFrequency]\u001b[0m\r\n",
      "                                                                    \u001b[35m[optional]\u001b[0m\r\n",
      "\u001b[1mParameters\u001b[0m:\r\n",
      "  \u001b[34m--p-classify--alpha\u001b[0m NUMBER\r\n",
      "                                                              \u001b[35m[default: 0.001]\u001b[0m\r\n",
      "  \u001b[34m--p-classify--chunk-size\u001b[0m INTEGER\r\n",
      "                                                              \u001b[35m[default: 20000]\u001b[0m\r\n",
      "  \u001b[34m--p-classify--class-prior\u001b[0m TEXT\r\n",
      "                                                             \u001b[35m[default: 'null']\u001b[0m\r\n",
      "  \u001b[34m--p-classify--fit-prior\u001b[0m / \u001b[34m--p-no-classify--fit-prior\u001b[0m\r\n",
      "                                                              \u001b[35m[default: False]\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--alternate-sign\u001b[0m / \u001b[34m--p-no-feat-ext--alternate-sign\u001b[0m\r\n",
      "                                                              \u001b[35m[default: False]\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--analyzer\u001b[0m TEXT\r\n",
      "                                                          \u001b[35m[default: 'char_wb']\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--binary\u001b[0m / \u001b[34m--p-no-feat-ext--binary\u001b[0m\r\n",
      "                                                              \u001b[35m[default: False]\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--decode-error\u001b[0m TEXT\r\n",
      "                                                           \u001b[35m[default: 'strict']\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--encoding\u001b[0m TEXT\r\n",
      "                                                            \u001b[35m[default: 'utf-8']\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--input\u001b[0m TEXT\r\n",
      "                                                          \u001b[35m[default: 'content']\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--lowercase\u001b[0m / \u001b[34m--p-no-feat-ext--lowercase\u001b[0m\r\n",
      "                                                               \u001b[35m[default: True]\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--n-features\u001b[0m INTEGER\r\n",
      "                                                               \u001b[35m[default: 8192]\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--ngram-range\u001b[0m TEXT\r\n",
      "                                                           \u001b[35m[default: '[7, 7]']\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--norm\u001b[0m TEXT\r\n",
      "                                                               \u001b[35m[default: 'l2']\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--preprocessor\u001b[0m TEXT\r\n",
      "                                                             \u001b[35m[default: 'null']\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--stop-words\u001b[0m TEXT\r\n",
      "                                                             \u001b[35m[default: 'null']\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--strip-accents\u001b[0m TEXT\r\n",
      "                                                             \u001b[35m[default: 'null']\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--token-pattern\u001b[0m TEXT\r\n",
      "                                                \u001b[35m[default: '(?u)\\\\b\\\\w\\\\w+\\\\b']\u001b[0m\r\n",
      "  \u001b[34m--p-feat-ext--tokenizer\u001b[0m TEXT\r\n",
      "                                                             \u001b[35m[default: 'null']\u001b[0m\r\n",
      "  \u001b[34m--p-verbose\u001b[0m / \u001b[34m--p-no-verbose\u001b[0m\r\n",
      "                                                              \u001b[35m[default: False]\u001b[0m\r\n",
      "\u001b[1mOutputs\u001b[0m:\r\n",
      "  \u001b[34m\u001b[4m--o-classifier\u001b[0m ARTIFACT \u001b[32mTaxonomicClassifier\u001b[0m\r\n",
      "                                                                    \u001b[35m[required]\u001b[0m\r\n",
      "\u001b[1mMiscellaneous\u001b[0m:\r\n",
      "  \u001b[34m--output-dir\u001b[0m PATH    Output unspecified results to a directory\r\n",
      "  \u001b[34m--verbose\u001b[0m / \u001b[34m--quiet\u001b[0m  Display verbose output to stdout and/or stderr during\r\n",
      "                       execution of this action. Or silence output if\r\n",
      "                       execution is successful (silence is golden).\r\n",
      "  \u001b[34m--examples\u001b[0m           Show usage examples and exit.\r\n",
      "  \u001b[34m--citations\u001b[0m          Show citations and exit.\r\n",
      "  \u001b[34m--help\u001b[0m               Show this message and exit.\r\n",
      "\r\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\r\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--i-reference-taxonomy': 'ref-taxonomy.qza' is not a\r\n",
      "  valid filepath\u001b[0m\r\n"
     ]
    }
   ],
   "source": [
    "!qiime feature-classifier fit-classifier-naive-bayes \\\n",
    "  --i-reference-reads ref-seqs.qza \\\n",
    "  --i-reference-taxonomy ref-taxonomy.qza \\\n",
    "  --o-classifier classifier.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] File b'best_blast_hits.csv' does not exist: b'best_blast_hits.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-2e9a33afdd40>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     10\u001b[0m \u001b[0mexisting_sequences_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'Feature ID'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 12\u001b[0;31m \u001b[0mbest_hits_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'best_blast_hits.csv'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     13\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;31m#combine taxonomy\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qiime2/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36mparser_f\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, squeeze, prefix, mangle_dupe_cols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, dialect, error_bad_lines, warn_bad_lines, delim_whitespace, low_memory, memory_map, float_precision)\u001b[0m\n\u001b[1;32m    683\u001b[0m         )\n\u001b[1;32m    684\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 685\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    686\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    687\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qiime2/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    455\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    456\u001b[0m     \u001b[0;31m# Create the parser.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 457\u001b[0;31m     \u001b[0mparser\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfp_or_buf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    458\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    459\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mchunksize\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0miterator\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qiime2/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m    893\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"has_index_names\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    894\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 895\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    896\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    897\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qiime2/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m_make_engine\u001b[0;34m(self, engine)\u001b[0m\n\u001b[1;32m   1133\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"c\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1134\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"c\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1135\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1136\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1137\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"python\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qiime2/lib/python3.6/site-packages/pandas/io/parsers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m   1915\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"usecols\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0musecols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1916\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1917\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparsers\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1918\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_reader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munnamed_cols\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1919\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/parsers.pyx\u001b[0m in \u001b[0;36mpandas._libs.parsers.TextReader._setup_parser_source\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] File b'best_blast_hits.csv' does not exist: b'best_blast_hits.csv'"
     ]
    }
   ],
   "source": [
    "#load the existing database taxonomy\n",
    "existing_taxonomy_path = '85_otu_taxonomy.qza'\n",
    "existing_sequences_path = '85_otus.qza'\n",
    "existing_taxonomy_qza = Artifact.load(existing_taxonomy_path)\n",
    "existing_sequences_qza = Artifact.load(existing_sequences_path)\n",
    "\n",
    "#convert to df\n",
    "existing_taxonomy_df = existing_taxonomy_qza.view(pd.DataFrame)\n",
    "existing_sequences_df = existing_sequences_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "existing_sequences_df.index.name = 'Feature ID'\n",
    "\n",
    "best_hits_df = pd.read_csv('best_blast_hits.csv')\n",
    "\n",
    "#combine taxonomy\n",
    "best_hits_df = best_hits_df[['Feature ID', 'taxonomy']]\n",
    "best_hits_df.columns = ['Feature ID', 'Taxon']  # Rename for consistency\n",
    "combined_taxonomy_df = pd.concat([existing_taxonomy_df, best_hits_df], ignore_index=True)\n",
    "combined_taxonomy_df.to_csv('combined_taxonomy.csv', index=False)\n",
    "\n",
    "#combine sequences\n",
    "combined_sequences_df = pd.concat([existing_sequences_df, best_hits_df[['Feature ID', 'Sequence']]], ignore_index=False)\n",
    "combined_sequences_df.to_csv('combined_sequences.fasta', index=True, header=False)\n",
    "\n",
    "\n",
    "#import combined taxonomy\n",
    "!qiime tools import \\\n",
    "  --type 'FeatureData[Taxonomy]' \\\n",
    "  --input-format HeaderlessTSVTaxonomyFormat \\\n",
    "  --input-path combined_taxonomy.csv \\\n",
    "  --output-path combined_taxonomy.qza\n",
    "\n",
    "#import combined sequences\n",
    "!qiime tools import \\\n",
    "  --type 'FeatureData[Sequence]' \\\n",
    "  --input-path combined_sequences.fasta \\\n",
    "  --output-path combined_sequences.qza\n",
    "\n",
    "#extract reference reads\n",
    "!qiime feature-classifier extract-reads \\\n",
    "  --i-sequences combined_sequences.qza \\\n",
    "  --p-f-primer GTGCCAGCMGCCGCGGTAA \\\n",
    "  --p-r-primer GGACTACHVGGGTWTCTAAT \\\n",
    "  --p-trunc-len 120 \\\n",
    "  --o-reads ref-seqs.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#training naive bayes classifier\n",
    "!qiime feature-classifier fit-classifier-naive-bayes \\\n",
    "  --i-reference-reads ref-seqs.qza \\\n",
    "  --i-reference-taxonomy combined_taxonomy.qza \\\n",
    "  --o-classifier classifier.qza"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2",
   "language": "python",
   "name": "qiime2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
