{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from qiime2 import Artifact\n",
    "import pandas as pd\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "k__Bacteria; p__Bacteroidetes; c__Bacteroidia; o__Bacteroidales; f__Porphyromonadaceae; g__Porphyromonas; s__\n"
     ]
    }
   ],
   "source": [
    "rep_seqs_qza = Artifact.load('qiime2_blast_data/SRA1038019-rep_seqs.qza') \n",
    "taxonomy_qza = Artifact.load('qiime2_blast_data/SRA1038019-taxonomy.qza')\n",
    "\n",
    "rep_seqs = rep_seqs_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "rep_seqs.index.name = 'Feature ID'\n",
    "tax = taxonomy_qza.view(pd.Series)\n",
    "tax.index.name = 'Feature ID' \n",
    "\n",
    "merged = pd.merge(rep_seqs, tax, left_index=True, right_index=True)\n",
    "merged.reset_index(inplace=True)\n",
    "unassigned = merged[merged['Taxon'].str.endswith('s__')]\n",
    "unique_values_count = unassigned.iloc[:, 0].nunique()\n",
    "first_sequence = unassigned['Taxon'].iloc[0]\n",
    "print (first_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#def fetch_taxonomy_data(taxon_id): \n",
    "    #try: \n",
    "        #handle = Entrez.efetch(db=\"taxonomy\", id=taxon_id, retmode=\"xml\") \n",
    "        #record = Entrez.read(handle) \n",
    "        #return record \n",
    "    #except Exception as e: \n",
    "        #print(f\"Error fetching taxonomy data: {e}\") \n",
    "        #return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature ID</th>\n",
       "      <th>expect</th>\n",
       "      <th>score</th>\n",
       "      <th>bits</th>\n",
       "      <th>align_length</th>\n",
       "      <th>identities</th>\n",
       "      <th>positives</th>\n",
       "      <th>gaps</th>\n",
       "      <th>query</th>\n",
       "      <th>sbjct</th>\n",
       "      <th>match</th>\n",
       "      <th>taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>917.0</td>\n",
       "      <td>828.131</td>\n",
       "      <td>461</td>\n",
       "      <td>460</td>\n",
       "      <td>460</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Uncultured bacterium clone ncd2437e12c2 16S ri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>917.0</td>\n",
       "      <td>828.131</td>\n",
       "      <td>461</td>\n",
       "      <td>460</td>\n",
       "      <td>460</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Uncultured bacterium clone ncd2286f10c1 16S ri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>917.0</td>\n",
       "      <td>828.131</td>\n",
       "      <td>461</td>\n",
       "      <td>460</td>\n",
       "      <td>460</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Uncultured bacterium clone ncd1245g01c1 16S ri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>912.0</td>\n",
       "      <td>823.622</td>\n",
       "      <td>461</td>\n",
       "      <td>459</td>\n",
       "      <td>459</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Uncultured bacterium clone ncd2787d04c1 16S ri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>912.0</td>\n",
       "      <td>823.622</td>\n",
       "      <td>461</td>\n",
       "      <td>459</td>\n",
       "      <td>459</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Uncultured bacterium clone ncd2554e10c1 16S ri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>788.457</td>\n",
       "      <td>439</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Uncultured organism clone ELU0143-T325-S-NIPCR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>788.457</td>\n",
       "      <td>439</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Levyella massiliensis strain 9403326 16S ribos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>788.457</td>\n",
       "      <td>439</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Uncultured bacterium clone ncd2716a04c1 16S ri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>788.457</td>\n",
       "      <td>439</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Uncultured bacterium clone ncd2714d06c1 16S ri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>9186e853f994fd8285d68ccd223cd34d</td>\n",
       "      <td>0.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>788.457</td>\n",
       "      <td>439</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "      <td>Uncultured bacterium clone ncd2714g04c1 16S ri...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>354 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Feature ID  expect  score     bits  align_length  \\\n",
       "0    4900b14ef37cec718bf18a0d92210a03     0.0  917.0  828.131           461   \n",
       "1    4900b14ef37cec718bf18a0d92210a03     0.0  917.0  828.131           461   \n",
       "2    4900b14ef37cec718bf18a0d92210a03     0.0  917.0  828.131           461   \n",
       "3    4900b14ef37cec718bf18a0d92210a03     0.0  912.0  823.622           461   \n",
       "4    4900b14ef37cec718bf18a0d92210a03     0.0  912.0  823.622           461   \n",
       "..                                ...     ...    ...      ...           ...   \n",
       "349  9186e853f994fd8285d68ccd223cd34d     0.0  873.0  788.457           439   \n",
       "350  9186e853f994fd8285d68ccd223cd34d     0.0  873.0  788.457           439   \n",
       "351  9186e853f994fd8285d68ccd223cd34d     0.0  873.0  788.457           439   \n",
       "352  9186e853f994fd8285d68ccd223cd34d     0.0  873.0  788.457           439   \n",
       "353  9186e853f994fd8285d68ccd223cd34d     0.0  873.0  788.457           439   \n",
       "\n",
       "     identities  positives  gaps  \\\n",
       "0           460        460     0   \n",
       "1           460        460     0   \n",
       "2           460        460     0   \n",
       "3           459        459     0   \n",
       "4           459        459     0   \n",
       "..          ...        ...   ...   \n",
       "349         438        438     0   \n",
       "350         438        438     0   \n",
       "351         438        438     0   \n",
       "352         438        438     0   \n",
       "353         438        438     0   \n",
       "\n",
       "                                                 query  \\\n",
       "0    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "1    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "2    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "3    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "4    GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "..                                                 ...   \n",
       "349  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "350  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "351  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "352  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "353  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "\n",
       "                                                 sbjct  \\\n",
       "0    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "1    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "2    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "3    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "4    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "..                                                 ...   \n",
       "349  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "350  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "351  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "352  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "353  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "\n",
       "                                                 match  \\\n",
       "0    |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "1    |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "2    |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "3    |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "4    |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "..                                                 ...   \n",
       "349  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "350  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "351  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "352  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "353  |||||| |||||||||||||||||||||||||||||||||||||||...   \n",
       "\n",
       "                                              taxonomy  \n",
       "0    Uncultured bacterium clone ncd2437e12c2 16S ri...  \n",
       "1    Uncultured bacterium clone ncd2286f10c1 16S ri...  \n",
       "2    Uncultured bacterium clone ncd1245g01c1 16S ri...  \n",
       "3    Uncultured bacterium clone ncd2787d04c1 16S ri...  \n",
       "4    Uncultured bacterium clone ncd2554e10c1 16S ri...  \n",
       "..                                                 ...  \n",
       "349  Uncultured organism clone ELU0143-T325-S-NIPCR...  \n",
       "350  Levyella massiliensis strain 9403326 16S ribos...  \n",
       "351  Uncultured bacterium clone ncd2716a04c1 16S ri...  \n",
       "352  Uncultured bacterium clone ncd2714d06c1 16S ri...  \n",
       "353  Uncultured bacterium clone ncd2714g04c1 16S ri...  \n",
       "\n",
       "[354 rows x 12 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def blast_classifier(rep_seqs_path, taxonomy_path, top_n=10):\n",
    "    # load in files\n",
    "    rep_seqs_qza = Artifact.load(rep_seqs_path) \n",
    "    taxonomy_qza = Artifact.load(taxonomy_path)\n",
    "\n",
    "    # convert taxonomy to df\n",
    "    taxonomy_df = taxonomy_qza.view(pd.DataFrame)\n",
    "\n",
    "    # convert req seqs to df\n",
    "    rep_seqs = rep_seqs_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "    rep_seqs.index.name = 'Feature ID'\n",
    "\n",
    "    # merge dfs on Feature ID\n",
    "    merged = pd.merge(rep_seqs, taxonomy_df, left_index=True, right_index=True)\n",
    "    merged.reset_index(inplace=True)\n",
    "\n",
    "    # filter for entries without species assignment\n",
    "    unassigned = merged[merged['Taxon'].str.endswith('s__')]\n",
    "\n",
    "    # empty list for BLAST results\n",
    "    blast_results = []\n",
    "\n",
    "    # perform BLAST for each unassigned sequencec\n",
    "    for _, row in unassigned.iterrows():\n",
    "        feature_id = row['Feature ID']\n",
    "        sequence = row['Sequence']\n",
    "        \n",
    "        # BLAST search\n",
    "        result_handle = NCBIWWW.qblast('blastn', 'nt', sequence)\n",
    "        blast_record = NCBIXML.read(result_handle)\n",
    "        if blast_record.alignments:\n",
    "            for alignment in blast_record.alignments[:top_n]:\n",
    "                for hsp in alignment.hsps:\n",
    "                    extracted_title = alignment.title.rsplit('|', 1)[-1].strip()\n",
    "                    blast_results.append({\n",
    "                        'Feature ID': feature_id,\n",
    "                        'expect': hsp.expect,\n",
    "                        'score': hsp.score,\n",
    "                        'bits': hsp.bits, \n",
    "                        'align_length': hsp.align_length,\n",
    "                        'identities': hsp.identities, \n",
    "                        'positives': hsp.positives,\n",
    "                        'gaps': hsp.gaps,\n",
    "                        'query': hsp.query,\n",
    "                        'sbjct': hsp.sbjct,\n",
    "                        'match': hsp.match,\n",
    "                        'taxonomy': extracted_title\n",
    "                    })\n",
    "                time.sleep(1)\n",
    "\n",
    "    blast_results_df = pd.DataFrame(blast_results)\n",
    "\n",
    "    return blast_results_df\n",
    "\n",
    "# file paths\n",
    "rep_seqs_path = 'qiime2_blast_data/SRA1038019-rep_seqs.qza'\n",
    "taxonomy_path = 'qiime2_blast_data/SRA1038019-taxonomy.qza'\n",
    "\n",
    "# Load the taxonomy artifact\n",
    "#taxonomy_qza_path = 'qiime2_blast_data/ERA1115542-taxonomy.qza'\n",
    "#taxonomy_qza = Artifact.load(taxonomy_qza_path)\n",
    "\n",
    "# Convert the taxonomy artifact to a DataFrame\n",
    "#taxonomy_df = taxonomy_qza.view(pd.DataFrame)\n",
    "\n",
    "# Display the first few rows of the taxonomy DataFrame\n",
    "#taxonomy_df.head()\n",
    "\n",
    "# Display the first few rows of the taxonomy DataFrame\n",
    "#print(taxonomy_df.head())\n",
    "\n",
    "# Save the DataFrame to a CSV file for further inspection if needed\n",
    "#taxonomy_df.to_csv('taxonomy.csv', index=False)\n",
    "\n",
    "# Perform BLAST classification and get results\n",
    "results_df = blast_classifier(rep_seqs_path, taxonomy_path) \n",
    "\n",
    "# Display the results\n",
    "results_df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature ID</th>\n",
       "      <th>expect</th>\n",
       "      <th>taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4900b14ef37cec718bf18a0d92210a03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Uncultured bacterium clone ncd2437e12c2 16S ri...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Feature ID  expect  \\\n",
       "0  4900b14ef37cec718bf18a0d92210a03     0.0   \n",
       "\n",
       "                                            taxonomy  \n",
       "0  Uncultured bacterium clone ncd2437e12c2 16S ri...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_best_blast_hits(results_df, expect_threshold=0.001):\n",
    "    best_hits = []\n",
    "\n",
    "    grouped = results_df.groupby('Feature ID')\n",
    "    for feature_id, group in grouped:\n",
    "        # Filter by expect threshold\n",
    "        filtered_group = group[group['expect'] <= expect_threshold]\n",
    "        if not filtered_group.empty:\n",
    "            # Select the best hit with the lowest expect value, then highest bit score\n",
    "            best_hit = filtered_group.loc[filtered_group.sort_values(['expect', 'bits'], ascending=[True, False]).index[0]]\n",
    "            best_hits.append(best_hit)\n",
    "\n",
    "    best_hits_df = pd.DataFrame(best_hits)\n",
    "    return best_hits_df\n",
    "\n",
    "best_hits_df = get_best_blast_hits(results_df, expect_threshold=0.001)\n",
    "\n",
    "# display the best hits\n",
    "best_hits_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gi|322196871|gb|JF211466.1| Uncultured bacterium clone ncd2437e12c2 16S ribosomal RNA gene, partial sequence\n"
     ]
    }
   ],
   "source": [
    "print (results_df['taxonomy'].iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Error message from NCBI: Message ID#24 Error: Failed to read the Blast query: Protein FASTA provided for nucleotide sequence",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-1c11c967664d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mfeature_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequence\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrep_seqs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m     \u001b[0mtaxonomy\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_blast_taxonomy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mtaxonomy\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m         \u001b[0motu_list\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfeature_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-6-1c11c967664d>\u001b[0m in \u001b[0;36mget_blast_taxonomy\u001b[0;34m(sequence, expect_threshold)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mget_blast_taxonomy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpect_threshold\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.001\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mresult_handle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mNCBIWWW\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqblast\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'blastn'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'nt'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0mblast_record\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mNCBIXML\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult_handle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0malignment\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mblast_record\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malignments\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mhsp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0malignment\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhsps\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qiime2/lib/python3.6/site-packages/Bio/Blast/NCBIWWW.py\u001b[0m in \u001b[0;36mqblast\u001b[0;34m(program, database, sequence, url_base, auto_format, composition_based_statistics, db_genetic_code, endpoints, entrez_query, expect, filter, gapcosts, genetic_code, hitlist_size, i_thresh, layout, lcase_mask, matrix_name, nucl_penalty, nucl_reward, other_advanced, perc_ident, phi_pattern, query_file, query_believe_defline, query_from, query_to, searchsp_eff, service, threshold, ungapped_alignment, word_size, short_query, alignments, alignment_view, descriptions, entrez_links_new_window, expect_low, expect_high, format_entrez_query, format_object, format_type, ncbi_gi, results_file, show_overview, megablast, template_type, template_length)\u001b[0m\n\u001b[1;32m    206\u001b[0m     \u001b[0;31m# Format the \"Get\" command, which gets the formatted results from qblast\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    207\u001b[0m     \u001b[0;31m# Parameters taken from http://www.ncbi.nlm.nih.gov/BLAST/Doc/node6.html on 9 July 2007\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 208\u001b[0;31m     \u001b[0mrid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrtoe\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_parse_qblast_ref_page\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    209\u001b[0m     parameters = [\n\u001b[1;32m    210\u001b[0m         \u001b[0;34m(\u001b[0m\u001b[0;34m\"ALIGNMENTS\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malignments\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qiime2/lib/python3.6/site-packages/Bio/Blast/NCBIWWW.py\u001b[0m in \u001b[0;36m_parse_qblast_ref_page\u001b[0;34m(handle)\u001b[0m\n\u001b[1;32m    316\u001b[0m             \u001b[0mmsg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"</p>\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\\n\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    317\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 318\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Error message from NCBI: %s\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    319\u001b[0m         \u001b[0;31m# Generic search based on the way the error messages start:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    320\u001b[0m         \u001b[0mi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0ms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Message ID#\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Error message from NCBI: Message ID#24 Error: Failed to read the Blast query: Protein FASTA provided for nucleotide sequence"
     ]
    }
   ],
   "source": [
    "#downloading reference sequences and taxonomy\n",
    "!wget -O \"85_otus.fasta\" \"https://data.qiime2.org/2024.5/tutorials/training-feature-classifiers/85_otus.fasta\"\n",
    "!wget -O \"85_otu_taxonomy.txt\" \"https://data.qiime2.org/2024.5/tutorials/training-feature-classifiers/85_otu_taxonomy.txt\"\n",
    "\n",
    "#importing into QIIME 2 artifact\n",
    "!qiime tools import \\\n",
    "  --type 'FeatureData[Sequence]' \\\n",
    "  --input-path 85_otus.fasta \\\n",
    "  --output-path 85_otus.qza\n",
    "\n",
    "#mport the downloaded taxonomy into a QIIME 2 artifact\n",
    "!qiime tools import \\\n",
    "  --type 'FeatureData[Taxonomy]' \\\n",
    "  --input-format HeaderlessTSVTaxonomyFormat \\\n",
    "  --input-path 85_otu_taxonomy.txt \\\n",
    "  --output-path 85_otu_taxonomy.qza\n",
    "\n",
    "#verify the imports by viewing the artifacts\n",
    "!qiime tools peek 85_otus.qza\n",
    "!qiime tools peek 85_otu_taxonomy.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-8-d3da90829954>, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-8-d3da90829954>\"\u001b[0;36m, line \u001b[0;32m3\u001b[0m\n\u001b[0;31m    from qiime tools import \\\u001b[0m\n\u001b[0m                   ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "#load the existing database taxonomy\n",
    "existing_taxonomy_path = '85_otu_taxonomy.qza'\n",
    "existing_sequences_path = '85_otus.qza'\n",
    "existing_taxonomy_qza = Artifact.load(existing_taxonomy_path)\n",
    "existing_sequences_qza = Artifact.load(existing_sequences_path)\n",
    "\n",
    "#convert to df\n",
    "existing_taxonomy_df = existing_taxonomy_qza.view(pd.DataFrame)\n",
    "existing_sequences_df = existing_sequences_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "existing_sequences_df.index.name = 'Feature ID'\n",
    "\n",
    "best_hits_df = pd.read_csv('best_blast_hits.csv')\n",
    "\n",
    "#combine taxonomy\n",
    "best_hits_df = best_hits_df[['Feature ID', 'taxonomy']]\n",
    "best_hits_df.columns = ['Feature ID', 'Taxon']  # Rename for consistency\n",
    "combined_taxonomy_df = pd.concat([existing_taxonomy_df, best_hits_df], ignore_index=True)\n",
    "combined_taxonomy_df.to_csv('combined_taxonomy.csv', index=False)\n",
    "\n",
    "#combine sequences\n",
    "combined_sequences_df = pd.concat([existing_sequences_df, best_hits_df[['Feature ID', 'Sequence']]], ignore_index=False)\n",
    "combined_sequences_df.to_csv('combined_sequences.fasta', index=True, header=False)\n",
    "\n",
    "\n",
    "#import combined taxonomy\n",
    "!qiime tools import \\\n",
    "  --type 'FeatureData[Taxonomy]' \\\n",
    "  --input-format HeaderlessTSVTaxonomyFormat \\\n",
    "  --input-path combined_taxonomy.csv \\\n",
    "  --output-path combined_taxonomy.qza\n",
    "\n",
    "#import combined sequences\n",
    "!qiime tools import \\\n",
    "  --type 'FeatureData[Sequence]' \\\n",
    "  --input-path combined_sequences.fasta \\\n",
    "  --output-path combined_sequences.qza\n",
    "\n",
    "#extract reference reads\n",
    "!qiime feature-classifier extract-reads \\\n",
    "  --i-sequences combined_sequences.qza \\\n",
    "  --p-f-primer GTGCCAGCMGCCGCGGTAA \\\n",
    "  --p-r-primer GGACTACHVGGGTWTCTAAT \\\n",
    "  --p-trunc-len 120 \\\n",
    "  --o-reads ref-seqs.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#training naive bayes classifier\n",
    "!qiime feature-classifier fit-classifier-naive-bayes \\\n",
    "  --i-reference-reads ref-seqs.qza \\\n",
    "  --i-reference-taxonomy combined_taxonomy.qza \\\n",
    "  --o-classifier classifier.qza"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2",
   "language": "python",
   "name": "qiime2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
