{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from qiime2 import Artifact\n",
    "import pandas as pd\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCCTCAGTGTCAGACATACCTTTGTGAGCTGCCTTCGCAATCGGAGTTCTGCGTAATATCTATGCATTTCACCGCTACACTACGCATTCCGCCCACATCATGTACTCTCAAGTCTTACAGTTTCAACGGCAAGGCTGAGGTTGAGCCTCAGCTTTTCACCACTGACTTGCAAGACCACCTGCGCACCCTTTAAACCCAATAAATCCGGATAACGCTCGTATCCTCCGTATTACCGCGGCTGCTGGCACGGAGTTAGCCGATACTTATTCTCGAGCTACTTTCAAACTCCCATCACGTGGAAGTCCTTACTCGCTCGCTAAAGAAGTTTACAATCCTTAGGACATTCTTCCTTCACGCGACTTGGCTGGTTCAGAGTTTCCTCCATTGACCAATATTCCTCACTGCTGCCTCCCGTAGG\n"
     ]
    }
   ],
   "source": [
    "rep_seqs_qza = Artifact.load('qiime2_blast_data/SRA1038019-rep_seqs.qza') \n",
    "taxonomy_qza = Artifact.load('qiime2_blast_data/SRA1038019-taxonomy.qza')\n",
    "\n",
    "rep_seqs = rep_seqs_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "rep_seqs.index.name = 'Feature ID'\n",
    "tax = taxonomy_qza.view(pd.Series)\n",
    "tax.index.name = 'Feature ID' \n",
    "\n",
    "merged = pd.merge(rep_seqs, tax, left_index=True, right_index=True)\n",
    "merged.reset_index(inplace=True)\n",
    "unassigned = merged[merged['Taxon'].str.endswith('s__')]\n",
    "unique_values_count = unassigned.iloc[:, 0].nunique()\n",
    "first_sequence = unassigned['Sequence'].iloc[0]\n",
    "print (first_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature ID</th>\n",
       "      <th>expect</th>\n",
       "      <th>score</th>\n",
       "      <th>bits</th>\n",
       "      <th>align_length</th>\n",
       "      <th>identities</th>\n",
       "      <th>positives</th>\n",
       "      <th>gaps</th>\n",
       "      <th>query</th>\n",
       "      <th>sbjct</th>\n",
       "      <th>match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>917.0</td>\n",
       "      <td>828.131</td>\n",
       "      <td>461</td>\n",
       "      <td>460</td>\n",
       "      <td>460</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>917.0</td>\n",
       "      <td>828.131</td>\n",
       "      <td>461</td>\n",
       "      <td>460</td>\n",
       "      <td>460</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>917.0</td>\n",
       "      <td>828.131</td>\n",
       "      <td>461</td>\n",
       "      <td>460</td>\n",
       "      <td>460</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>912.0</td>\n",
       "      <td>823.622</td>\n",
       "      <td>461</td>\n",
       "      <td>459</td>\n",
       "      <td>459</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>912.0</td>\n",
       "      <td>823.622</td>\n",
       "      <td>461</td>\n",
       "      <td>459</td>\n",
       "      <td>459</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>53</td>\n",
       "      <td>0.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>788.457</td>\n",
       "      <td>439</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>53</td>\n",
       "      <td>0.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>788.457</td>\n",
       "      <td>439</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>341</th>\n",
       "      <td>53</td>\n",
       "      <td>0.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>788.457</td>\n",
       "      <td>439</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>342</th>\n",
       "      <td>53</td>\n",
       "      <td>0.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>788.457</td>\n",
       "      <td>439</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>53</td>\n",
       "      <td>0.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>788.457</td>\n",
       "      <td>439</td>\n",
       "      <td>438</td>\n",
       "      <td>438</td>\n",
       "      <td>0</td>\n",
       "      <td>GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...</td>\n",
       "      <td>|||||| |||||||||||||||||||||||||||||||||||||||...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>344 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Feature ID  expect  score     bits  align_length  identities  positives  \\\n",
       "0             3     0.0  917.0  828.131           461         460        460   \n",
       "1             3     0.0  917.0  828.131           461         460        460   \n",
       "2             3     0.0  917.0  828.131           461         460        460   \n",
       "3             3     0.0  912.0  823.622           461         459        459   \n",
       "4             3     0.0  912.0  823.622           461         459        459   \n",
       "..          ...     ...    ...      ...           ...         ...        ...   \n",
       "339          53     0.0  873.0  788.457           439         438        438   \n",
       "340          53     0.0  873.0  788.457           439         438        438   \n",
       "341          53     0.0  873.0  788.457           439         438        438   \n",
       "342          53     0.0  873.0  788.457           439         438        438   \n",
       "343          53     0.0  873.0  788.457           439         438        438   \n",
       "\n",
       "     gaps                                              query  \\\n",
       "0       0  GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "1       0  GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "2       0  GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "3       0  GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "4       0  GACTACAAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "..    ...                                                ...   \n",
       "339     0  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "340     0  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "341     0  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "342     0  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "343     0  GACTACAAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "\n",
       "                                                 sbjct  \\\n",
       "0    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "1    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "2    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "3    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "4    GACTACCAGGGTATCTAATCCTGTTCGCTCCCCACGCTTTCGTGCC...   \n",
       "..                                                 ...   \n",
       "339  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "340  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "341  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "342  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "343  GACTACCAGGGTATCTAATCCTGTTTGCTCCCCACGCTTTCGTACC...   \n",
       "\n",
       "                                                 match  \n",
       "0    |||||| |||||||||||||||||||||||||||||||||||||||...  \n",
       "1    |||||| |||||||||||||||||||||||||||||||||||||||...  \n",
       "2    |||||| |||||||||||||||||||||||||||||||||||||||...  \n",
       "3    |||||| |||||||||||||||||||||||||||||||||||||||...  \n",
       "4    |||||| |||||||||||||||||||||||||||||||||||||||...  \n",
       "..                                                 ...  \n",
       "339  |||||| |||||||||||||||||||||||||||||||||||||||...  \n",
       "340  |||||| |||||||||||||||||||||||||||||||||||||||...  \n",
       "341  |||||| |||||||||||||||||||||||||||||||||||||||...  \n",
       "342  |||||| |||||||||||||||||||||||||||||||||||||||...  \n",
       "343  |||||| |||||||||||||||||||||||||||||||||||||||...  \n",
       "\n",
       "[344 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def blast_classifier(rep_seqs_path, taxonomy_path, top_n=10):\n",
    "    '''\n",
    "    This classifier classifies the sequences by performing BLAST search on the unassigned sequences.\n",
    "    \n",
    "    Parameters:\n",
    "    rep_seqs_path (str): path to the representative sequences file\n",
    "    taxonomy_path (str): path to the taxonomy file\n",
    "    top_n (int): number of top alignments to consider each BLAST search\n",
    "     max_results (int): number of unassigned sequences to process\n",
    "      \n",
    "    Returns:\n",
    "    pd.DataFrame: DataFrame containing BLAST results.  \n",
    "    '''\n",
    "    # load in files\n",
    "    rep_seqs_qza = Artifact.load(rep_seqs_path) \n",
    "    taxonomy_qza = Artifact.load(taxonomy_path)\n",
    "\n",
    "    # convert taxonomy to df\n",
    "    taxonomy_df = taxonomy_qza.view(pd.DataFrame)\n",
    "\n",
    "    # convert req seqs to df\n",
    "    rep_seqs = rep_seqs_qza.view(pd.Series).to_frame(name='Sequence')\n",
    "    rep_seqs.index.name = 'Feature ID'\n",
    "\n",
    "    # merge dfs on Feature ID\n",
    "    merged = pd.merge(rep_seqs, taxonomy_df, left_index=True, right_index=True)\n",
    "    merged.reset_index(inplace=True)\n",
    "\n",
    "    # filter for entries without species assignment\n",
    "    unassigned = merged[merged['Taxon'].str.endswith('s__')]\n",
    "\n",
    "    # empty list for BLAST results\n",
    "    blast_results = []\n",
    "\n",
    "    # perform BLAST for each unassigned sequencec\n",
    "    for feature_id, row in unassigned.iterrows():\n",
    "        sequence = row['Sequence']\n",
    "        \n",
    "        # BLAST search\n",
    "        result_handle = NCBIWWW.qblast('blastn', 'nt', sequence)\n",
    "        blast_record = NCBIXML.read(result_handle)\n",
    "        if blast_record.alignments:\n",
    "            for alignment in blast_record.alignments[:top_n]:\n",
    "                for hsp in alignment.hsps:\n",
    "                    #taxonomy = row['Taxon']\n",
    "                    blast_results.append({\n",
    "                        'Feature ID': feature_id,\n",
    "                        'expect': hsp.expect,\n",
    "                        'score': hsp.score,\n",
    "                        'bits': hsp.bits, \n",
    "                        'align_length': hsp.align_length,\n",
    "                        'identities': hsp.identities, \n",
    "                        'positives': hsp.positives,\n",
    "                        'gaps': hsp.gaps,\n",
    "                        'query': hsp.query,\n",
    "                        'sbjct': hsp.sbjct,\n",
    "                        'match': hsp.match,\n",
    "                        'staxids': hsp.staxids\n",
    "                        #'taxonomy': taxonomy  \n",
    "                    })\n",
    "                time.sleep(1)\n",
    "\n",
    "    blast_results_df = pd.DataFrame(blast_results)\n",
    "\n",
    "    return blast_results_df\n",
    "\n",
    "# file paths\n",
    "rep_seqs_path = 'qiime2_blast_data/SRA1038019-rep_seqs.qza'\n",
    "taxonomy_path = 'qiime2_blast_data/SRA1038019-taxonomy.qza'\n",
    "\n",
    "# Load the taxonomy artifact\n",
    "#taxonomy_qza_path = 'qiime2_blast_data/ERA1115542-taxonomy.qza'\n",
    "#taxonomy_qza = Artifact.load(taxonomy_qza_path)\n",
    "\n",
    "# Convert the taxonomy artifact to a DataFrame\n",
    "#taxonomy_df = taxonomy_qza.view(pd.DataFrame)\n",
    "\n",
    "# Display the first few rows of the taxonomy DataFrame\n",
    "#taxonomy_df.head()\n",
    "\n",
    "# Display the first few rows of the taxonomy DataFrame\n",
    "#print(taxonomy_df.head())\n",
    "\n",
    "# Save the DataFrame to a CSV file for further inspection if needed\n",
    "#taxonomy_df.to_csv('taxonomy.csv', index=False)\n",
    "\n",
    "# Perform BLAST classification and get results\n",
    "results_df = blast_classifier(rep_seqs_path, taxonomy_path) \n",
    "\n",
    "# Display the results\n",
    "results_df\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'results_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-bf89044b1c7a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mmin_expect\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mresults_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'expect'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mmin_expect\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'results_df' is not defined"
     ]
    }
   ],
   "source": [
    "min_expect = results_df['expect'].min()\n",
    "min_expect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Error message from NCBI: Message ID#24 Error: Failed to read the Blast query: Protein FASTA provided for nucleotide sequence",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-1c11c967664d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mfeature_id\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequence\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrep_seqs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m     \u001b[0mtaxonomy\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_blast_taxonomy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mtaxonomy\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m         \u001b[0motu_list\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfeature_id\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-6-1c11c967664d>\u001b[0m in \u001b[0;36mget_blast_taxonomy\u001b[0;34m(sequence, expect_threshold)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mget_blast_taxonomy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexpect_threshold\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0.001\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mresult_handle\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mNCBIWWW\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqblast\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'blastn'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'nt'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m     \u001b[0mblast_record\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mNCBIXML\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult_handle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0malignment\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mblast_record\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0malignments\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mhsp\u001b[0m \u001b[0;32min\u001b[0m \u001b[0malignment\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhsps\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qiime2/lib/python3.6/site-packages/Bio/Blast/NCBIWWW.py\u001b[0m in \u001b[0;36mqblast\u001b[0;34m(program, database, sequence, url_base, auto_format, composition_based_statistics, db_genetic_code, endpoints, entrez_query, expect, filter, gapcosts, genetic_code, hitlist_size, i_thresh, layout, lcase_mask, matrix_name, nucl_penalty, nucl_reward, other_advanced, perc_ident, phi_pattern, query_file, query_believe_defline, query_from, query_to, searchsp_eff, service, threshold, ungapped_alignment, word_size, short_query, alignments, alignment_view, descriptions, entrez_links_new_window, expect_low, expect_high, format_entrez_query, format_object, format_type, ncbi_gi, results_file, show_overview, megablast, template_type, template_length)\u001b[0m\n\u001b[1;32m    206\u001b[0m     \u001b[0;31m# Format the \"Get\" command, which gets the formatted results from qblast\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    207\u001b[0m     \u001b[0;31m# Parameters taken from http://www.ncbi.nlm.nih.gov/BLAST/Doc/node6.html on 9 July 2007\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 208\u001b[0;31m     \u001b[0mrid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mrtoe\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_parse_qblast_ref_page\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhandle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    209\u001b[0m     parameters = [\n\u001b[1;32m    210\u001b[0m         \u001b[0;34m(\u001b[0m\u001b[0;34m\"ALIGNMENTS\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0malignments\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/qiime2/lib/python3.6/site-packages/Bio/Blast/NCBIWWW.py\u001b[0m in \u001b[0;36m_parse_qblast_ref_page\u001b[0;34m(handle)\u001b[0m\n\u001b[1;32m    316\u001b[0m             \u001b[0mmsg\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"</p>\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\\n\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    317\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 318\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Error message from NCBI: %s\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    319\u001b[0m         \u001b[0;31m# Generic search based on the way the error messages start:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    320\u001b[0m         \u001b[0mi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0ms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Message ID#\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Error message from NCBI: Message ID#24 Error: Failed to read the Blast query: Protein FASTA provided for nucleotide sequence"
     ]
    }
   ],
   "source": [
    "def get_blast_taxonomy(sequence, expect_threshold=0.001):\n",
    "    result_handle = NCBIWWW.qblast('blastn', 'nt', sequence)\n",
    "    blast_record = NCBIXML.read(result_handle)\n",
    "    for alignment in blast_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            if hsp.expect < expect_threshold:\n",
    "                taxonomy = f\"k__{alignment.hit_def.split(';')[0]}; p__{alignment.hit_def.split(';')[1]}; \" \\\n",
    "                           f\"c__{alignment.hit_def.split(';')[2]}; o__{alignment.hit_def.split(';')[3]}; \" \\\n",
    "                           f\"f__{alignment.hit_def.split(';')[4]}; g__{alignment.hit_def.split(';')[5]}; \" \\\n",
    "                           f\"s__{alignment.hit_def.split(';')[6]}\"\n",
    "                return taxonomy\n",
    "    return None\n",
    "\n",
    "# perform BLAST for each sequence and create reference database files\n",
    "otu_list = []\n",
    "taxonomy_list = []\n",
    "\n",
    "for feature_id, sequence in rep_seqs.items():\n",
    "    taxonomy = get_blast_taxonomy(sequence)\n",
    "    if taxonomy:\n",
    "        otu_list.append(feature_id)\n",
    "        taxonomy_list.append(taxonomy)\n",
    "\n",
    "# df for OTUs and Taxonomy\n",
    "otu_df = pd.DataFrame({'Feature ID': otu_list, 'Taxonomy': taxonomy_list})\n",
    "otu_artifact = Artifact.import_data('FeatureData[Sequence]', otu_df)\n",
    "taxonomy_artifact = Artifact.import_data('FeatureData[Taxonomy]', otu_df)\n",
    "\n",
    "# save artifacts\n",
    "otu_artifact.save('otus.qza')\n",
    "taxonomy_artifact.save('taxonomy.qza')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-8-d3da90829954>, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-8-d3da90829954>\"\u001b[0;36m, line \u001b[0;32m3\u001b[0m\n\u001b[0;31m    from qiime tools import \\\u001b[0m\n\u001b[0m                   ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "#naive bayes\n",
    "# import necessary plugins\n",
    "mkdir training-feature-classifiers\n",
    "cd training-feature-classifiers\n",
    "\n",
    "qiime tools import \\\n",
    "  --type 'FeatureData[Sequence]' \\\n",
    "  --input-path otus.qza \\\n",
    "  --output-path ref-seqs.qza\n",
    "\n",
    "qiime tools import \\\n",
    "  --type 'FeatureData[Taxonomy]' \\\n",
    "  --input-path taxonomy.qza \\\n",
    "  --output-path ref-taxonomy.qza\n",
    "\n",
    "# Train the classifier\n",
    "qiime feature-classifier fit-classifier-naive-bayes \\\n",
    "  --i-reference-reads ref-seqs.qza \\\n",
    "  --i-reference-taxonomy ref-taxonomy.qza \\\n",
    "  --o-classifier classifier.qza\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qiime2",
   "language": "python",
   "name": "qiime2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
